<!doctype html>
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="generator" content="Hugo 0.80.0" /><META NAME="ROBOTS" CONTENT="INDEX, FOLLOW">


<link rel="shortcut icon" href="../../../favicons/favicon.ico" >
<link rel="apple-touch-icon" href="../../../favicons/apple-touch-icon-180x180.png" sizes="180x180">
<link rel="icon" type="image/png" href="../../../favicons/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="../../../favicons/favicon-32x32.png" sizes="32x32">
<link rel="icon" type="image/png" href="../../../favicons/android-36x36.png" sizes="36x36">
<link rel="icon" type="image/png" href="../../../favicons/android-48x48.png" sizes="48x48">
<link rel="icon" type="image/png" href="../../../favicons/android-72x72.png" sizes="72x72">
<link rel="icon" type="image/png" href="../../../favicons/android-96x96.png" sizes="96x96">
<link rel="icon" type="image/png" href="../../../favicons/android-144x144.png" sizes="144x144">
<link rel="icon" type="image/png" href="../../../favicons/android-192x192.png" sizes="192x192">

<title>Programming in R | GEN242</title><meta property="og:title" content="Programming in R" />
<meta property="og:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Overview One of the main attractions of using the R (http://cran.at.r-project.org) environment is the ease with which users can write their own programs and custom functions. The R programming syntax is extremely easy to learn, even for users with no previous programming experience. Once the basic R programming control structures are understood, users can use the R language as a powerful environment to perform complex custom analyses of almost any type of data (Gentleman 2008)." />
<meta property="og:type" content="article" />
<meta property="og:url" content="/manuals/rprogramming/rprogramming/" />
<meta property="article:modified_time" content="2021-05-01T15:18:21-07:00" /><meta property="og:site_name" content="GEN242" />
<meta itemprop="name" content="Programming in R">
<meta itemprop="description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Overview One of the main attractions of using the R (http://cran.at.r-project.org) environment is the ease with which users can write their own programs and custom functions. The R programming syntax is extremely easy to learn, even for users with no previous programming experience. Once the basic R programming control structures are understood, users can use the R language as a powerful environment to perform complex custom analyses of almost any type of data (Gentleman 2008).">
<meta itemprop="dateModified" content="2021-05-01T15:18:21-07:00" />
<meta itemprop="wordCount" content="3700">



<meta itemprop="keywords" content="" />
<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Programming in R"/>
<meta name="twitter:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Overview One of the main attractions of using the R (http://cran.at.r-project.org) environment is the ease with which users can write their own programs and custom functions. The R programming syntax is extremely easy to learn, even for users with no previous programming experience. Once the basic R programming control structures are understood, users can use the R language as a powerful environment to perform complex custom analyses of almost any type of data (Gentleman 2008)."/>


<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-00000000-0', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>






<link rel="preload" href="../../../scss/main.min.1f416e6a5f646f231d085681bf2b4ebdad5d41f149a3977625affc97f735ab7e.css" as="style">
<link href="../../../scss/main.min.1f416e6a5f646f231d085681bf2b4ebdad5d41f149a3977625affc97f735ab7e.css" rel="stylesheet" integrity="">


<script
  src="https://code.jquery.com/jquery-3.5.1.min.js"
  integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0="
  crossorigin="anonymous"></script>

<script
  src="https://unpkg.com/lunr@2.3.8/lunr.min.js"
  integrity="sha384-vRQ9bDyE0Wnu+lMfm57BlYLO0/XauFuKpVsZPs7KEDwYKktWi5+Kz3MP8++DFlRY"
  crossorigin="anonymous"></script>



<link rel="stylesheet" href="../../../css/prism.css"/>


<script src="../../../js/driver0.9.8.min.js"></script>
<link rel="stylesheet" href="../../../css/driver0.9.8.min.css"/>


<link rel="stylesheet" href="../../../css/site.css"/>
<script src="../../../js/site.js"></script>



  </head>
  <body class="td-page line-numbers">
    <header>
     
<nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar">
  <a class="navbar-brand" href="../../../">
		<span class="font-weight-bold">GEN242</span>
	</a>
	<row class="header-switch">
	  	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="sidebarSwitch" checked=true>
        <label class="custom-control-label" for="sidebarSwitch">sidebar</label>
      </div>
    	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="tocSwitch" checked=true>
        <label class="custom-control-label" for="tocSwitch">toc</label>
      </div>
	</row>
	<div class="td-navbar-nav-scroll ml-md-auto" id="main_navbar">
		<ul class="navbar-nav mt-2 mt-lg-0">
			
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../about/" id="navbarDropdown" role="button">
          	About
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/introduction/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/syllabus/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Syllabus
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/schedule/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Schedule
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/internal/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Internal
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/internal_resources/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Resources
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/install/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Installation
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../slides/" id="navbarDropdown" role="button">
          	Slides
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_01"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Course Introduction (S1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_02"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Genome Basics (S2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_03"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Databases and Software (S3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_04"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 DNA Sequencing (S4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_05"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (S5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_06"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sequence Alignments (S6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_07"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (S7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_08"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Multiple Alignments (S8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_09"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Short Read Alignments (S9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_10"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R on HPC (S10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_11"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (S11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_12"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Gene Expression Analysis (S12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_13"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Workflows (S13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_14/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Overview (S14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_15/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Enrichment Analysis (S15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_16/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis (S16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_17/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics in R (S17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_18/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R Packages and dplyr (S18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_19/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Shiny Apps (S19)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_20/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Profile HMMs (S20)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_21/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sequence Assembly (S21)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../tutorials/" id="navbarDropdown" role="button">
          	Tutorials
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/github/github/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Introduction (T1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/linux/linux/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux and HPCC Cluster (T2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rbasics/rbasics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (T3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rprogramming/rprogramming/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (T4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rparallel/rparallel/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Parallel R (T5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rsequences/rsequences/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (T6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/systempiper/systempiper/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflows (T7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/sprnaseq/sprnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 RNA-Seq Workflow (T8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/spchipseq/spchipseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Workflow (T9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rmarkdown/rmarkdown/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R Markdown (T10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rfea/rfea/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Functional Enrichment Analysis (T11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rclustering/rclustering/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis and Data Mining (T12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (T13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rgraphics/rgraphics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics and Visualization (T14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/shinyapps/shinyapps/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Shiny Apps (T15)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../manuals/" id="navbarDropdown" role="button">
          	Manuals
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/github/github/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/linux/linux/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux &amp; HPCC Cluster
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rbasics/rbasics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rbasics/sample/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sample R Markdown
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rprogramming/rprogramming/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rparallel/rparallel/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Parallel Evaluations in R
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rsequences/rsequences/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/systempiper/systempiper/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Workflow Overview
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/sprnaseq/sprnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 RNA-Seq Workflow
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/spchipseq/spchipseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Workflow
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rfea/rfea/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Functional Enrichment Analysis
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rclustering/rclustering/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis and Data Mining
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rgraphics/rgraphics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics and Visualization
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../assignments/" id="navbarDropdown" role="button">
          	Assignments
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/homework/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Homework
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw01/hw01/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW1: Intro &amp; GitHub
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw02/hw02/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW2: Linux &amp; HPC
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw03/hw03/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW3: R Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw04/hw04/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW4: Alignments
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw05/hw05"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW5: R Programming
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw06/hw06"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW6: NGS Analysis Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw07/hw07"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW7: RNA-Seq Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw08/hw08"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW8: R Markdown
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_overview/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Overview of Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/01_rnaseq_aligners/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 RNA-Seq Aligners
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/02_rnaseq_deg/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEG Methods
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/03_cluster_analysis/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Cluster Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/04_scrnaseq_embedding/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 scRNA-Seq Embedding
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/05_chipseq_peakcaller/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 ChIP-Seq Peak Callers
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/06_functional_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Functional Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/07_motif_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Motif Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/08_genome_graphics/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Genome Graphics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Project Data
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Presentations
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/paper_presentations/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Paper Presentations
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../external_resources/" id="navbarDropdown" role="button">
          	Links
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="https://piazza.com/ucr/spring2021/gen242/home"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Piazza
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://classroom.github.com/classrooms/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Classroom
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://github.com/tgirke/GEN242"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Repo
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://bioconductor.org/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Bioconductor
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://www.dcassol.com/docsy_Rmarkdown/rmarkdown/install/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Hugo, Docsy &amp; R
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
				
				
					<a class="nav-link" href="../../../blog/" ><span>News</span></a>
				
			</li>
			
			
			
		</ul>
	</div>
	<div class="navbar-nav d-none d-lg-block">



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.18c6eeb6e844cb515c3d2f03590d5c67.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>

</div>
</nav>

    </header>
    <div class="container-fluid td-outer">
      <div class="td-main">
        <div class="row flex-xl-nowrap">
          <div class="col-12 col-md-3 col-xl-2 td-sidebar d-print-none">
            




<div id="td-sidebar-menu" class="td-sidebar__inner">
  
  <form class="td-sidebar__search d-flex align-items-center">
    



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.18c6eeb6e844cb515c3d2f03590d5c67.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>


    <button class="btn btn-link td-sidebar__toggle d-md-none p-0 ml-3 fas fa-bars" type="button" data-toggle="collapse" data-target="#td-section-nav" aria-controls="td-docs-nav" aria-expanded="false" aria-label="Toggle section navigation">
    </button>
  </form>
  
  <nav class="collapse td-sidebar-nav" id="td-section-nav">
    
    
    






<ul class="td-sidebar-nav__section pr-md-3">
  <li class="td-sidebar-nav__section-title">
    <a  href="../../../manuals/" class="align-left pl-0 pr-2 active td-sidebar-link td-sidebar-link__section">Manuals</a>
  </li>
  <ul>
    <li class="collapse show" id="manuals">
      
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsgithubgithub" href="../../../manuals/github/github/">GitHub</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualslinuxlinux" href="../../../manuals/linux/linux/">Linux &amp; HPCC Cluster</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrbasicsrbasics" href="../../../manuals/rbasics/rbasics/">Introduction to R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page  active" id="m-manualsrprogrammingrprogramming" href="../../../manuals/rprogramming/rprogramming/">Programming in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrbasicssample" href="../../../manuals/rbasics/sample/">Sample R Markdown</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrparallelrparallel" href="../../../manuals/rparallel/rparallel/">Parallel Evaluations in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrsequencesrsequences" href="../../../manuals/rsequences/rsequences/">NGS Analysis Basics</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualssystempipersystempiper" href="../../../manuals/systempiper/systempiper/">systemPipeR: Workflow design and reporting generation environment</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualssprnaseqsprnaseq" href="../../../manuals/sprnaseq/sprnaseq/">RNA-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsspchipseqspchipseq" href="../../../manuals/spchipseq/spchipseq/">ChIP-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrfearfea" href="../../../manuals/rfea/rfea/">Functional Enrichment Analysis</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrclusteringrclustering" href="../../../manuals/rclustering/rclustering/">Cluster Analysis in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrgraphicsrgraphics" href="../../../manuals/rgraphics/rgraphics/">Graphics and Data Visualization in R</a>
        
      
      
    </li>
  </ul>
</ul>

  </nav>
</div>




          </div>
          
          

          
            
          

          
          
              <div class="d-none d-xl-block col-xl-2 td-toc d-print-none">
                
  
  
  
  
  
  <div class="td-page-meta ml-2 pb-1 pt-2 mb-0">
  
    
    
      
    
    
    
    
    
    
    
    


    
    
      <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/manuals/rprogramming/rprogramming.markdown" target="_blank" class="source-link"><i class="fa fa-code fa-fw"></i> View Rmd source</a>
    

    <a href="https://github.com/tgirke/GEN242//edit/main/content/en/manuals/rprogramming/rprogramming.markdown" target="_blank"><i class="fa fa-edit fa-fw"></i> Edit this page</a>
    <a href="https://github.com/tgirke/GEN242//new/main/content/en/manuals/rprogramming/rprogramming.markdown?filename=change-me.md&amp;value=---%0Atitle%3A&#43;%22Long&#43;Page&#43;Title%22%0AlinkTitle%3A&#43;%22Short&#43;Nav&#43;Title%22%0Aweight%3A&#43;100%0Adescription%3A&#43;%3E-%0A&#43;&#43;&#43;&#43;&#43;Page&#43;description&#43;for&#43;heading&#43;and&#43;indexes.%0A---%0A%0A%23%23&#43;Heading%0A%0AEdit&#43;this&#43;template&#43;to&#43;create&#43;your&#43;new&#43;page.%0A%0A%2A&#43;Give&#43;it&#43;a&#43;good&#43;name%2C&#43;ending&#43;in&#43;%60.md%60&#43;-&#43;e.g.&#43;%60getting-started.md%60%0A%2A&#43;Edit&#43;the&#43;%22front&#43;matter%22&#43;section&#43;at&#43;the&#43;top&#43;of&#43;the&#43;page&#43;%28weight&#43;controls&#43;how&#43;its&#43;ordered&#43;amongst&#43;other&#43;pages&#43;in&#43;the&#43;same&#43;directory%3B&#43;lowest&#43;number&#43;first%29.%0A%2A&#43;Add&#43;a&#43;good&#43;commit&#43;message&#43;at&#43;the&#43;bottom&#43;of&#43;the&#43;page&#43;%28%3C80&#43;characters%3B&#43;use&#43;the&#43;extended&#43;description&#43;field&#43;for&#43;more&#43;detail%29.%0A%2A&#43;Create&#43;a&#43;new&#43;branch&#43;so&#43;you&#43;can&#43;preview&#43;your&#43;new&#43;file&#43;and&#43;request&#43;a&#43;review&#43;via&#43;Pull&#43;Request.%0A" target="_blank"><i class="fa fa-edit fa-fw"></i> Create child page</a>
    <a href="https://github.com/tgirke/GEN242//issues/new?title=Programming%20in%20R" target="_blank"><i class="fab fa-github fa-fw"></i> Create documentation issue</a>
    
      
      <a href="https://github.com/tgirke/GEN242/issues/new" target="_blank"><i class="fas fa-tasks fa-fw"></i> Create project issue</a>
    

  
  
  </div>





<nav id="TableOfContents">
  <ul>
    <li><a href="#overview">Overview</a></li>
    <li><a href="#why-programming-in-r">Why Programming in R?</a></li>
    <li><a href="#r-basics">R Basics</a></li>
    <li><a href="#code-editors-for-r">Code Editors for R</a></li>
    <li><a href="#finding-help">Finding Help</a></li>
    <li><a href="#control-structures">Control Structures</a>
      <ul>
        <li><a href="#important-operators">Important Operators</a>
          <ul>
            <li><a href="#comparison-operators">Comparison operators</a></li>
            <li><a href="#logical-operators">Logical operators</a></li>
          </ul>
        </li>
        <li><a href="#conditional-executions-if-statements">Conditional Executions: <code>if</code> Statements</a></li>
        <li><a href="#conditional-executions-ifelse-statements">Conditional Executions: <code>ifelse</code> Statements</a></li>
      </ul>
    </li>
    <li><a href="#loops">Loops</a>
      <ul>
        <li><a href="#for-loop"><code>for</code> loop</a>
          <ul>
            <li><a href="#conditional-stop-of-loops">Conditional Stop of Loops</a></li>
          </ul>
        </li>
        <li><a href="#while-loop"><code>while</code> loop</a></li>
        <li><a href="#the-apply-function-family">The <code>apply</code> Function Family</a>
          <ul>
            <li><a href="#apply"><code>apply</code></a></li>
            <li><a href="#tapply"><code>tapply</code></a></li>
            <li><a href="#sapply-lapply-and-vapply"><code>sapply</code>, <code>lapply</code> and <code>vapply</code></a></li>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#functions">Functions</a>
      <ul>
        <li><a href="#function-overview">Function Overview</a></li>
        <li><a href="#function-syntax-rules">Function Syntax Rules</a></li>
        <li><a href="#examples">Examples</a></li>
      </ul>
    </li>
    <li><a href="#useful-utilities">Useful Utilities</a>
      <ul>
        <li><a href="#debugging-utilities">Debugging Utilities</a></li>
        <li><a href="#regular-expressions">Regular Expressions</a>
          <ul>
            <li><a href="#string-matching-with-grep">String matching with <code>grep</code></a></li>
            <li><a href="#string-substitution-with-gsub">String substitution with <code>gsub</code></a></li>
          </ul>
        </li>
        <li><a href="#interpreting-a-character-string-as-expression">Interpreting a Character String as Expression</a></li>
        <li><a href="#replacement-split-and-paste-functions-for-strings">Replacement, Split and Paste Functions for Strings</a></li>
        <li><a href="#time-date-and-sleep">Time, Date and Sleep</a>
          <ul>
            <li><a href="#example">Example</a>
              <ul>
                <li><a href="#import-of-specific-file-lines-with-regular-expression">Import of Specific File Lines with Regular Expression</a></li>
              </ul>
            </li>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#calling-external-software">Calling External Software</a></li>
    <li><a href="#running-r-scripts">Running R Scripts</a>
      <ul>
        <li><a href="#possibilities-for-executing-r-scripts">Possibilities for Executing R Scripts</a>
          <ul>
            <li><a href="#r-console">R console</a></li>
            <li><a href="#command-line">Command-line</a></li>
            <li><a href="#passing-arguments-from-command-line-to-r">Passing arguments from command-line to R</a></li>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#building-r-packages">Building R Packages</a>
      <ul>
        <li><a href="#short-overview-of-package-building-process">Short Overview of Package Building Process</a></li>
      </ul>
    </li>
    <li><a href="#programming-exercises">Programming Exercises</a>
      <ul>
        <li><a href="#exercise-1">Exercise 1</a>
          <ul>
            <li><a href="#for-loop-1"><code>for</code> loop</a></li>
            <li><a href="#while-loop-1"><code>while</code> loop</a></li>
            <li><a href="#apply-loop"><code>apply</code> loop</a></li>
            <li><a href="#avoiding-loops">Avoiding loops</a></li>
          </ul>
        </li>
        <li><a href="#exercise-2">Exercise 2</a>
          <ul>
            <li><a href="#custom-functions">Custom functions</a></li>
          </ul>
        </li>
        <li><a href="#exercise-3">Exercise 3</a>
          <ul>
            <li><a href="#nested-loops-to-generate-similarity-matrices">Nested loops to generate similarity matrices</a></li>
          </ul>
        </li>
        <li><a href="#exercise-4">Exercise 4</a>
          <ul>
            <li><a href="#build-your-own-r-package">Build your own R package</a></li>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#homework-5">Homework 5</a>
      <ul>
        <li><a href="#reverse-and-complement-of-dna">Reverse and complement of DNA</a></li>
        <li><a href="#translate-dna-into-protein">Translate DNA into Protein</a></li>
        <li><a href="#homework-submission">Homework submission</a></li>
        <li><a href="#due-date">Due date</a></li>
        <li><a href="#homework-solutions">Homework Solutions</a></li>
      </ul>
    </li>
    <li><a href="#session-info">Session Info</a></li>
    <li><a href="#references">References</a></li>
  </ul>
</nav>



              </div>
              <main class="col-12 col-md-9 col-xl-8 pl-md-5" role="main">
                
  

                <nav aria-label="breadcrumb" class="d-none d-md-block d-print-none">
	<ol class="breadcrumb spb-1">
		







<li class="breadcrumb-item" >
	<a href="../../../manuals/">Manuals</a>
</li>




<li class="breadcrumb-item active" aria-current="page">
	<a href="../../../manuals/rprogramming/rprogramming/">Programming in R</a>
</li>

	</ol>
</nav	>

                
<div class="td-content">
	<h1>Programming in R</h1>
    
	
        <p class="reading-time"><i class="fa fa-clock" aria-hidden="true"></i> 18 minute read</p>
           
	<!--
- Compile from command-line
Rscript -e "rmarkdown::render('Programming_in_R.Rmd', c('html_document'), clean=F); knitr::knit('Programming_in_R.Rmd', tangle=TRUE)"; Rscript ../md2jekyll.R Programming_in_R.knit.md 9; Rscript -e "rmarkdown::render('Programming_in_R.Rmd', c('pdf_document'))"
-->
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>
<h2 id="overview">Overview</h2>
<p>One of the main attractions of using the R
(<a href="http://cran.at.r-project.org">http://cran.at.r-project.org</a>) environment is
the ease with which users can write their own programs and custom functions.
The R programming syntax is extremely easy to learn, even for users with no
previous programming experience. Once the basic R programming control
structures are understood, users can use the R language as a powerful
environment to perform complex custom analyses of almost any type of data (Gentleman 2008).</p>
<h2 id="why-programming-in-r">Why Programming in R?</h2>
<ul>
<li>Powerful statistical environment and programming language</li>
<li>Facilitates reproducible research</li>
<li>Efficient data structures make programming very easy</li>
<li>Ease of implementing custom functions</li>
<li>Powerful graphics</li>
<li>Access to fast growing number of analysis packages</li>
<li>Most widely used language in bioinformatics</li>
<li>Is standard for data mining and biostatistical analysis</li>
<li>Technical advantages: free, open-source, available for all OSs</li>
</ul>
<h2 id="r-basics">R Basics</h2>
<p>The previous <a href="http://girke.bioinformatics.ucr.edu/GEN242/mydoc/mydoc_Rbasics_01.html">Rbasics</a> tutorial provides a general introduction to the usage of the R environment and its basic command syntax.
More details can be found in the R &amp; BioConductor manual <a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual">here</a>.</p>
<h2 id="code-editors-for-r">Code Editors for R</h2>
<p>Several excellent code editors are available that provide functionalities like R syntax highlighting, auto code indenting and utilities to send code/functions to the R console.</p>
<ul>
<li><a href="https://www.rstudio.com/products/rstudio/features/">RStudio</a>: GUI-based IDE for R</li>
<li><a href="http://manuals.bioinformatics.ucr.edu/home/programming-in-r/vim-r">Vim-R-Tmux</a>: R working environment based on vim and tmux</li>
<li><a href="http://www.xemacs.org/Download/index.html">Emacs</a> (<a href="http://ess.r-project.org/">ESS add-on package</a>)</li>
<li><a href="https://wiki.gnome.org/Apps/Gedit">gedit</a> and <a href="https://wiki.gnome.org/Apps/Gedit">Rgedit</a></li>
<li><a href="https://rkward.kde.org/">RKWard</a></li>
<li><a href="http://www.walware.de/goto/statet">Eclipse</a></li>
<li><a href="http://jekyll.math.byuh.edu/other/howto/tinnr/install.shtml">Tinn-R</a></li>
<li><a href="https://sourceforge.net/projects/npptor/">Notepad++ (NppToR)</a></li>
</ul>
<center>
Programming in R using RStudio
</center>
<center>
<img title="R_Interfaces" src="../images/rstudio.png"/>
</center>
<center>
Programming in R using Vim or Emacs
</center>
<center>
<img title="vim-r" src="../images/vimR.png"/>
</center>
<h2 id="finding-help">Finding Help</h2>
<p>Reference list on R programming (selection)</p>
<ul>
<li><a href="http://adv-r.had.co.nz/">Advanced R</a>, by Hadley Wickham</li>
<li><a href="http://master.bioconductor.org/help/publications/books/r-programming-for-bioinformatics/">R Programming for Bioinformatics</a>, by Robert Gentleman</li>
<li><a href="http://www.stats.ox.ac.uk/pub/MASS3/Sprog/">S Programming</a>, by W. N. Venables and B. D. Ripley</li>
<li><a href="http://www.amazon.com/Programming-Data-Language-Lecture-Economics/dp/0387985034">Programming with Data</a>, by John M. Chambers</li>
<li><a href="http://www1.maths.lth.se/help/R/">R Help</a> &amp; <a href="http://www1.maths.lth.se/help/R/RCC/">R Coding Conventions</a>, Henrik Bengtsson, Lund University</li>
<li><a href="http://zoonek2.free.fr/UNIX/48_R/02.html">Programming in R</a> (Vincent Zoonekynd)</li>
<li><a href="http://www2.warwick.ac.uk/fac/sci/moac/people/students/peter_cock/r">Peter’s R Programming Pages</a>, University of Warwick</li>
<li><a href="http://pj.freefaculty.org/R/statsRus.html">Rtips</a>, Paul Johnsson, University of Kansas</li>
<li><a href="http://heather.cs.ucdavis.edu/~matloff/r.html">R for Programmers</a>, Norm Matloff, UC Davis</li>
<li><a href="http://www.statistik.uni-dortmund.de/useR-2008/tutorials/useR2008introhighperfR.pdf">High-Performance R</a>, Dirk Eddelbuettel tutorial presented at useR-2008</li>
<li><a href="http://www.stat.harvard.edu/ccr2005/index.html">C/C++ level programming for R</a>, Gopi Goswami</li>
</ul>
<h2 id="control-structures">Control Structures</h2>
<h3 id="important-operators">Important Operators</h3>
<h4 id="comparison-operators">Comparison operators</h4>
<ul>
<li><code>==</code> (equal)</li>
<li><code>!=</code> (not equal)</li>
<li><code>&gt;</code> (greater than)</li>
<li><code>&gt;=</code> (greater than or equal)</li>
<li><code>&lt;</code> (less than)</li>
<li><code>&lt;=</code> (less than or equal)</li>
</ul>
<h4 id="logical-operators">Logical operators</h4>
<ul>
<li><code>&amp;</code> (and)</li>
<li><code>|</code> (or)</li>
<li><code>!</code> (not)</li>
</ul>
<h3 id="conditional-executions-if-statements">Conditional Executions: <code>if</code> Statements</h3>
<p>An <code>if</code> statement operates on length-one logical vectors.</p>
<p><strong>Syntax</strong></p>
<pre><code class="language-r">if (TRUE) { 
    statements_1 
} else { 
    statements_2 
}
</code></pre>
<p>In the <code>else</code> component, avoid inserting newlines between <code>} else</code>. For details on how to best and consistently format R code,
this <a href="http://adv-r.had.co.nz/Style.html">style guide</a> is a good start. In addition, the <a href="https://yihui.org/formatr/"><code>formatR</code></a> package can be helpful.</p>
<p><strong>Example</strong></p>
<pre><code class="language-r">if (1==0) { 
    print(1) 
} else { 
    print(2) 
}
</code></pre>
<pre><code>## [1] 2
</code></pre>
<h3 id="conditional-executions-ifelse-statements">Conditional Executions: <code>ifelse</code> Statements</h3>
<p>The <code>ifelse</code> statement operates on vectors.</p>
<p><strong>Syntax</strong></p>
<pre><code class="language-r">ifelse(test, true_value, false_value)
</code></pre>
<p><strong>Example</strong></p>
<pre><code class="language-r">x &lt;- 1:10 
ifelse(x&lt;5, x, 0)
</code></pre>
<pre><code>##  [1] 1 2 3 4 0 0 0 0 0 0
</code></pre>
<h2 id="loops">Loops</h2>
<h3 id="for-loop"><code>for</code> loop</h3>
<p><code>for</code> loops iterate over elements of a looping vector.</p>
<p><strong>Syntax</strong></p>
<pre><code class="language-r">for(variable in sequence) { 
    statements 
}
</code></pre>
<p><strong>Example</strong></p>
<pre><code class="language-r">mydf &lt;- iris
myve &lt;- NULL
for(i in seq(along=mydf[,1])) {
    myve &lt;- c(myve, mean(as.numeric(mydf[i,1:3])))
}
myve[1:8]
</code></pre>
<pre><code>## [1] 3.333333 3.100000 3.066667 3.066667 3.333333 3.666667 3.133333 3.300000
</code></pre>
<p><strong>Note:</strong> Inject into objecs is much faster than append approach with <code>c</code>, <code>cbind</code>, etc.</p>
<p><strong>Example</strong></p>
<pre><code class="language-r">myve &lt;- numeric(length(mydf[,1]))
for(i in seq(along=myve)) {
    myve[i] &lt;- mean(as.numeric(mydf[i,1:3]))
}
myve[1:8]
</code></pre>
<pre><code>## [1] 3.333333 3.100000 3.066667 3.066667 3.333333 3.666667 3.133333 3.300000
</code></pre>
<h4 id="conditional-stop-of-loops">Conditional Stop of Loops</h4>
<p>The <code>stop</code> function can be used to break out of a loop (or a function) when a condition becomes <code>TRUE</code>. In addition, an error message will be printed.</p>
<p><strong>Example</strong></p>
<pre><code class="language-r">x &lt;- 1:10
z &lt;- NULL
for(i in seq(along=x)) { 
    if(x[i] &lt; 5) { 
        z &lt;- c(z, x[i]-1)  
    } else { 
        stop(&quot;values need to be &lt; 5&quot;) 
    }
}
</code></pre>
<h3 id="while-loop"><code>while</code> loop</h3>
<p>Iterates as long as a condition is true.</p>
<p><strong>Syntax</strong></p>
<pre><code class="language-r">while(condition) {
    statements
}
</code></pre>
<p><strong>Example</strong></p>
<pre><code class="language-r">z &lt;- 0
while(z&lt;5) { 
    z &lt;- z + 2
    print(z)  
}
</code></pre>
<pre><code>## [1] 2
## [1] 4
## [1] 6
</code></pre>
<h3 id="the-apply-function-family">The <code>apply</code> Function Family</h3>
<h4 id="apply"><code>apply</code></h4>
<p><strong>Syntax</strong></p>
<pre><code class="language-r">apply(X, MARGIN, FUN, ARGs)
</code></pre>
<p><strong>Arguments</strong></p>
<ul>
<li><code>X</code>: <code>array</code>, <code>matrix</code> or <code>data.frame</code></li>
<li><code>MARGIN</code>: <code>1</code> for rows, <code>2</code> for columns</li>
<li><code>FUN</code>: one or more functions</li>
<li><code>ARGs</code>: possible arguments for functions</li>
</ul>
<p><strong>Example</strong></p>
<pre><code class="language-r">apply(iris[1:8,1:3], 1, mean)
</code></pre>
<pre><code>##        1        2        3        4        5        6        7        8 
## 3.333333 3.100000 3.066667 3.066667 3.333333 3.666667 3.133333 3.300000
</code></pre>
<h4 id="tapply"><code>tapply</code></h4>
<p>Applies a function to vector components that are defined by a factor.</p>
<p><strong>Syntax</strong></p>
<pre><code class="language-r">tapply(vector, factor, FUN)
</code></pre>
<p><strong>Example</strong></p>
<pre><code class="language-r">iris[1:2,]
</code></pre>
<pre><code>##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 1          5.1         3.5          1.4         0.2  setosa
## 2          4.9         3.0          1.4         0.2  setosa
</code></pre>
<pre><code class="language-r">tapply(iris$Sepal.Length, iris$Species, mean)
</code></pre>
<pre><code>##     setosa versicolor  virginica 
##      5.006      5.936      6.588
</code></pre>
<h4 id="sapply-lapply-and-vapply"><code>sapply</code>, <code>lapply</code> and <code>vapply</code></h4>
<p>All three apply a function to vector or list. The <code>lapply</code> function always returns a list, while <code>sapply</code> returns <code>vector</code> or <code>matrix</code> objects when possible. If not then a list is returned.
The <code>vapply</code> function returns a vector or array of type matching the <code>FUN.VALUE</code>. Compared to <code>sappy</code>, <code>vapply</code> is a safer choice with respect to controlling specific output
types to avoid exception handling problems.</p>
<p><strong>Examples</strong></p>
<pre><code class="language-r">l &lt;- list(a = 1:10, beta = exp(-3:3), logic = c(TRUE,FALSE,FALSE,TRUE))
lapply(l, mean)
</code></pre>
<pre><code>## $a
## [1] 5.5
## 
## $beta
## [1] 4.535125
## 
## $logic
## [1] 0.5
</code></pre>
<pre><code class="language-r">sapply(l, mean)
</code></pre>
<pre><code>##        a     beta    logic 
## 5.500000 4.535125 0.500000
</code></pre>
<pre><code class="language-r">vapply(l, mean, FUN.VALUE=numeric(1))
</code></pre>
<pre><code>##        a     beta    logic 
## 5.500000 4.535125 0.500000
</code></pre>
<p>Often used in combination with a function definition:</p>
<pre><code class="language-r">lapply(names(l), function(x) mean(l[[x]]))
sapply(names(l), function(x) mean(l[[x]]))
vapply(names(l), function(x) mean(l[[x]]), FUN.VALUE=numeric(1))
</code></pre>
<h2 id="functions">Functions</h2>
<h3 id="function-overview">Function Overview</h3>
<p>A very useful feature of the R environment is the possibility to expand existing functions and to easily write custom functions. In fact, most of the R software can be viewed as a series of R functions.</p>
<p><strong>Syntax</strong> to define function</p>
<pre><code class="language-r">myfct &lt;- function(arg1, arg2, ...) { 
    function_body 
}
</code></pre>
<p><strong>Syntax</strong> to call functions</p>
<pre><code class="language-r">myfct(arg1=..., arg2=...)
</code></pre>
<p>The value returned by a function is the value of the function body, which is usually an unassigned final expression, <em>e.g.</em>: <code>return()</code></p>
<h3 id="function-syntax-rules">Function Syntax Rules</h3>
<p><strong>General</strong></p>
<ul>
<li>Functions are defined by
<ol>
<li>The assignment with the keyword <code>function</code></li>
<li>The declaration of arguments/variables (<code>arg1, arg2, ...</code>)</li>
<li>The definition of operations (<code>function_body</code>) that perform computations on the provided arguments. A function name needs to be assigned to call the function.</li>
</ol>
</li>
</ul>
<p><strong>Naming</strong></p>
<ul>
<li>Function names can be almost anything. However, the usage of names of existing functions should be avoided.</li>
</ul>
<p><strong>Arguments</strong></p>
<ul>
<li>It is often useful to provide default values for arguments (<em>e.g.</em>: <code>arg1=1:10</code>). This way they don’t need to be provided in a function call. The argument list can also be left empty (<code>myfct &lt;- function() { fct_body }</code>) if a function is expected to return always the same value(s). The argument <code>...</code> can be used to allow one function to pass on argument settings to another.</li>
</ul>
<p><strong>Body</strong></p>
<ul>
<li>The actual expressions (commands/operations) are defined in the function body which should be enclosed by braces. The individual commands are separated by semicolons or new lines (preferred).</li>
</ul>
<p><strong>Usage</strong></p>
<ul>
<li>Functions are called by their name followed by parentheses containing possible argument names. Empty parenthesis after the function name will result in an error message when a function requires certain arguments to be provided by the user. The function name alone will print the definition of a function.</li>
</ul>
<p><strong>Scope</strong></p>
<ul>
<li>Variables created inside a function exist only for the life time of a function. Thus, they are not accessible outside of the function. To force variables in functions to exist globally, one can use the double assignment operator: <code>&lt;&lt;-</code></li>
</ul>
<h3 id="examples">Examples</h3>
<p><strong>Define sample function</strong></p>
<pre><code class="language-r">myfct &lt;- function(x1, x2=5) { 
    z1 &lt;- x1 / x1
    z2 &lt;- x2 * x2
        myvec &lt;- c(z1, z2) 
        return(myvec)
} 
</code></pre>
<p><strong>Function usage</strong></p>
<p>Apply function to values <code>2</code> and <code>5</code></p>
<pre><code class="language-r">myfct(x1=2, x2=5) 
</code></pre>
<pre><code>## [1]  1 25
</code></pre>
<p>Run without argument names</p>
<pre><code class="language-r">myfct(2, 5) 
</code></pre>
<pre><code>## [1]  1 25
</code></pre>
<p>Makes use of default value <code>5</code></p>
<pre><code class="language-r">myfct(x1=2) 
</code></pre>
<pre><code>## [1]  1 25
</code></pre>
<p>Print function definition (often unintended)</p>
<pre><code class="language-r">myfct 
</code></pre>
<pre><code>## function(x1, x2=5) { 
##  z1 &lt;- x1 / x1
##  z2 &lt;- x2 * x2
##         myvec &lt;- c(z1, z2) 
##         return(myvec)
## }
## &lt;bytecode: 0x575bef4fbd40&gt;
</code></pre>
<h2 id="useful-utilities">Useful Utilities</h2>
<h3 id="debugging-utilities">Debugging Utilities</h3>
<p>Several debugging utilities are available for R. They include:</p>
<ul>
<li><code>traceback</code></li>
<li><code>browser</code></li>
<li><code>options(error=recover)</code></li>
<li><code>options(error=NULL)</code></li>
<li><code>debug</code></li>
</ul>
<p>The <a href="http://www.stats.uwo.ca/faculty/murdoch/software/debuggingR/">Debugging in R page</a> provides an overview of the available resources.</p>
<h3 id="regular-expressions">Regular Expressions</h3>
<p>R’s regular expression utilities work similar as in other languages. To learn how to use them in R, one can consult the main help page on this topic with <code>?regexp</code>.</p>
<h4 id="string-matching-with-grep">String matching with <code>grep</code></h4>
<p>The grep function can be used for finding patterns in strings, here letter <code>A</code> in vector <code>month.name</code>.</p>
<pre><code class="language-r">month.name[grep(&quot;A&quot;, month.name)] 
</code></pre>
<pre><code>## [1] &quot;April&quot;  &quot;August&quot;
</code></pre>
<h4 id="string-substitution-with-gsub">String substitution with <code>gsub</code></h4>
<p>Example for using regular expressions to substitute a pattern by another one using a back reference. Remember: single escapes <code>\</code> need to be double escaped <code>\\</code> in R.</p>
<pre><code class="language-r">gsub('(i.*a)', 'xxx_\\1', &quot;virginica&quot;, perl = TRUE) 
</code></pre>
<pre><code>## [1] &quot;vxxx_irginica&quot;
</code></pre>
<h3 id="interpreting-a-character-string-as-expression">Interpreting a Character String as Expression</h3>
<p>Some useful examples</p>
<p>Generates vector of object names in session</p>
<pre><code class="language-r">mylist &lt;- ls()
mylist[1] 
</code></pre>
<pre><code>## [1] &quot;i&quot;
</code></pre>
<p>Executes 1st entry as expression</p>
<pre><code class="language-r">get(mylist[1])
</code></pre>
<pre><code>## [1] 150
</code></pre>
<p>Alternative approach</p>
<pre><code class="language-r">eval(parse(text=mylist[1])) 
</code></pre>
<pre><code>## [1] 150
</code></pre>
<h3 id="replacement-split-and-paste-functions-for-strings">Replacement, Split and Paste Functions for Strings</h3>
<p><strong>Selected examples</strong></p>
<p>Substitution with back reference which inserts in this example <code>_</code> character</p>
<pre><code class="language-r">x &lt;- gsub(&quot;(a)&quot;,&quot;\\1_&quot;, month.name[1], perl=T) 
x
</code></pre>
<pre><code>## [1] &quot;Ja_nua_ry&quot;
</code></pre>
<p>Split string on inserted character from above</p>
<pre><code class="language-r">strsplit(x,&quot;_&quot;)
</code></pre>
<pre><code>## [[1]]
## [1] &quot;Ja&quot;  &quot;nua&quot; &quot;ry&quot;
</code></pre>
<p>Reverse a character string by splitting first all characters into vector fields</p>
<pre><code class="language-r">paste(rev(unlist(strsplit(x, NULL))), collapse=&quot;&quot;) 
</code></pre>
<pre><code>## [1] &quot;yr_aun_aJ&quot;
</code></pre>
<h3 id="time-date-and-sleep">Time, Date and Sleep</h3>
<p><strong>Selected examples</strong></p>
<p>Return CPU (and other) times that an expression used (here ls)</p>
<pre><code class="language-r">system.time(ls()) 
</code></pre>
<pre><code>##    user  system elapsed 
##       0       0       0
</code></pre>
<p>Return the current system date and time</p>
<pre><code class="language-r">date() 
</code></pre>
<pre><code>## [1] &quot;Thu Feb 18 14:46:40 2021&quot;
</code></pre>
<p>Pause execution of R expressions for a given number of seconds (e.g. in loop)</p>
<pre><code class="language-r">Sys.sleep(1) 
</code></pre>
<h4 id="example">Example</h4>
<h5 id="import-of-specific-file-lines-with-regular-expression">Import of Specific File Lines with Regular Expression</h5>
<p>The following example demonstrates the retrieval of specific lines from an external file with a regular expression. First, an external file is created with the <code>cat</code> function, all lines of this file are imported into a vector with <code>readLines</code>, the specific elements (lines) are then retieved with the <code>grep</code> function, and the resulting lines are split into vector fields with <code>strsplit</code>.</p>
<pre><code class="language-r">cat(month.name, file=&quot;zzz.txt&quot;, sep=&quot;\n&quot;)
x &lt;- readLines(&quot;zzz.txt&quot;)
x[1:6] 
</code></pre>
<pre><code>## [1] &quot;January&quot;  &quot;February&quot; &quot;March&quot;    &quot;April&quot;    &quot;May&quot;      &quot;June&quot;
</code></pre>
<pre><code class="language-r">x &lt;- x[c(grep(&quot;^J&quot;, as.character(x), perl = TRUE))]
t(as.data.frame(strsplit(x, &quot;u&quot;)))
</code></pre>
<pre><code>##                 [,1]  [,2] 
## c..Jan....ary.. &quot;Jan&quot; &quot;ary&quot;
## c..J....ne..    &quot;J&quot;   &quot;ne&quot; 
## c..J....ly..    &quot;J&quot;   &quot;ly&quot;
</code></pre>
<h2 id="calling-external-software">Calling External Software</h2>
<p>External command-line software can be called with <code>system</code>. The following example calls <code>blastall</code> from R</p>
<pre><code class="language-r">system(&quot;blastall -p blastp -i seq.fasta -d uniprot -o seq.blastp&quot;)
</code></pre>
<pre><code>## Warning in system(&quot;blastall -p blastp -i seq.fasta -d uniprot -o seq.blastp&quot;): error in running
## command
</code></pre>
<h2 id="running-r-scripts">Running R Scripts</h2>
<h3 id="possibilities-for-executing-r-scripts">Possibilities for Executing R Scripts</h3>
<h4 id="r-console">R console</h4>
<pre><code class="language-r">source(&quot;my_script.R&quot;)
</code></pre>
<h4 id="command-line">Command-line</h4>
<pre><code class="language-sh">Rscript my_script.R # or just ./myscript.R after making it executable
R CMD BATCH my_script.R # Alternative way 1 
R --slave &lt; my_script.R # Alternative way 2
</code></pre>
<h4 id="passing-arguments-from-command-line-to-r">Passing arguments from command-line to R</h4>
<p>Create an R script named <code>test.R</code> with the following content:</p>
<pre><code class="language-sh">myarg &lt;- commandArgs()
print(iris[1:myarg[6], ])
</code></pre>
<p>Then run it from the command-line like this:</p>
<pre><code class="language-sh">Rscript test.R 10
</code></pre>
<p>In the given example the number <code>10</code> is passed on from the command-line as an argument to the R script which is used to return to <code>STDOUT</code> the first 10 rows of the <code>iris</code> sample data. If several arguments are provided, they will be interpreted as one string and need to be split in R with the strsplit function. A more detailed example can be found <a href="http://manuals.bioinformatics.ucr.edu/home/ht-seq#TOC-Quality-Reports-of-FASTQ-Files-">here</a>.</p>
<h2 id="building-r-packages">Building R Packages</h2>
<h3 id="short-overview-of-package-building-process">Short Overview of Package Building Process</h3>
<p>R packages can be built with the <code>package.skeleton</code> function. The given example will create a directory named <code>mypackage</code> containing the skeleton of the package for all functions, methods and classes defined in the R script(s) passed on to the <code>code_files</code> argument. The basic structure of the package directory is described <a href="http://manuals.bioinformatics.ucr.edu/home/programming-in-r#Progr_pack">here</a>. The package directory will also contain a file named <code>Read-and-delete-me</code> with instructions for completing the package:</p>
<pre><code class="language-r">package.skeleton(name=&quot;mypackage&quot;, code_files=c(&quot;script1.R&quot;, &quot;script2.R&quot;))
</code></pre>
<p>Once a package skeleton is available one can build the package from the command-line (Linux/OS X). This will create a tarball of the package with its version number encoded in the file name. Subequently, the package tarball needs to be checked for errors with:</p>
<pre><code class="language-r">R CMD build mypackage
R CMD check mypackage_1.0.tar.gz
</code></pre>
<p>Install package from source</p>
<pre><code class="language-r">install.packages(&quot;mypackage_1.0.tar.gz&quot;, repos=NULL) 
</code></pre>
<p>For more details see <a href="http://manuals.bioinformatics.ucr.edu/home/programming-in-r#TOC-Building-R-Packages">here</a></p>
<h2 id="programming-exercises">Programming Exercises</h2>
<h3 id="exercise-1">Exercise 1</h3>
<h4 id="for-loop-1"><code>for</code> loop</h4>
<p><strong>Task 1.1</strong>: Compute the mean of each row in <code>myMA</code> by applying the mean function in a <code>for</code> loop.</p>
<pre><code class="language-r">myMA &lt;- matrix(rnorm(500), 100, 5, dimnames=list(1:100, paste(&quot;C&quot;, 1:5, sep=&quot;&quot;)))
myve_for &lt;- NULL
for(i in seq(along=myMA[,1])) {
    myve_for &lt;- c(myve_for, mean(as.numeric(myMA[i, ])))
}
myResult &lt;- cbind(myMA, mean_for=myve_for)
myResult[1:4, ]
</code></pre>
<pre><code>##           C1          C2         C3         C4          C5   mean_for
## 1 -1.5890569 -0.00759616 -0.7651252 -0.1334947  1.75902044 -0.1472505
## 2 -0.5760445 -0.07502956  1.8156748 -0.6022573  0.04071099  0.1206109
## 3  0.1452555 -0.85005686  0.8514295 -0.4692688 -0.95188121 -0.2549044
## 4  0.9678927 -0.48747853  0.5058947  0.5961237 -0.84373458  0.1477396
</code></pre>
<h4 id="while-loop-1"><code>while</code> loop</h4>
<p><strong>Task 1.2</strong>: Compute the mean of each row in <code>myMA</code> by applying the mean function in a <code>while</code> loop.</p>
<pre><code class="language-r">z &lt;- 1
myve_while &lt;- NULL
while(z &lt;= length(myMA[,1])) {
    myve_while &lt;- c(myve_while, mean(as.numeric(myMA[z, ])))
    z &lt;- z + 1
}
myResult &lt;- cbind(myMA, mean_for=myve_for, mean_while=myve_while)
myResult[1:4, -c(1,2)]
</code></pre>
<pre><code>##           C3         C4          C5   mean_for mean_while
## 1 -0.7651252 -0.1334947  1.75902044 -0.1472505 -0.1472505
## 2  1.8156748 -0.6022573  0.04071099  0.1206109  0.1206109
## 3  0.8514295 -0.4692688 -0.95188121 -0.2549044 -0.2549044
## 4  0.5058947  0.5961237 -0.84373458  0.1477396  0.1477396
</code></pre>
<p><strong>Task 1.3</strong>: Confirm that the results from both mean calculations are identical</p>
<pre><code class="language-r">all(myResult[,6] == myResult[,7])
</code></pre>
<pre><code>## [1] TRUE
</code></pre>
<h4 id="apply-loop"><code>apply</code> loop</h4>
<p><strong>Task 1.4</strong>: Compute the mean of each row in myMA by applying the mean function in an <code>apply</code> loop</p>
<pre><code class="language-r">myve_apply &lt;- apply(myMA, 1, mean)
myResult &lt;- cbind(myMA, mean_for=myve_for, mean_while=myve_while, mean_apply=myve_apply)
myResult[1:4, -c(1,2)]
</code></pre>
<pre><code>##           C3         C4          C5   mean_for mean_while mean_apply
## 1 -0.7651252 -0.1334947  1.75902044 -0.1472505 -0.1472505 -0.1472505
## 2  1.8156748 -0.6022573  0.04071099  0.1206109  0.1206109  0.1206109
## 3  0.8514295 -0.4692688 -0.95188121 -0.2549044 -0.2549044 -0.2549044
## 4  0.5058947  0.5961237 -0.84373458  0.1477396  0.1477396  0.1477396
</code></pre>
<h4 id="avoiding-loops">Avoiding loops</h4>
<p><strong>Task 1.5</strong>: When operating on large data sets it is much faster to use the rowMeans function</p>
<pre><code class="language-r">mymean &lt;- rowMeans(myMA)
myResult &lt;- cbind(myMA, mean_for=myve_for, mean_while=myve_while, mean_apply=myve_apply, mean_int=mymean)
myResult[1:4, -c(1,2,3)]
</code></pre>
<pre><code>##           C4          C5   mean_for mean_while mean_apply   mean_int
## 1 -0.1334947  1.75902044 -0.1472505 -0.1472505 -0.1472505 -0.1472505
## 2 -0.6022573  0.04071099  0.1206109  0.1206109  0.1206109  0.1206109
## 3 -0.4692688 -0.95188121 -0.2549044 -0.2549044 -0.2549044 -0.2549044
## 4  0.5961237 -0.84373458  0.1477396  0.1477396  0.1477396  0.1477396
</code></pre>
<h3 id="exercise-2">Exercise 2</h3>
<h4 id="custom-functions">Custom functions</h4>
<p><strong>Task 2.1</strong>: Use the following code as basis to implement a function that allows the user to compute the mean for any combination of columns in a matrix or data frame. The first argument of this function should specify the input data set, the second the mathematical function to be passed on (<em>e.g.</em> <code>mean</code>, <code>sd</code>, <code>max</code>) and the third one should allow the selection of the columns by providing a grouping vector.</p>
<pre><code class="language-r">myMA &lt;- matrix(rnorm(100000), 10000, 10, dimnames=list(1:10000, paste(&quot;C&quot;, 1:10, sep=&quot;&quot;)))
myMA[1:2,]
</code></pre>
<pre><code>##           C1        C2         C3        C4        C5        C6         C7        C8         C9
## 1 -0.1430870 1.6518732 -0.4030777 1.3449119 -1.275383 0.8054689  0.7659832 0.2528063  0.2783535
## 2 -0.6469048 0.5299854 -0.4038505 0.4141411 -1.097646 1.9669993 -2.5523817 1.6555577 -1.4068711
##           C10
## 1 -0.03573911
## 2 -0.58036942
</code></pre>
<pre><code class="language-r">myList &lt;- tapply(colnames(myMA), c(1,1,1,2,2,2,3,3,4,4), list) 
names(myList) &lt;- sapply(myList, paste, collapse=&quot;_&quot;)
myMAmean &lt;- sapply(myList, function(x) apply(myMA[,x], 1, mean))
myMAmean[1:4,] 
</code></pre>
<pre><code>##     C1_C2_C3    C4_C5_C6      C7_C8       C9_C10
## 1  0.3685695  0.29166595  0.5093947  0.121307169
## 2 -0.1735900  0.42783144 -0.4484120 -0.993620282
## 3 -0.5667075 -0.04833491  0.6504228  0.006118701
## 4  0.7495088 -0.44156279  0.5179593  0.082354279
</code></pre>
<!---
Solution

-->
<h3 id="exercise-3">Exercise 3</h3>
<h4 id="nested-loops-to-generate-similarity-matrices">Nested loops to generate similarity matrices</h4>
<p><strong>Task 3.1</strong>: Create a sample list populated with character vectors of different lengths</p>
<pre><code class="language-r">setlist &lt;- lapply(11:30, function(x) sample(letters, x, replace=TRUE))
names(setlist) &lt;- paste(&quot;S&quot;, seq(along=setlist), sep=&quot;&quot;) 
setlist[1:6]
</code></pre>
<pre><code>## $S1
##  [1] &quot;s&quot; &quot;a&quot; &quot;m&quot; &quot;n&quot; &quot;w&quot; &quot;s&quot; &quot;m&quot; &quot;t&quot; &quot;n&quot; &quot;y&quot; &quot;c&quot;
## 
## $S2
##  [1] &quot;x&quot; &quot;r&quot; &quot;p&quot; &quot;v&quot; &quot;a&quot; &quot;n&quot; &quot;d&quot; &quot;b&quot; &quot;o&quot; &quot;d&quot; &quot;g&quot; &quot;d&quot;
## 
## $S3
##  [1] &quot;e&quot; &quot;p&quot; &quot;k&quot; &quot;q&quot; &quot;x&quot; &quot;p&quot; &quot;u&quot; &quot;s&quot; &quot;z&quot; &quot;t&quot; &quot;j&quot; &quot;r&quot; &quot;a&quot;
## 
## $S4
##  [1] &quot;c&quot; &quot;f&quot; &quot;e&quot; &quot;j&quot; &quot;k&quot; &quot;j&quot; &quot;c&quot; &quot;q&quot; &quot;b&quot; &quot;j&quot; &quot;o&quot; &quot;x&quot; &quot;n&quot; &quot;x&quot;
## 
## $S5
##  [1] &quot;q&quot; &quot;d&quot; &quot;a&quot; &quot;f&quot; &quot;j&quot; &quot;m&quot; &quot;m&quot; &quot;o&quot; &quot;c&quot; &quot;k&quot; &quot;c&quot; &quot;q&quot; &quot;s&quot; &quot;u&quot; &quot;s&quot;
## 
## $S6
##  [1] &quot;v&quot; &quot;i&quot; &quot;f&quot; &quot;z&quot; &quot;d&quot; &quot;m&quot; &quot;w&quot; &quot;f&quot; &quot;u&quot; &quot;b&quot; &quot;l&quot; &quot;c&quot; &quot;g&quot; &quot;f&quot; &quot;c&quot; &quot;u&quot;
</code></pre>
<p><strong>Task 3.2</strong>: Compute the length for all pairwise intersects of the vectors stored in <code>setlist</code>. The intersects can be determined with the <code>%in%</code> function like this: <code>sum(setlist[[1]] %in% setlist[[2]])</code></p>
<pre><code class="language-r">setlist &lt;- sapply(setlist, unique)
olMA &lt;- sapply(names(setlist), function(x) sapply(names(setlist), 
               function(y) sum(setlist[[x]] %in% setlist[[y]])))
olMA[1:12,] 
</code></pre>
<pre><code>##     S1 S2 S3 S4 S5 S6 S7 S8 S9 S10 S11 S12 S13 S14 S15 S16 S17 S18 S19 S20
## S1   8  2  3  2  4  3  3  4  6   3   3   5   5   8   4   8   6   5   8   5
## S2   2 10  4  4  3  4  3  4  4   8   4   7   6   6   6   8   6   6   6   7
## S3   3  4 12  5  6  2  5  6  6   7   7   7   7   6   6   9  12   8   9   8
## S4   2  4  5 10  6  3  2  4  4   8   2   7   6   8   6   7   8   6   6   6
## S5   4  3  6  6 11  5  3  6  7   8   5   6   5   8   6  10   9   7   9   7
## S6   3  4  2  3  5 12  6  8  5   7   4   6   6   8  10   8   7  10   9   8
## S7   3  3  5  2  3  6 11  5  7   5   4   5   6   5   8   8   6   7   7   8
## S8   4  4  6  4  6  8  5 13  5   8   5   7   9  10   9  10   8  11  12   7
## S9   6  4  6  4  7  5  7  5 13   8   5   8   8   7   8  11   9   6   8  11
## S10  3  8  7  8  8  7  5  8  8  16   8  10  11  10  10  12  11   9   9  10
## S11  3  4  7  2  5  4  4  5  5   8  11   6   6   6   7   7  10   7   9   9
## S12  5  7  7  7  6  6  5  7  8  10   6  15  10  11  12  11  11   9  11  11
</code></pre>
<p><strong>Task 3.3</strong> Plot the resulting intersect matrix as heat map. The <code>image</code> or the <code>heatmap.2</code> function from the <code>gplots</code> library can be used for this.</p>
<pre><code class="language-r">image(olMA)
</code></pre>
<p><img src="../../../en/manuals/rprogramming/rprogramming_files/figure-html/nested_loops3-1.png" width="672" /></p>
<h3 id="exercise-4">Exercise 4</h3>
<h4 id="build-your-own-r-package">Build your own R package</h4>
<p><strong>Task 4.1</strong>: Save one or more of your functions to a file called <code>script.R</code> and build the package with the <code>package.skeleton</code> function.</p>
<pre><code class="language-r">package.skeleton(name=&quot;mypackage&quot;, code_files=c(&quot;script1.R&quot;))
</code></pre>
<p><strong>Task 4.2</strong>: Build tarball of the package</p>
<pre><code class="language-r">system(&quot;R CMD build mypackage&quot;)
</code></pre>
<p><strong>Task 4.3</strong>: Install and use package</p>
<pre><code class="language-r">install.packages(&quot;mypackage_1.0.tar.gz&quot;, repos=NULL, type=&quot;source&quot;)
library(mypackage)
?myMAcomp # Opens help for function defined by mypackage
</code></pre>
<h2 id="homework-5">Homework 5</h2>
<h3 id="reverse-and-complement-of-dna">Reverse and complement of DNA</h3>
<p><strong>Task 1</strong>: Write a <code>RevComp</code> function that returns the reverse and complement of a DNA sequence string. Include an argument that will allow to return only the reversed sequence, the complemented sequence or the reversed and complemented sequence. The following R functions will be useful for the implementation:</p>
<pre><code class="language-r">x &lt;- c(&quot;ATGCATTGGACGTTAG&quot;)  
x
</code></pre>
<pre><code>## [1] &quot;ATGCATTGGACGTTAG&quot;
</code></pre>
<pre><code class="language-r">x &lt;- substring(x, 1:nchar(x), 1:nchar(x)) 
x
</code></pre>
<pre><code>##  [1] &quot;A&quot; &quot;T&quot; &quot;G&quot; &quot;C&quot; &quot;A&quot; &quot;T&quot; &quot;T&quot; &quot;G&quot; &quot;G&quot; &quot;A&quot; &quot;C&quot; &quot;G&quot; &quot;T&quot; &quot;T&quot; &quot;A&quot; &quot;G&quot;
</code></pre>
<pre><code class="language-r">x &lt;- rev(x) 
x
</code></pre>
<pre><code>##  [1] &quot;G&quot; &quot;A&quot; &quot;T&quot; &quot;T&quot; &quot;G&quot; &quot;C&quot; &quot;A&quot; &quot;G&quot; &quot;G&quot; &quot;T&quot; &quot;T&quot; &quot;A&quot; &quot;C&quot; &quot;G&quot; &quot;T&quot; &quot;A&quot;
</code></pre>
<pre><code class="language-r">x &lt;- paste(x, collapse=&quot;&quot;)
x
</code></pre>
<pre><code>## [1] &quot;GATTGCAGGTTACGTA&quot;
</code></pre>
<pre><code class="language-r">chartr(&quot;ATGC&quot;, &quot;TACG&quot;, x) 
</code></pre>
<pre><code>## [1] &quot;CTAACGTCCAATGCAT&quot;
</code></pre>
<p><strong>Task 2</strong>: Write a function that applies the <code>RevComp</code> function to many sequences stored in a vector.</p>
<h3 id="translate-dna-into-protein">Translate DNA into Protein</h3>
<p><strong>Task 3</strong>: Write a function that will translate one or many DNA sequences in all three reading frames into proteins. The following commands will simplify this task:</p>
<pre><code class="language-r">AAdf &lt;- read.table(file=&quot;http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/My_R_Scripts/AA.txt&quot;, header=TRUE, sep=&quot;\t&quot;) 
AAdf[1:4,]
</code></pre>
<pre><code>##   Codon AA_1 AA_3 AA_Full AntiCodon
## 1   TCA    S  Ser  Serine       TGA
## 2   TCG    S  Ser  Serine       CGA
## 3   TCC    S  Ser  Serine       GGA
## 4   TCT    S  Ser  Serine       AGA
</code></pre>
<pre><code class="language-r">AAv &lt;- as.character(AAdf[,2]) 
names(AAv) &lt;- AAdf[,1] 
AAv
</code></pre>
<pre><code>## TCA TCG TCC TCT TTT TTC TTA TTG TAT TAC TAA TAG TGT TGC TGA TGG CTA CTG CTC CTT CCA CCG CCC CCT CAT 
## &quot;S&quot; &quot;S&quot; &quot;S&quot; &quot;S&quot; &quot;F&quot; &quot;F&quot; &quot;L&quot; &quot;L&quot; &quot;Y&quot; &quot;Y&quot; &quot;*&quot; &quot;*&quot; &quot;C&quot; &quot;C&quot; &quot;*&quot; &quot;W&quot; &quot;L&quot; &quot;L&quot; &quot;L&quot; &quot;L&quot; &quot;P&quot; &quot;P&quot; &quot;P&quot; &quot;P&quot; &quot;H&quot; 
## CAC CAA CAG CGA CGG CGC CGT ATT ATC ATA ATG ACA ACG ACC ACT AAT AAC AAA AAG AGT AGC AGA AGG GTA GTG 
## &quot;H&quot; &quot;Q&quot; &quot;Q&quot; &quot;R&quot; &quot;R&quot; &quot;R&quot; &quot;R&quot; &quot;I&quot; &quot;I&quot; &quot;I&quot; &quot;M&quot; &quot;T&quot; &quot;T&quot; &quot;T&quot; &quot;T&quot; &quot;N&quot; &quot;N&quot; &quot;K&quot; &quot;K&quot; &quot;S&quot; &quot;S&quot; &quot;R&quot; &quot;R&quot; &quot;V&quot; &quot;V&quot; 
## GTC GTT GCA GCG GCC GCT GAT GAC GAA GAG GGA GGG GGC GGT 
## &quot;V&quot; &quot;V&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot; &quot;D&quot; &quot;D&quot; &quot;E&quot; &quot;E&quot; &quot;G&quot; &quot;G&quot; &quot;G&quot; &quot;G&quot;
</code></pre>
<pre><code class="language-r">y &lt;- gsub(&quot;(...)&quot;, &quot;\\1_&quot;, x) 
y &lt;- unlist(strsplit(y, &quot;_&quot;)) 
y &lt;- y[grep(&quot;^...$&quot;, y)] 
AAv[y] 
</code></pre>
<pre><code>## GAT TGC AGG TTA CGT 
## &quot;D&quot; &quot;C&quot; &quot;R&quot; &quot;L&quot; &quot;R&quot;
</code></pre>
<h3 id="homework-submission">Homework submission</h3>
<p>Submit the 3 functions in one well structured and annotated R script to the instructor. The script should include instructions on how to use the functions.</p>
<h3 id="due-date">Due date</h3>
<p>This homework is due on Thu, April 26th at 6:00 PM.</p>
<h3 id="homework-solutions">Homework Solutions</h3>
<p>See homework section <a href="https://girke.bioinformatics.ucr.edu/GEN242/assignments/homework/hw05/hw05/">here</a></p>
<h2 id="session-info">Session Info</h2>
<pre><code class="language-r">sessionInfo()
</code></pre>
<pre><code>## R version 4.0.4 (2021-02-15)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 10 (buster)
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.8.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.8.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
##  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] ggplot2_3.3.2    limma_3.46.0     BiocStyle_2.18.0
## 
## loaded via a namespace (and not attached):
##  [1] knitr_1.30          magrittr_2.0.1      tidyselect_1.1.0    munsell_0.5.0      
##  [5] colorspace_2.0-0    R6_2.5.0            rlang_0.4.8         dplyr_1.0.2        
##  [9] stringr_1.4.0       tools_4.0.4         grid_4.0.4          gtable_0.3.0       
## [13] xfun_0.20           withr_2.3.0         ellipsis_0.3.1      htmltools_0.5.1.1  
## [17] yaml_2.2.1          digest_0.6.27       tibble_3.0.4        lifecycle_0.2.0    
## [21] crayon_1.3.4        bookdown_0.21       purrr_0.3.4         BiocManager_1.30.10
## [25] codetools_0.2-18    vctrs_0.3.5         glue_1.4.2          evaluate_0.14      
## [29] rmarkdown_2.5       blogdown_1.1.7      stringi_1.5.3       pillar_1.4.7       
## [33] compiler_4.0.4      generics_0.1.0      scales_1.1.1        pkgconfig_2.0.3
</code></pre>
<h2 id="references">References</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Gentleman2008-xo" class="csl-entry">
<p>Gentleman, Robert. 2008. <em>R Programming for Bioinformatics (Chapman &amp; Hall/CRC Computer Science &amp; Data Analysis)</em>. 1 edition. Chapman; Hall/CRC. <a href="http://www.amazon.com/Programming-Bioinformatics-Chapman-Computer-Analysis/dp/1420063677">http://www.amazon.com/Programming-Bioinformatics-Chapman-Computer-Analysis/dp/1420063677</a>.</p>
</div>
</div>

	
	
	<div class="text-muted mt-5 pt-3 border-top">Last modified 2021-05-01: <a  href="https://github.com/tgirke/GEN242//commit/33fa745d116679e34731b56225f58ff5e13abeca">some edits (33fa745d1)</a>
</div>
</div>


              </main>
          


        </div>
      </div>
      
<footer class="bg-dark py-5 row d-print-none">
  <div class="container-fluid mx-sm-5">
    <div class="row">
      <div class="col-6 col-sm-4 text-xs-center order-sm-2">
        
      </div>
      <div class="col-6 col-sm-4 text-right text-xs-center order-sm-3">
        
      </div>
      <div class="col-12 col-sm-4 text-center py-2 order-sm-2">
        <small class="text-white">&copy; 2021 The GEN242 Instructors All Rights Reserved</small>
        <small class="ml-1"><a href="https://www.ucr.edu/privacy.html" target="_blank">Privacy Policy</a></small>
	
      </div>
    </div>
  </div>
</footer>


    </div>
    
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>


<script>
  (function() {
    var i, text, code, codes = document.getElementsByTagName('code');
    for (i = 0; i < codes.length;) {
      code = codes[i];
      if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
        text = code.textContent;
        if (/^\$[^$]/.test(text) && /[^$]\$$/.test(text)) {
          text = text.replace(/^\$/, '\\(').replace(/\$$/, '\\)');
          code.textContent = text;
        }
        if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
            /^\$(.|\s)+\$$/.test(text) ||
            /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
          code.outerHTML = code.innerHTML;  
          continue;
        }
      }
      i++;
    }
  })();
</script>

<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script>













<script src="../../../js/main.min.f820bfc507c3ebefe17f8881ec4ee10013f3b89ad9555a50acf9bb1771e6823f.js" integrity="sha256-&#43;CC/xQfD6&#43;/hf4iB7E7hABPzuJrZVVpQrPm7F3Hmgj8=" crossorigin="anonymous"></script>



<script src='../../../js/prism.js'></script>



  </body>
</html>
