<!doctype html>
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="generator" content="Hugo 0.80.0" /><META NAME="ROBOTS" CONTENT="INDEX, FOLLOW">


<link rel="shortcut icon" href="../../../favicons/favicon.ico" >
<link rel="apple-touch-icon" href="../../../favicons/apple-touch-icon-180x180.png" sizes="180x180">
<link rel="icon" type="image/png" href="../../../favicons/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="../../../favicons/favicon-32x32.png" sizes="32x32">
<link rel="icon" type="image/png" href="../../../favicons/android-36x36.png" sizes="36x36">
<link rel="icon" type="image/png" href="../../../favicons/android-48x48.png" sizes="48x48">
<link rel="icon" type="image/png" href="../../../favicons/android-72x72.png" sizes="72x72">
<link rel="icon" type="image/png" href="../../../favicons/android-96x96.png" sizes="96x96">
<link rel="icon" type="image/png" href="../../../favicons/android-144x144.png" sizes="144x144">
<link rel="icon" type="image/png" href="../../../favicons/android-192x192.png" sizes="192x192">

<title>ChIP-Seq Workflow Template | GEN242</title><meta property="og:title" content="ChIP-Seq Workflow Template" />
<meta property="og:description" content="pre code { white-space: pre !important; overflow-x: scroll !important; word-break: keep-all !important; word-wrap: initial !important; }  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Introduction Users want to provide here background information about the design of their ChIP-Seq project.
Background and objectives This report describes the analysis of several ChIP-Seq experiments studying the DNA binding patterns of the transcriptions factors … from organism …." />
<meta property="og:type" content="article" />
<meta property="og:url" content="/manuals/spchipseq/spchipseq/" />
<meta property="article:modified_time" content="2021-05-01T15:18:21-07:00" /><meta property="og:site_name" content="GEN242" />
<meta itemprop="name" content="ChIP-Seq Workflow Template">
<meta itemprop="description" content="pre code { white-space: pre !important; overflow-x: scroll !important; word-break: keep-all !important; word-wrap: initial !important; }  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Introduction Users want to provide here background information about the design of their ChIP-Seq project.
Background and objectives This report describes the analysis of several ChIP-Seq experiments studying the DNA binding patterns of the transcriptions factors … from organism ….">
<meta itemprop="dateModified" content="2021-05-01T15:18:21-07:00" />
<meta itemprop="wordCount" content="3659">



<meta itemprop="keywords" content="" />
<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="ChIP-Seq Workflow Template"/>
<meta name="twitter:description" content="pre code { white-space: pre !important; overflow-x: scroll !important; word-break: keep-all !important; word-wrap: initial !important; }  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Introduction Users want to provide here background information about the design of their ChIP-Seq project.
Background and objectives This report describes the analysis of several ChIP-Seq experiments studying the DNA binding patterns of the transcriptions factors … from organism …."/>


<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-00000000-0', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>






<link rel="preload" href="../../../scss/main.min.1f416e6a5f646f231d085681bf2b4ebdad5d41f149a3977625affc97f735ab7e.css" as="style">
<link href="../../../scss/main.min.1f416e6a5f646f231d085681bf2b4ebdad5d41f149a3977625affc97f735ab7e.css" rel="stylesheet" integrity="">


<script
  src="https://code.jquery.com/jquery-3.5.1.min.js"
  integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0="
  crossorigin="anonymous"></script>

<script
  src="https://unpkg.com/lunr@2.3.8/lunr.min.js"
  integrity="sha384-vRQ9bDyE0Wnu+lMfm57BlYLO0/XauFuKpVsZPs7KEDwYKktWi5+Kz3MP8++DFlRY"
  crossorigin="anonymous"></script>



<link rel="stylesheet" href="../../../css/prism.css"/>


<script src="../../../js/driver0.9.8.min.js"></script>
<link rel="stylesheet" href="../../../css/driver0.9.8.min.css"/>


<link rel="stylesheet" href="../../../css/site.css"/>
<script src="../../../js/site.js"></script>



  </head>
  <body class="td-page line-numbers">
    <header>
     
<nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar">
  <a class="navbar-brand" href="../../../">
		<span class="font-weight-bold">GEN242</span>
	</a>
	<row class="header-switch">
	  	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="sidebarSwitch" checked=true>
        <label class="custom-control-label" for="sidebarSwitch">sidebar</label>
      </div>
    	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="tocSwitch" checked=true>
        <label class="custom-control-label" for="tocSwitch">toc</label>
      </div>
	</row>
	<div class="td-navbar-nav-scroll ml-md-auto" id="main_navbar">
		<ul class="navbar-nav mt-2 mt-lg-0">
			
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../about/" id="navbarDropdown" role="button">
          	About
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/introduction/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/syllabus/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Syllabus
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/schedule/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Schedule
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/internal/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Internal
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/internal_resources/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Resources
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/install/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Installation
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../slides/" id="navbarDropdown" role="button">
          	Slides
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_01"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Course Introduction (S1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_02"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Genome Basics (S2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_03"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Databases and Software (S3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_04"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 DNA Sequencing (S4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_05"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (S5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_06"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sequence Alignments (S6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_07"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (S7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_08"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Multiple Alignments (S8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_09"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Short Read Alignments (S9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_10"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R on HPC (S10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_11"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (S11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_12"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Gene Expression Analysis (S12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_13"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Workflows (S13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_14/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Overview (S14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_15/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Enrichment Analysis (S15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_16/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis (S16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_17/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R Packages and dplyr (S17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_18/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Shiny Apps (S18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_19/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Profile HMMs (S19)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_20/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sequence Assembly (S20)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../tutorials/" id="navbarDropdown" role="button">
          	Tutorials
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/github/github/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Introduction (T1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/linux/linux/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux and HPCC Cluster (T2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rbasics/rbasics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (T3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rprogramming/rprogramming/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (T4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rparallel/rparallel/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Parallel R (T5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rsequences/rsequences/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (T6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/systempiper/systempiper/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflows (T7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/sprnaseq/sprnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 RNA-Seq Workflow (T8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/spchipseq/spchipseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Workflow (T9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/tutorial_09"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 VAR-Seq Workflow (T10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/tutorial_10"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics and Visualization (T11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/tutorial_12"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis and Data Mining (T11)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../manuals/" id="navbarDropdown" role="button">
          	Manuals
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/github/github/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/linux/linux/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux &amp; HPCC Cluster
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rbasics/rbasics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rbasics/sample/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sample R Markdown
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rprogramming/rprogramming/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rparallel/rparallel/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Parallel Evaluations in R
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rsequences/rsequences/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/systempiper/systempiper/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Workflow Overview
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/sprnaseq/sprnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 RNA-Seq Workflow
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/spchipseq/spchipseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Workflow
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rgraphics/rgraphics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics and Visualization
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../assignments/" id="navbarDropdown" role="button">
          	Assignments
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/homework/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Homework
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw01/hw01/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW1: Intro &amp; GitHub
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw02/hw02/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW2: Linux &amp; HPC
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw03/hw03/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW3: R Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw04/hw04/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW4: Alignments
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw05/hw05"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW5: R Programming
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw06/hw06"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW6: NGS Analysis Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw07/hw07"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW7: RNA-Seq Analysis
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_overview/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Overview of Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/01_rnaseq_aligners/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 RNA-Seq Aligners
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/02_rnaseq_deg/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEG Methods
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/03_cluster_analysis/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Cluster Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/04_scrnaseq_embedding/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 scRNA-Seq Embedding
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/05_chipseq_peakcaller/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 ChIP-Seq Peak Callers
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/06_functional_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Functional Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/07_motif_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Motif Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/08_genome_graphics/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Genome Graphics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Project Data
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Presentations
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/paper_presentations/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Paper Presentations
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../external_resources/" id="navbarDropdown" role="button">
          	Links
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="https://piazza.com/ucr/spring2021/gen242/home"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Piazza
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://classroom.github.com/classrooms/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Classroom
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://github.com/tgirke/GEN242"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Repo
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://bioconductor.org/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Bioconductor
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://www.dcassol.com/docsy_Rmarkdown/rmarkdown/install/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Hugo, Docsy &amp; R
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
				
				
					<a class="nav-link" href="../../../blog/" ><span>News</span></a>
				
			</li>
			
			
			
		</ul>
	</div>
	<div class="navbar-nav d-none d-lg-block">



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.6b8c347c43c93c00b1245c91f31641a0.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>

</div>
</nav>

    </header>
    <div class="container-fluid td-outer">
      <div class="td-main">
        <div class="row flex-xl-nowrap">
          <div class="col-12 col-md-3 col-xl-2 td-sidebar d-print-none">
            




<div id="td-sidebar-menu" class="td-sidebar__inner">
  
  <form class="td-sidebar__search d-flex align-items-center">
    



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.6b8c347c43c93c00b1245c91f31641a0.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>


    <button class="btn btn-link td-sidebar__toggle d-md-none p-0 ml-3 fas fa-bars" type="button" data-toggle="collapse" data-target="#td-section-nav" aria-controls="td-docs-nav" aria-expanded="false" aria-label="Toggle section navigation">
    </button>
  </form>
  
  <nav class="collapse td-sidebar-nav" id="td-section-nav">
    
    
    






<ul class="td-sidebar-nav__section pr-md-3">
  <li class="td-sidebar-nav__section-title">
    <a  href="../../../manuals/" class="align-left pl-0 pr-2 active td-sidebar-link td-sidebar-link__section">Manuals</a>
  </li>
  <ul>
    <li class="collapse show" id="manuals">
      
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsgithubgithub" href="../../../manuals/github/github/">GitHub</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualslinuxlinux" href="../../../manuals/linux/linux/">Linux &amp; HPCC Cluster</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrbasicsrbasics" href="../../../manuals/rbasics/rbasics/">Introduction to R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrprogrammingrprogramming" href="../../../manuals/rprogramming/rprogramming/">Programming in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrbasicssample" href="../../../manuals/rbasics/sample/">Sample R Markdown</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrparallelrparallel" href="../../../manuals/rparallel/rparallel/">Parallel Evaluations in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrsequencesrsequences" href="../../../manuals/rsequences/rsequences/">NGS Analysis Basics</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualssystempipersystempiper" href="../../../manuals/systempiper/systempiper/">systemPipeR: Workflow design and reporting generation environment</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualssprnaseqsprnaseq" href="../../../manuals/sprnaseq/sprnaseq/">RNA-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page  active" id="m-manualsspchipseqspchipseq" href="../../../manuals/spchipseq/spchipseq/">ChIP-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrgraphicsrgraphics" href="../../../manuals/rgraphics/rgraphics/">Graphics and Data Visualization in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrclusteringrclustering" href="../../../manuals/rclustering/rclustering/">Cluster Analysis in R</a>
        
      
      
    </li>
  </ul>
</ul>

  </nav>
</div>




          </div>
          
          

          
            
          

          
          
              <div class="d-none d-xl-block col-xl-2 td-toc d-print-none">
                
  
  
  
  
  
  <div class="td-page-meta ml-2 pb-1 pt-2 mb-0">
  
    
    
      
    
    
    
    
    
    
    
    


    
    
      <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/manuals/spchipseq/spchipseq.Rmd" target="_blank" class="source-link"><i class="fa fa-code fa-fw"></i> View Rmd source</a>
    

    <a href="https://github.com/tgirke/GEN242//edit/main/content/en/manuals/spchipseq/spchipseq.md" target="_blank"><i class="fa fa-edit fa-fw"></i> Edit this page</a>
    <a href="https://github.com/tgirke/GEN242//new/main/content/en/manuals/spchipseq/spchipseq.md?filename=change-me.md&amp;value=---%0Atitle%3A&#43;%22Long&#43;Page&#43;Title%22%0AlinkTitle%3A&#43;%22Short&#43;Nav&#43;Title%22%0Aweight%3A&#43;100%0Adescription%3A&#43;%3E-%0A&#43;&#43;&#43;&#43;&#43;Page&#43;description&#43;for&#43;heading&#43;and&#43;indexes.%0A---%0A%0A%23%23&#43;Heading%0A%0AEdit&#43;this&#43;template&#43;to&#43;create&#43;your&#43;new&#43;page.%0A%0A%2A&#43;Give&#43;it&#43;a&#43;good&#43;name%2C&#43;ending&#43;in&#43;%60.md%60&#43;-&#43;e.g.&#43;%60getting-started.md%60%0A%2A&#43;Edit&#43;the&#43;%22front&#43;matter%22&#43;section&#43;at&#43;the&#43;top&#43;of&#43;the&#43;page&#43;%28weight&#43;controls&#43;how&#43;its&#43;ordered&#43;amongst&#43;other&#43;pages&#43;in&#43;the&#43;same&#43;directory%3B&#43;lowest&#43;number&#43;first%29.%0A%2A&#43;Add&#43;a&#43;good&#43;commit&#43;message&#43;at&#43;the&#43;bottom&#43;of&#43;the&#43;page&#43;%28%3C80&#43;characters%3B&#43;use&#43;the&#43;extended&#43;description&#43;field&#43;for&#43;more&#43;detail%29.%0A%2A&#43;Create&#43;a&#43;new&#43;branch&#43;so&#43;you&#43;can&#43;preview&#43;your&#43;new&#43;file&#43;and&#43;request&#43;a&#43;review&#43;via&#43;Pull&#43;Request.%0A" target="_blank"><i class="fa fa-edit fa-fw"></i> Create child page</a>
    <a href="https://github.com/tgirke/GEN242//issues/new?title=ChIP-Seq%20Workflow%20Template" target="_blank"><i class="fab fa-github fa-fw"></i> Create documentation issue</a>
    
      
      <a href="https://github.com/tgirke/GEN242/issues/new" target="_blank"><i class="fas fa-tasks fa-fw"></i> Create project issue</a>
    

  
  
  </div>





<nav id="TableOfContents">
  <ul>
    <li><a href="#introduction">Introduction</a>
      <ul>
        <li><a href="#background-and-objectives">Background and objectives</a></li>
        <li><a href="#experimental-design">Experimental design</a></li>
      </ul>
    </li>
    <li><a href="#workflow-environment">Workflow environment</a>
      <ul>
        <li><a href="#load-packages">Load packages</a></li>
        <li><a href="#generate-workflow-environment">Generate workflow environment</a></li>
        <li><a href="#run-workflow">Run workflow</a></li>
      </ul>
    </li>
    <li><a href="#read-preprocessing">Read preprocessing</a>
      <ul>
        <li><a href="#experiment-definition-provided-by-targets-file">Experiment definition provided by <code>targets</code> file</a></li>
        <li><a href="#read-quality-filtering-and-trimming">Read quality filtering and trimming</a></li>
        <li><a href="#fastq-quality-report">FASTQ quality report</a></li>
      </ul>
    </li>
    <li><a href="#alignments">Alignments</a>
      <ul>
        <li><a href="#read-mapping-with-bowtie2">Read mapping with <code>Bowtie2</code></a></li>
        <li><a href="#read-and-alignment-stats">Read and alignment stats</a></li>
        <li><a href="#create-symbolic-links-for-viewing-bam-files-in-igv">Create symbolic links for viewing BAM files in IGV</a></li>
      </ul>
    </li>
    <li><a href="#utilities-for-coverage-data">Utilities for coverage data</a>
      <ul>
        <li><a href="#rle-object-stores-coverage-information">Rle object stores coverage information</a></li>
        <li><a href="#resizing-aligned-reads">Resizing aligned reads</a></li>
        <li><a href="#naive-peak-calling">Naive peak calling</a></li>
        <li><a href="#plot-coverage-for-defined-region">Plot coverage for defined region</a></li>
      </ul>
    </li>
    <li><a href="#peak-calling-with-macs2">Peak calling with MACS2</a>
      <ul>
        <li><a href="#merge-bam-files-of-replicates-prior-to-peak-calling">Merge BAM files of replicates prior to peak calling</a></li>
        <li><a href="#peak-calling-without-inputreference-sample">Peak calling without input/reference sample</a></li>
        <li><a href="#peak-calling-with-inputreference-sample">Peak calling with input/reference sample</a></li>
        <li><a href="#identify-consensus-peaks">Identify consensus peaks</a></li>
      </ul>
    </li>
    <li><a href="#annotate-peaks-with-genomic-context">Annotate peaks with genomic context</a>
      <ul>
        <li><a href="#annotation-with-chippeakanno-package">Annotation with <code>ChIPpeakAnno</code> package</a></li>
        <li><a href="#annotation-with-chipseeker-package">Annotation with <code>ChIPseeker</code> package</a></li>
      </ul>
    </li>
    <li><a href="#count-reads-overlapping-peaks">Count reads overlapping peaks</a></li>
    <li><a href="#differential-binding-analysis">Differential binding analysis</a></li>
    <li><a href="#go-term-enrichment-analysis">GO term enrichment analysis</a></li>
    <li><a href="#motif-analysis">Motif analysis</a>
      <ul>
        <li><a href="#parse-dna-sequences-of-peak-regions-from-genome">Parse DNA sequences of peak regions from genome</a></li>
        <li><a href="#motif-discovery-with-bcrank">Motif discovery with <code>BCRANK</code></a></li>
      </ul>
    </li>
    <li><a href="#version-information">Version Information</a></li>
    <li><a href="#funding">Funding</a></li>
    <li><a href="#references">References</a></li>
  </ul>
</nav>



              </div>
              <main class="col-12 col-md-9 col-xl-8 pl-md-5" role="main">
                
  

                <nav aria-label="breadcrumb" class="d-none d-md-block d-print-none">
	<ol class="breadcrumb spb-1">
		







<li class="breadcrumb-item" >
	<a href="../../../manuals/">Manuals</a>
</li>




<li class="breadcrumb-item active" aria-current="page">
	<a href="../../../manuals/spchipseq/spchipseq/">ChIP-Seq Workflow Template</a>
</li>

	</ol>
</nav	>

                
<div class="td-content">
	<h1>ChIP-Seq Workflow Template</h1>
    
	
        <p class="reading-time"><i class="fa fa-clock" aria-hidden="true"></i> 18 minute read</p>
           
	<style type="text/css">
pre code {
white-space: pre !important;
overflow-x: scroll !important;
word-break: keep-all !important;
word-wrap: initial !important;
}
</style>
<!--
- Compile from command-line
Rscript -e "rmarkdown::render('spchipseq.Rmd', c('BiocStyle::html_document'), clean=F); knitr::knit('spchipseq.Rmd', tangle=TRUE)"; Rscript -e "rmarkdown::render('spchipseq.Rmd', c('BiocStyle::pdf_document'))"
-->
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>
<h2 id="introduction">Introduction</h2>
<p>Users want to provide here background information about the design of their ChIP-Seq project.</p>
<h3 id="background-and-objectives">Background and objectives</h3>
<p>This report describes the analysis of several ChIP-Seq experiments
studying the DNA binding patterns of the transcriptions factors … from <em>organism</em> ….</p>
<h3 id="experimental-design">Experimental design</h3>
<p>Typically, users want to specify here all information relevant for the
analysis of their NGS study. This includes detailed descriptions of
FASTQ files, experimental design, reference genome, gene annotations,
etc.</p>
<h2 id="workflow-environment">Workflow environment</h2>
<h3 id="load-packages">Load packages</h3>
<p>The <code>systemPipeR</code> package needs to be loaded to perform the analysis
steps shown in this report (H Backman and Girke 2016). The package allows users
to run the entire analysis workflow interactively or with a single command
while also generating the corresponding analysis report. For details
see <code>systemPipeR's</code> main <a href="http://www.bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html">vignette</a>.</p>
<pre><code class="language-r">library(systemPipeR)
</code></pre>
<h3 id="generate-workflow-environment">Generate workflow environment</h3>
<p><a href="http://bioconductor.org/packages/release/data/experiment/html/systemPipeRdata.html"><em>systemPipeRdata</em></a> package is a helper package to generate a fully populated <a href="http://bioconductor.org/packages/release/bioc/html/systemPipeR.html"><em>systemPipeR</em></a>
workflow environment in the current working directory with a single command.
All the instruction for generating the workflow template are provide in the <em>systemPipeRdata</em> vignette <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRdata.html#1_Introduction">here</a>.</p>
<p>After building and loading the workflow environment generated by <code>genWorkenvir</code>
from <em>systemPipeRdata</em> all data inputs are stored in
a <code>data/</code> directory and all analysis results will be written to a separate
<code>results/</code> directory, while the <code>systemPipeChIPseq.Rmd</code> script and the <code>targets</code> file are expected to be located in the parent directory. The R session is expected to run from this parent
directory. Additional parameter files are stored under <code>param/</code>.</p>
<p>To work with real data, users want to organize their own data similarly
and substitute all test data for their own data. To rerun an established
workflow on new data, the initial <code>targets</code> file along with the corresponding
FASTQ files are usually the only inputs the user needs to provide.</p>
<p>For more details, please consult the documentation
<a href="http://www.bioconductor.org/packages/release/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html">here</a>. More information about the <code>targets</code> files from <em>systemPipeR</em> can be found <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html#25_structure_of_targets_file">here</a>.</p>
<h3 id="run-workflow">Run workflow</h3>
<p>Now open the R markdown script <code>systemPipeRIBOseq.Rmd</code>in your R IDE (<em>e.g.</em> vim-r or RStudio) and
run the workflow as outlined below.</p>
<p>Here pair-end workflow example is provided. Please refer to the main vignette
<code>systemPipeR.Rmd</code> for running the workflow with single-end data.</p>
<p>If you are running on a single machine, use following code as an example to check
if some tools used in this workflow are in your environment <strong>PATH</strong>. No warning
message should be shown if all tools are installed.</p>
<h2 id="read-preprocessing">Read preprocessing</h2>
<h3 id="experiment-definition-provided-by-targets-file">Experiment definition provided by <code>targets</code> file</h3>
<p>The <code>targets</code> file defines all FASTQ files and sample comparisons of the analysis workflow.</p>
<pre><code class="language-r">targetspath &lt;- system.file(&quot;extdata&quot;, &quot;targetsPE_chip.txt&quot;, package = &quot;systemPipeR&quot;)
targets &lt;- read.delim(targetspath, comment.char = &quot;#&quot;)
targets[1:4, -c(5, 6)]
</code></pre>
<pre><code>##                     FileName1                   FileName2
## 1 ./data/SRR446027_1.fastq.gz ./data/SRR446027_2.fastq.gz
## 2 ./data/SRR446028_1.fastq.gz ./data/SRR446028_2.fastq.gz
## 3 ./data/SRR446029_1.fastq.gz ./data/SRR446029_2.fastq.gz
## 4 ./data/SRR446030_1.fastq.gz ./data/SRR446030_2.fastq.gz
##   SampleName Factor        Date SampleReference
## 1        M1A     M1 23-Mar-2012                
## 2        M1B     M1 23-Mar-2012                
## 3        A1A     A1 23-Mar-2012             M1A
## 4        A1B     A1 23-Mar-2012             M1B
</code></pre>
<h3 id="read-quality-filtering-and-trimming">Read quality filtering and trimming</h3>
<p>The following example shows how one can design a custom read
preprocessing function using utilities provided by the <code>ShortRead</code> package, and then
apply it with <code>preprocessReads</code> in batch mode to all FASTQ samples referenced in the
corresponding <code>SYSargs2</code> instance (<code>trim</code> object below). More detailed information on
read preprocessing is provided in <code>systemPipeR's</code> main vignette.</p>
<p>First, we construct <em><code>SYSargs2</code></em> object from <em><code>cwl</code></em> and <em><code>yml</code></em> param and <em><code>targets</code></em> files.</p>
<pre><code class="language-r">dir_path &lt;- system.file(&quot;extdata/cwl/preprocessReads/trim-pe&quot;, 
    package = &quot;systemPipeR&quot;)
trim &lt;- loadWF(targets = targetspath, wf_file = &quot;trim-pe.cwl&quot;, 
    input_file = &quot;trim-pe.yml&quot;, dir_path = dir_path)
trim &lt;- renderWF(trim, inputvars = c(FileName1 = &quot;_FASTQ_PATH1_&quot;, 
    FileName2 = &quot;_FASTQ_PATH2_&quot;, SampleName = &quot;_SampleName_&quot;))
trim
output(trim)[1:2]
</code></pre>
<p>Next, we execute the code for trimming all the raw data.</p>
<pre><code class="language-r">filterFct &lt;- function(fq, cutoff = 20, Nexceptions = 0) {
    qcount &lt;- rowSums(as(quality(fq), &quot;matrix&quot;) &lt;= cutoff, na.rm = TRUE)
    fq[qcount &lt;= Nexceptions]
    # Retains reads where Phred scores are &gt;= cutoff with N
    # exceptions
}
preprocessReads(args = trim, Fct = &quot;filterFct(fq, cutoff=20, Nexceptions=0)&quot;, 
    batchsize = 1e+05)
writeTargetsout(x = trim, file = &quot;targets_chip_trimPE.txt&quot;, step = 1, 
    new_col = c(&quot;FileName1&quot;, &quot;FileName2&quot;), new_col_output_index = c(1, 
        2), overwrite = TRUE)
</code></pre>
<h3 id="fastq-quality-report">FASTQ quality report</h3>
<p>The following <code>seeFastq</code> and <code>seeFastqPlot</code> functions generate and plot a series
of useful quality statistics for a set of FASTQ files including per cycle quality box
plots, base proportions, base-level quality trends, relative k-mer
diversity, length and occurrence distribution of reads, number of reads
above quality cutoffs and mean quality distribution. The results are
written to a PDF file named <code>fastqReport.pdf</code>.
Parallelization of FASTQ quality report via scheduler (<em>e.g.</em> Slurm) across several compute nodes.</p>
<pre><code class="language-r">library(BiocParallel)
library(batchtools)
f &lt;- function(x) {
    library(systemPipeR)
    targets &lt;- system.file(&quot;extdata&quot;, &quot;targetsPE_chip.txt&quot;, package = &quot;systemPipeR&quot;)
    dir_path &lt;- system.file(&quot;extdata/cwl/preprocessReads/trim-pe&quot;, 
        package = &quot;systemPipeR&quot;)
    trim &lt;- loadWorkflow(targets = targets, wf_file = &quot;trim-pe.cwl&quot;, 
        input_file = &quot;trim-pe.yml&quot;, dir_path = dir_path)
    trim &lt;- renderWF(trim, inputvars = c(FileName1 = &quot;_FASTQ_PATH1_&quot;, 
        FileName2 = &quot;_FASTQ_PATH2_&quot;, SampleName = &quot;_SampleName_&quot;))
    seeFastq(fastq = infile1(trim)[x], batchsize = 1e+05, klength = 8)
}

resources &lt;- list(walltime = 120, ntasks = 1, ncpus = 4, memory = 1024)
param &lt;- BatchtoolsParam(workers = 4, cluster = &quot;slurm&quot;, template = &quot;batchtools.slurm.tmpl&quot;, 
    resources = resources)
fqlist &lt;- bplapply(seq(along = trim), f, BPPARAM = param)

pdf(&quot;./results/fastqReport.pdf&quot;, height = 18, width = 4 * length(fqlist))
seeFastqPlot(unlist(fqlist, recursive = FALSE))
dev.off()
</code></pre>
<p><img src="../results/fastqReport.png" alt=""></p>
<div align="center">
<p>Figure 1: FASTQ quality report for 18 samples</p>
</div>
</br>
<h2 id="alignments">Alignments</h2>
<h3 id="read-mapping-with-bowtie2">Read mapping with <code>Bowtie2</code></h3>
<p>The NGS reads of this project will be aligned with <code>Bowtie2</code> against the
reference genome sequence (Langmead and Salzberg 2012). The parameter settings of the
aligner are defined in the <code>bowtie2-index.cwl</code> and <code>bowtie2-index.yml</code> files.
In ChIP-Seq experiments it is usually more appropriate to eliminate reads mapping
to multiple locations. To achieve this, users want to remove the argument setting
<code>-k 50 non-deterministic</code> in the configuration files.</p>
<p>Building the index:</p>
<pre><code class="language-r">dir_path &lt;- system.file(&quot;extdata/cwl/bowtie2/bowtie2-idx&quot;, package = &quot;systemPipeR&quot;)
idx &lt;- loadWorkflow(targets = NULL, wf_file = &quot;bowtie2-index.cwl&quot;, 
    input_file = &quot;bowtie2-index.yml&quot;, dir_path = dir_path)
idx &lt;- renderWF(idx)
idx
cmdlist(idx)

### Run in single machine
runCommandline(idx, make_bam = FALSE)
</code></pre>
<p>The following submits 18 alignment jobs via a scheduler to a computer cluster.</p>
<pre><code class="language-r">targets &lt;- system.file(&quot;extdata&quot;, &quot;targetsPE_chip.txt&quot;, package = &quot;systemPipeR&quot;)
dir_path &lt;- system.file(&quot;extdata/cwl/bowtie2/bowtie2-pe&quot;, package = &quot;systemPipeR&quot;)
args &lt;- loadWF(targets = targets, wf_file = &quot;bowtie2-mapping-pe.cwl&quot;, 
    input_file = &quot;bowtie2-mapping-pe.yml&quot;, dir_path = dir_path)
args &lt;- renderWF(args, inputvars = c(FileName1 = &quot;_FASTQ_PATH1_&quot;, 
    FileName2 = &quot;_FASTQ_PATH2_&quot;, SampleName = &quot;_SampleName_&quot;))
args
cmdlist(args)[1:2]
output(args)[1:2]
</code></pre>
<pre><code class="language-r">moduleload(modules(args))  # Skip if a module system is not used
resources &lt;- list(walltime = 120, ntasks = 1, ncpus = 4, memory = 1024)
reg &lt;- clusterRun(args, FUN = runCommandline, more.args = list(args = args, 
    dir = FALSE), conffile = &quot;.batchtools.conf.R&quot;, template = &quot;batchtools.slurm.tmpl&quot;, 
    Njobs = 18, runid = &quot;01&quot;, resourceList = resources)
getStatus(reg = reg)
waitForJobs(reg = reg)
args &lt;- output_update(args, dir = FALSE, replace = TRUE, extension = c(&quot;.sam&quot;, 
    &quot;.bam&quot;))  ## Updates the output(args) to the right location in the subfolders
output(args)
</code></pre>
<p>Alternatively, one can run the alignments sequentially on a single system.</p>
<pre><code class="language-r">args &lt;- runCommandline(args, force = F)
</code></pre>
<p>Check whether all BAM files have been created and write out the new targets file.</p>
<pre><code class="language-r">writeTargetsout(x = args, file = &quot;targets_bam.txt&quot;, step = 1, 
    new_col = &quot;FileName&quot;, new_col_output_index = 1, overwrite = TRUE)
outpaths &lt;- subsetWF(args, slot = &quot;output&quot;, subset = 1, index = 1)
file.exists(outpaths)
</code></pre>
<h3 id="read-and-alignment-stats">Read and alignment stats</h3>
<p>The following provides an overview of the number of reads in each sample
and how many of them aligned to the reference.</p>
<pre><code class="language-r">read_statsDF &lt;- alignStats(args = args)
write.table(read_statsDF, &quot;results/alignStats.xls&quot;, row.names = FALSE, 
    quote = FALSE, sep = &quot;\t&quot;)
read.delim(&quot;results/alignStats.xls&quot;)
</code></pre>
<h3 id="create-symbolic-links-for-viewing-bam-files-in-igv">Create symbolic links for viewing BAM files in IGV</h3>
<p>The <code>symLink2bam</code> function creates symbolic links to view the BAM alignment files in a
genome browser such as IGV without moving these large files to a local
system. The corresponding URLs are written to a file with a path
specified under <code>urlfile</code>, here <code>IGVurl.txt</code>. Please replace the directory and the user name.</p>
<pre><code class="language-r">symLink2bam(sysargs = args, htmldir = c(&quot;~/.html/&quot;, &quot;somedir/&quot;), 
    urlbase = &quot;http://cluster.hpcc.ucr.edu/~tgirke/&quot;, urlfile = &quot;./results/IGVurl.txt&quot;)
</code></pre>
<h2 id="utilities-for-coverage-data">Utilities for coverage data</h2>
<p>The following introduces several utilities useful for ChIP-Seq data. They are not part of the actual workflow.</p>
<h3 id="rle-object-stores-coverage-information">Rle object stores coverage information</h3>
<pre><code class="language-r">library(rtracklayer)
library(GenomicRanges)
library(Rsamtools)
library(GenomicAlignments)
outpaths &lt;- subsetWF(args, slot = &quot;output&quot;, subset = 1, index = 1)
aligns &lt;- readGAlignments(outpaths[1])
cov &lt;- coverage(aligns)
cov
</code></pre>
<h3 id="resizing-aligned-reads">Resizing aligned reads</h3>
<pre><code class="language-r">trim(resize(as(aligns, &quot;GRanges&quot;), width = 200))
</code></pre>
<h3 id="naive-peak-calling">Naive peak calling</h3>
<pre><code class="language-r">islands &lt;- slice(cov, lower = 15)
islands[[1]]
</code></pre>
<h3 id="plot-coverage-for-defined-region">Plot coverage for defined region</h3>
<pre><code class="language-r">library(ggbio)
myloc &lt;- c(&quot;Chr1&quot;, 1, 1e+05)
ga &lt;- readGAlignments(outpaths[1], use.names = TRUE, param = ScanBamParam(which = GRanges(myloc[1], 
    IRanges(as.numeric(myloc[2]), as.numeric(myloc[3])))))
autoplot(ga, aes(color = strand, fill = strand), facets = strand ~ 
    seqnames, stat = &quot;coverage&quot;)
</code></pre>
<h2 id="peak-calling-with-macs2">Peak calling with MACS2</h2>
<h3 id="merge-bam-files-of-replicates-prior-to-peak-calling">Merge BAM files of replicates prior to peak calling</h3>
<p>Merging BAM files of technical and/or biological replicates can improve
the sensitivity of the peak calling by increasing the depth of read
coverage. The <code>mergeBamByFactor</code> function merges BAM files based on grouping information
specified by a <code>factor</code>, here the <code>Factor</code> column of the imported targets file. It
also returns an updated <code>SYSargs2</code> object containing the paths to the
merged BAM files as well as to any unmerged files without replicates.
This step can be skipped if merging of BAM files is not desired.</p>
<pre><code class="language-r">dir_path &lt;- system.file(&quot;extdata/cwl/mergeBamByFactor&quot;, package = &quot;systemPipeR&quot;)
args &lt;- loadWF(targets = &quot;targets_bam.txt&quot;, wf_file = &quot;merge-bam.cwl&quot;, 
    input_file = &quot;merge-bam.yml&quot;, dir_path = dir_path)
args &lt;- renderWF(args, inputvars = c(FileName = &quot;_BAM_PATH_&quot;, 
    SampleName = &quot;_SampleName_&quot;))

args_merge &lt;- mergeBamByFactor(args = args, overwrite = TRUE)
writeTargetsout(x = args_merge, file = &quot;targets_mergeBamByFactor.txt&quot;, 
    step = 1, new_col = &quot;FileName&quot;, new_col_output_index = 1, 
    overwrite = TRUE)
</code></pre>
<pre><code class="language-r">## Skip if a module system is not used
module(&quot;list&quot;)
module(&quot;unload&quot;, &quot;miniconda2&quot;)
module(&quot;load&quot;, &quot;python/2.7.14&quot;)  # Make sure to set up your enviroment variable for MACS2 
</code></pre>
<h3 id="peak-calling-without-inputreference-sample">Peak calling without input/reference sample</h3>
<p>MACS2 can perform peak calling on ChIP-Seq data with and without input
samples (Zhang et al. 2008). The following performs peak calling without
input on all samples specified in the corresponding <code>args</code> object. Note, due to
the small size of the sample data, MACS2 needs to be run here with the
<code>nomodel</code> setting. For real data sets, users want to remove this parameter
in the corresponding <code>*.param</code> file(s).</p>
<pre><code class="language-r">dir_path &lt;- system.file(&quot;extdata/cwl/MACS2/MACS2-noinput/&quot;, package = &quot;systemPipeR&quot;)
args &lt;- loadWF(targets = &quot;targets_mergeBamByFactor.txt&quot;, wf_file = &quot;macs2.cwl&quot;, 
    input_file = &quot;macs2.yml&quot;, dir_path = dir_path)
args &lt;- renderWF(args, inputvars = c(FileName = &quot;_FASTQ_PATH1_&quot;, 
    SampleName = &quot;_SampleName_&quot;))

runCommandline(args, make_bam = FALSE, force = T)
outpaths &lt;- subsetWF(args, slot = &quot;output&quot;, subset = 1, index = 1)
file.exists(outpaths)
writeTargetsout(x = args, file = &quot;targets_macs.txt&quot;, step = 1, 
    new_col = &quot;FileName&quot;, new_col_output_index = 1, overwrite = TRUE)
</code></pre>
<h3 id="peak-calling-with-inputreference-sample">Peak calling with input/reference sample</h3>
<p>To perform peak calling with input samples, they can be most
conveniently specified in the <code>SampleReference</code> column of the initial
<code>targets</code> file. The <code>writeTargetsRef</code> function uses this information to create a <code>targets</code>
file intermediate for running MACS2 with the corresponding input samples.</p>
<pre><code class="language-r">writeTargetsRef(infile = &quot;targets_mergeBamByFactor.txt&quot;, outfile = &quot;targets_bam_ref.txt&quot;, 
    silent = FALSE, overwrite = TRUE)
dir_path &lt;- system.file(&quot;extdata/cwl/MACS2/MACS2-input/&quot;, package = &quot;systemPipeR&quot;)
args_input &lt;- loadWF(targets = &quot;targets_bam_ref.txt&quot;, wf_file = &quot;macs2-input.cwl&quot;, 
    input_file = &quot;macs2.yml&quot;, dir_path = dir_path)
args_input &lt;- renderWF(args_input, inputvars = c(FileName1 = &quot;_FASTQ_PATH1_&quot;, 
    FileName2 = &quot;_FASTQ_PATH2_&quot;, SampleName = &quot;_SampleName_&quot;))
cmdlist(args_input)[1]
### Run
args_input &lt;- runCommandline(args_input, make_bam = FALSE, force = T)
outpaths_input &lt;- subsetWF(args_input, slot = &quot;output&quot;, subset = 1, 
    index = 1)
file.exists(outpaths_input)
writeTargetsout(x = args_input, file = &quot;targets_macs_input.txt&quot;, 
    step = 1, new_col = &quot;FileName&quot;, new_col_output_index = 1, 
    overwrite = TRUE)
</code></pre>
<p>The peak calling results from MACS2 are written for each sample to
separate files in the <code>results</code> directory. They are named after the corresponding
files with extensions used by MACS2.</p>
<h3 id="identify-consensus-peaks">Identify consensus peaks</h3>
<p>The following example shows how one can identify consensus preaks among two peak sets sharing either a minimum absolute overlap and/or minimum relative overlap using the <code>subsetByOverlaps</code> or <code>olRanges</code> functions, respectively. Note, the latter is a custom function imported below by sourcing it.</p>
<pre><code class="language-r">## source('http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/My_R_Scripts/rangeoverlapper.R')
outpaths &lt;- subsetWF(args, slot = &quot;output&quot;, subset = 1, index = 1)  ## escolher um dos outputs index
peak_M1A &lt;- outpaths[&quot;M1A&quot;]
peak_M1A &lt;- as(read.delim(peak_M1A, comment = &quot;#&quot;)[, 1:3], &quot;GRanges&quot;)
peak_A1A &lt;- outpaths[&quot;A1A&quot;]
peak_A1A &lt;- as(read.delim(peak_A1A, comment = &quot;#&quot;)[, 1:3], &quot;GRanges&quot;)
(myol1 &lt;- subsetByOverlaps(peak_M1A, peak_A1A, minoverlap = 1))
# Returns any overlap
myol2 &lt;- olRanges(query = peak_M1A, subject = peak_A1A, output = &quot;gr&quot;)
# Returns any overlap with OL length information
myol2[values(myol2)[&quot;OLpercQ&quot;][, 1] &gt;= 50]
# Returns only query peaks with a minimum overlap of 50%
</code></pre>
<h2 id="annotate-peaks-with-genomic-context">Annotate peaks with genomic context</h2>
<h3 id="annotation-with-chippeakanno-package">Annotation with <code>ChIPpeakAnno</code> package</h3>
<p>The following annotates the identified peaks with genomic context information using the <code>ChIPpeakAnno</code> and <code>ChIPseeker</code> packages, respectively (Zhu et al. 2010; Yu, Wang, and He 2015).</p>
<pre><code class="language-r">library(ChIPpeakAnno)
library(GenomicFeatures)
dir_path &lt;- system.file(&quot;extdata/cwl/annotate_peaks&quot;, package = &quot;systemPipeR&quot;)
args &lt;- loadWF(targets = &quot;targets_macs.txt&quot;, wf_file = &quot;annotate-peaks.cwl&quot;, 
    input_file = &quot;annotate-peaks.yml&quot;, dir_path = dir_path)
args &lt;- renderWF(args, inputvars = c(FileName = &quot;_FASTQ_PATH1_&quot;, 
    SampleName = &quot;_SampleName_&quot;))

txdb &lt;- makeTxDbFromGFF(file = &quot;data/tair10.gff&quot;, format = &quot;gff&quot;, 
    dataSource = &quot;TAIR&quot;, organism = &quot;Arabidopsis thaliana&quot;)
ge &lt;- genes(txdb, columns = c(&quot;tx_name&quot;, &quot;gene_id&quot;, &quot;tx_type&quot;))
for (i in seq(along = args)) {
    peaksGR &lt;- as(read.delim(infile1(args)[i], comment = &quot;#&quot;), 
        &quot;GRanges&quot;)
    annotatedPeak &lt;- annotatePeakInBatch(peaksGR, AnnotationData = genes(txdb))
    df &lt;- data.frame(as.data.frame(annotatedPeak), as.data.frame(values(ge[values(annotatedPeak)$feature, 
        ])))
    outpaths &lt;- subsetWF(args, slot = &quot;output&quot;, subset = 1, index = 1)
    write.table(df, outpaths[i], quote = FALSE, row.names = FALSE, 
        sep = &quot;\t&quot;)
}
writeTargetsout(x = args, file = &quot;targets_peakanno.txt&quot;, step = 1, 
    new_col = &quot;FileName&quot;, new_col_output_index = 1, overwrite = TRUE)
</code></pre>
<p>The peak annotation results are written for each peak set to separate
files in the <code>results</code> directory. They are named after the corresponding peak
files with extensions specified in the <code>annotate_peaks.param</code> file,
here <code>*.peaks.annotated.xls</code>.</p>
<h3 id="annotation-with-chipseeker-package">Annotation with <code>ChIPseeker</code> package</h3>
<p>Same as in previous step but using the <code>ChIPseeker</code> package for annotating the peaks.</p>
<pre><code class="language-r">library(ChIPseeker)
for (i in seq(along = args)) {
    peakAnno &lt;- annotatePeak(infile1(args)[i], TxDb = txdb, verbose = FALSE)
    df &lt;- as.data.frame(peakAnno)
    outpaths &lt;- subsetWF(args, slot = &quot;output&quot;, subset = 1, index = 1)
    write.table(df, outpaths[i], quote = FALSE, row.names = FALSE, 
        sep = &quot;\t&quot;)
}
writeTargetsout(x = args, file = &quot;targets_peakanno.txt&quot;, step = 1, 
    new_col = &quot;FileName&quot;, new_col_output_index = 1, overwrite = TRUE)
</code></pre>
<p>Summary plots provided by the <code>ChIPseeker</code> package. Here applied only to one sample
for demonstration purposes.</p>
<pre><code class="language-r">peak &lt;- readPeakFile(infile1(args)[1])
covplot(peak, weightCol = &quot;X.log10.pvalue.&quot;)
outpaths &lt;- subsetWF(args, slot = &quot;output&quot;, subset = 1, index = 1)
peakHeatmap(outpaths[1], TxDb = txdb, upstream = 1000, downstream = 1000, 
    color = &quot;red&quot;)
plotAvgProf2(outpaths[1], TxDb = txdb, upstream = 1000, downstream = 1000, 
    xlab = &quot;Genomic Region (5'-&gt;3')&quot;, ylab = &quot;Read Count Frequency&quot;)
</code></pre>
<h2 id="count-reads-overlapping-peaks">Count reads overlapping peaks</h2>
<p>The <code>countRangeset</code> function is a convenience wrapper to perform read counting
iteratively over serveral range sets, here peak range sets. Internally,
the read counting is performed with the <code>summarizeOverlaps</code> function from the
<code>GenomicAlignments</code> package. The resulting count tables are directly saved to
files, one for each peak set.</p>
<pre><code class="language-r">library(GenomicRanges)
dir_path &lt;- system.file(&quot;extdata/cwl/count_rangesets&quot;, package = &quot;systemPipeR&quot;)
args &lt;- loadWF(targets = &quot;targets_macs.txt&quot;, wf_file = &quot;count_rangesets.cwl&quot;, 
    input_file = &quot;count_rangesets.yml&quot;, dir_path = dir_path)
args &lt;- renderWF(args, inputvars = c(FileName = &quot;_FASTQ_PATH1_&quot;, 
    SampleName = &quot;_SampleName_&quot;))

### Bam Files
targets &lt;- system.file(&quot;extdata&quot;, &quot;targetsPE_chip.txt&quot;, package = &quot;systemPipeR&quot;)
dir_path &lt;- system.file(&quot;extdata/cwl/bowtie2/bowtie2-pe&quot;, package = &quot;systemPipeR&quot;)
args_bam &lt;- loadWF(targets = targets, wf_file = &quot;bowtie2-mapping-pe.cwl&quot;, 
    input_file = &quot;bowtie2-mapping-pe.yml&quot;, dir_path = dir_path)
args_bam &lt;- renderWF(args_bam, inputvars = c(FileName = &quot;_FASTQ_PATH1_&quot;, 
    SampleName = &quot;_SampleName_&quot;))
args_bam &lt;- output_update(args_bam, dir = FALSE, replace = TRUE, 
    extension = c(&quot;.sam&quot;, &quot;.bam&quot;))
outpaths &lt;- subsetWF(args_bam, slot = &quot;output&quot;, subset = 1, index = 1)

bfl &lt;- BamFileList(outpaths, yieldSize = 50000, index = character())
countDFnames &lt;- countRangeset(bfl, args, mode = &quot;Union&quot;, ignore.strand = TRUE)
writeTargetsout(x = args, file = &quot;targets_countDF.txt&quot;, step = 1, 
    new_col = &quot;FileName&quot;, new_col_output_index = 1, overwrite = TRUE)
</code></pre>
<h2 id="differential-binding-analysis">Differential binding analysis</h2>
<p>The <code>runDiff</code> function performs differential binding analysis in batch mode for
several count tables using <code>edgeR</code> or <code>DESeq2</code> (Robinson, McCarthy, and Smyth 2010; Love, Huber, and Anders 2014).
Internally, it calls the functions <code>run_edgeR</code> and <code>run_DESeq2</code>. It also returns
the filtering results and plots from the downstream <code>filterDEGs</code> function using
the fold change and FDR cutoffs provided under the <code>dbrfilter</code> argument.</p>
<pre><code class="language-r">dir_path &lt;- system.file(&quot;extdata/cwl/rundiff&quot;, package = &quot;systemPipeR&quot;)
args_diff &lt;- loadWF(targets = &quot;targets_countDF.txt&quot;, wf_file = &quot;rundiff.cwl&quot;, 
    input_file = &quot;rundiff.yml&quot;, dir_path = dir_path)
args_diff &lt;- renderWF(args_diff, inputvars = c(FileName = &quot;_FASTQ_PATH1_&quot;, 
    SampleName = &quot;_SampleName_&quot;))

cmp &lt;- readComp(file = args_bam, format = &quot;matrix&quot;)
dbrlist &lt;- runDiff(args = args_diff, diffFct = run_edgeR, targets = targets.as.df(targets(args_bam)), 
    cmp = cmp[[1]], independent = TRUE, dbrfilter = c(Fold = 2, 
        FDR = 1))
writeTargetsout(x = args_diff, file = &quot;targets_rundiff.txt&quot;, 
    step = 1, new_col = &quot;FileName&quot;, new_col_output_index = 1, 
    overwrite = TRUE)
</code></pre>
<h2 id="go-term-enrichment-analysis">GO term enrichment analysis</h2>
<p>The following performs GO term enrichment analysis for each annotated peak set.</p>
<pre><code class="language-r">dir_path &lt;- system.file(&quot;extdata/cwl/annotate_peaks&quot;, package = &quot;systemPipeR&quot;)
args &lt;- loadWF(targets = &quot;targets_bam_ref.txt&quot;, wf_file = &quot;annotate-peaks.cwl&quot;, 
    input_file = &quot;annotate-peaks.yml&quot;, dir_path = dir_path)
args &lt;- renderWF(args, inputvars = c(FileName1 = &quot;_FASTQ_PATH1_&quot;, 
    FileName2 = &quot;_FASTQ_PATH2_&quot;, SampleName = &quot;_SampleName_&quot;))

args_anno &lt;- loadWF(targets = &quot;targets_macs.txt&quot;, wf_file = &quot;annotate-peaks.cwl&quot;, 
    input_file = &quot;annotate-peaks.yml&quot;, dir_path = dir_path)
args_anno &lt;- renderWF(args_anno, inputvars = c(FileName = &quot;_FASTQ_PATH1_&quot;, 
    SampleName = &quot;_SampleName_&quot;))
annofiles &lt;- subsetWF(args_anno, slot = &quot;output&quot;, subset = 1, 
    index = 1)
gene_ids &lt;- sapply(names(annofiles), function(x) unique(as.character(read.delim(annofiles[x])[, 
    &quot;geneId&quot;])), simplify = FALSE)
load(&quot;data/GO/catdb.RData&quot;)
BatchResult &lt;- GOCluster_Report(catdb = catdb, setlist = gene_ids, 
    method = &quot;all&quot;, id_type = &quot;gene&quot;, CLSZ = 2, cutoff = 0.9, 
    gocats = c(&quot;MF&quot;, &quot;BP&quot;, &quot;CC&quot;), recordSpecGO = NULL)
</code></pre>
<h2 id="motif-analysis">Motif analysis</h2>
<h3 id="parse-dna-sequences-of-peak-regions-from-genome">Parse DNA sequences of peak regions from genome</h3>
<p>Enrichment analysis of known DNA binding motifs or <em>de novo</em> discovery
of novel motifs requires the DNA sequences of the identified peak
regions. To parse the corresponding sequences from the reference genome,
the <code>getSeq</code> function from the <code>Biostrings</code> package can be used. The
following example parses the sequences for each peak set and saves the
results to separate FASTA files, one for each peak set. In addition, the
sequences in the FASTA files are ranked (sorted) by increasing p-values
as expected by some motif discovery tools, such as <code>BCRANK</code>.</p>
<pre><code class="language-r">library(Biostrings)
library(seqLogo)
library(BCRANK)
dir_path &lt;- system.file(&quot;extdata/cwl/annotate_peaks&quot;, package = &quot;systemPipeR&quot;)
args &lt;- loadWF(targets = &quot;targets_macs.txt&quot;, wf_file = &quot;annotate-peaks.cwl&quot;, 
    input_file = &quot;annotate-peaks.yml&quot;, dir_path = dir_path)
args &lt;- renderWF(args, inputvars = c(FileName = &quot;_FASTQ_PATH1_&quot;, 
    SampleName = &quot;_SampleName_&quot;))

rangefiles &lt;- infile1(args)
for (i in seq(along = rangefiles)) {
    df &lt;- read.delim(rangefiles[i], comment = &quot;#&quot;)
    peaks &lt;- as(df, &quot;GRanges&quot;)
    names(peaks) &lt;- paste0(as.character(seqnames(peaks)), &quot;_&quot;, 
        start(peaks), &quot;-&quot;, end(peaks))
    peaks &lt;- peaks[order(values(peaks)$X.log10.pvalue., decreasing = TRUE)]
    pseq &lt;- getSeq(FaFile(&quot;./data/tair10.fasta&quot;), peaks)
    names(pseq) &lt;- names(peaks)
    writeXStringSet(pseq, paste0(rangefiles[i], &quot;.fasta&quot;))
}
</code></pre>
<h3 id="motif-discovery-with-bcrank">Motif discovery with <code>BCRANK</code></h3>
<p>The Bioconductor package <code>BCRANK</code> is one of the many tools available for
<em>de novo</em> discovery of DNA binding motifs in peak regions of ChIP-Seq
experiments. The given example applies this method on the first peak
sample set and plots the sequence logo of the highest ranking motif.</p>
<pre><code class="language-r">set.seed(0)
BCRANKout &lt;- bcrank(paste0(rangefiles[1], &quot;.fasta&quot;), restarts = 25, 
    use.P1 = TRUE, use.P2 = TRUE)
toptable(BCRANKout)
topMotif &lt;- toptable(BCRANKout, 1)
weightMatrix &lt;- pwm(topMotif, normalize = FALSE)
weightMatrixNormalized &lt;- pwm(topMotif, normalize = TRUE)
pdf(&quot;results/seqlogo.pdf&quot;)
seqLogo(weightMatrixNormalized)
dev.off()
</code></pre>
<p><img src="../results/seqlogo.png" alt=""></p>
<div align="center">
<p>Figure 2: One of the motifs identified by <code>BCRANK</code></p>
</div>
</br>
<h2 id="version-information">Version Information</h2>
<pre><code class="language-r">sessionInfo()
</code></pre>
<pre><code>## R version 4.0.5 (2021-03-31)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 10 (buster)
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.8.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.8.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices
## [6] utils     datasets  methods   base     
## 
## other attached packages:
##  [1] systemPipeR_1.24.5          ShortRead_1.48.0           
##  [3] GenomicAlignments_1.26.0    SummarizedExperiment_1.20.0
##  [5] Biobase_2.50.0              MatrixGenerics_1.2.0       
##  [7] matrixStats_0.57.0          BiocParallel_1.24.1        
##  [9] Rsamtools_2.6.0             Biostrings_2.58.0          
## [11] XVector_0.30.0              GenomicRanges_1.42.0       
## [13] GenomeInfoDb_1.26.1         IRanges_2.24.0             
## [15] S4Vectors_0.28.0            BiocGenerics_0.36.0        
## [17] BiocStyle_2.18.0           
## 
## loaded via a namespace (and not attached):
##   [1] colorspace_2.0-0         rjson_0.2.20            
##   [3] hwriter_1.3.2            ellipsis_0.3.1          
##   [5] bit64_4.0.5              AnnotationDbi_1.52.0    
##   [7] xml2_1.3.2               codetools_0.2-18        
##   [9] splines_4.0.5            knitr_1.30              
##  [11] jsonlite_1.7.1           annotate_1.68.0         
##  [13] GO.db_3.12.1             dbplyr_2.0.0            
##  [15] png_0.1-7                pheatmap_1.0.12         
##  [17] graph_1.68.0             BiocManager_1.30.10     
##  [19] compiler_4.0.5           httr_1.4.2              
##  [21] backports_1.2.0          GOstats_2.56.0          
##  [23] assertthat_0.2.1         Matrix_1.3-2            
##  [25] limma_3.46.0             formatR_1.7             
##  [27] htmltools_0.5.1.1        prettyunits_1.1.1       
##  [29] tools_4.0.5              gtable_0.3.0            
##  [31] glue_1.4.2               GenomeInfoDbData_1.2.4  
##  [33] Category_2.56.0          dplyr_1.0.2             
##  [35] rsvg_2.1                 batchtools_0.9.14       
##  [37] rappdirs_0.3.1           V8_3.4.0                
##  [39] Rcpp_1.0.5               jquerylib_0.1.3         
##  [41] vctrs_0.3.5              blogdown_1.2            
##  [43] rtracklayer_1.50.0       xfun_0.22               
##  [45] stringr_1.4.0            lifecycle_0.2.0         
##  [47] XML_3.99-0.5             edgeR_3.32.0            
##  [49] zlibbioc_1.36.0          scales_1.1.1            
##  [51] BSgenome_1.58.0          VariantAnnotation_1.36.0
##  [53] hms_0.5.3                RBGL_1.66.0             
##  [55] RColorBrewer_1.1-2       yaml_2.2.1              
##  [57] curl_4.3                 memoise_1.1.0           
##  [59] ggplot2_3.3.2            sass_0.3.1              
##  [61] biomaRt_2.46.0           latticeExtra_0.6-29     
##  [63] stringi_1.5.3            RSQLite_2.2.1           
##  [65] genefilter_1.72.0        checkmate_2.0.0         
##  [67] GenomicFeatures_1.42.1   DOT_0.1                 
##  [69] rlang_0.4.8              pkgconfig_2.0.3         
##  [71] bitops_1.0-6             evaluate_0.14           
##  [73] lattice_0.20-41          purrr_0.3.4             
##  [75] bit_4.0.4                tidyselect_1.1.0        
##  [77] GSEABase_1.52.0          AnnotationForge_1.32.0  
##  [79] magrittr_2.0.1           bookdown_0.21           
##  [81] R6_2.5.0                 generics_0.1.0          
##  [83] base64url_1.4            DelayedArray_0.16.0     
##  [85] DBI_1.1.0                withr_2.3.0             
##  [87] pillar_1.4.7             survival_3.2-10         
##  [89] RCurl_1.98-1.2           tibble_3.0.4            
##  [91] crayon_1.3.4             BiocFileCache_1.14.0    
##  [93] rmarkdown_2.7            jpeg_0.1-8.1            
##  [95] progress_1.2.2           locfit_1.5-9.4          
##  [97] grid_4.0.5               data.table_1.13.2       
##  [99] blob_1.2.1               Rgraphviz_2.34.0        
## [101] digest_0.6.27            xtable_1.8-4            
## [103] brew_1.0-6               openssl_1.4.3           
## [105] munsell_0.5.0            bslib_0.2.4             
## [107] askpass_1.1
</code></pre>
<h2 id="funding">Funding</h2>
<p>This project was supported by funds from the National Institutes of
Health (NIH) and the National Science Foundation (NSF).</p>
<h2 id="references">References</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-H_Backman2016-bt" class="csl-entry">
<p>H Backman, Tyler W, and Thomas Girke. 2016. “<span class="nocase">systemPipeR: NGS workflow and report generation environment</span>.” <em>BMC Bioinformatics</em> 17 (1): 388. <a href="https://doi.org/10.1186/s12859-016-1241-0">https://doi.org/10.1186/s12859-016-1241-0</a>.</p>
</div>
<div id="ref-Langmead2012-bs" class="csl-entry">
<p>Langmead, Ben, and Steven L Salzberg. 2012. “Fast Gapped-Read Alignment with Bowtie 2.” <em>Nat. Methods</em> 9 (4): 357–59. <a href="https://doi.org/10.1038/nmeth.1923">https://doi.org/10.1038/nmeth.1923</a>.</p>
</div>
<div id="ref-Love2014-sh" class="csl-entry">
<p>Love, Michael, Wolfgang Huber, and Simon Anders. 2014. “Moderated Estimation of Fold Change and Dispersion for <span class="nocase">RNA-seq</span> Data with DESeq2.” <em>Genome Biol.</em> 15 (12): 550. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.</p>
</div>
<div id="ref-Robinson2010-uk" class="csl-entry">
<p>Robinson, M D, D J McCarthy, and G K Smyth. 2010. “edgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.” <em>Bioinformatics</em> 26 (1): 139–40. <a href="https://doi.org/10.1093/bioinformatics/btp616">https://doi.org/10.1093/bioinformatics/btp616</a>.</p>
</div>
<div id="ref-Yu2015-xu" class="csl-entry">
<p>Yu, Guangchuang, Li-Gen Wang, and Qing-Yu He. 2015. “ChIPseeker: An R/Bioconductor Package for ChIP Peak Annotation, Comparison and Visualization.” <em>Bioinformatics</em> 31 (14): 2382–83. <a href="https://doi.org/10.1093/bioinformatics/btv145">https://doi.org/10.1093/bioinformatics/btv145</a>.</p>
</div>
<div id="ref-Zhang2008-pc" class="csl-entry">
<p>Zhang, Y, T Liu, C A Meyer, J Eeckhoute, D S Johnson, B E Bernstein, C Nussbaum, et al. 2008. “Model-Based Analysis of ChIP-Seq (MACS).” <em>Genome Biol.</em> 9 (9). <a href="https://doi.org/10.1186/gb-2008-9-9-r137">https://doi.org/10.1186/gb-2008-9-9-r137</a>.</p>
</div>
<div id="ref-Zhu2010-zo" class="csl-entry">
<p>Zhu, Lihua J, Claude Gazin, Nathan D Lawson, Hervé Pagès, Simon M Lin, David S Lapointe, and Michael R Green. 2010. “ChIPpeakAnno: A Bioconductor Package to Annotate <span class="nocase">ChIP-seq</span> and <span class="nocase">ChIP-chip</span> Data.” <em>BMC Bioinformatics</em> 11: 237. <a href="https://doi.org/10.1186/1471-2105-11-237">https://doi.org/10.1186/1471-2105-11-237</a>.</p>
</div>
</div>

	
	
	<div class="text-muted mt-5 pt-3 border-top">Last modified 2021-05-01: <a  href="https://github.com/tgirke/GEN242//commit/33fa745d116679e34731b56225f58ff5e13abeca">some edits (33fa745d)</a>
</div>
</div>


              </main>
          


        </div>
      </div>
      
<footer class="bg-dark py-5 row d-print-none">
  <div class="container-fluid mx-sm-5">
    <div class="row">
      <div class="col-6 col-sm-4 text-xs-center order-sm-2">
        
      </div>
      <div class="col-6 col-sm-4 text-right text-xs-center order-sm-3">
        
      </div>
      <div class="col-12 col-sm-4 text-center py-2 order-sm-2">
        <small class="text-white">&copy; 2021 The GEN242 Instructors All Rights Reserved</small>
        <small class="ml-1"><a href="https://www.ucr.edu/privacy.html" target="_blank">Privacy Policy</a></small>
	
      </div>
    </div>
  </div>
</footer>


    </div>
    
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>


<script>
  (function() {
    var i, text, code, codes = document.getElementsByTagName('code');
    for (i = 0; i < codes.length;) {
      code = codes[i];
      if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
        text = code.textContent;
        if (/^\$[^$]/.test(text) && /[^$]\$$/.test(text)) {
          text = text.replace(/^\$/, '\\(').replace(/\$$/, '\\)');
          code.textContent = text;
        }
        if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
            /^\$(.|\s)+\$$/.test(text) ||
            /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
          code.outerHTML = code.innerHTML;  
          continue;
        }
      }
      i++;
    }
  })();
</script>

<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script>













<script src="../../../js/main.min.f820bfc507c3ebefe17f8881ec4ee10013f3b89ad9555a50acf9bb1771e6823f.js" integrity="sha256-&#43;CC/xQfD6&#43;/hf4iB7E7hABPzuJrZVVpQrPm7F3Hmgj8=" crossorigin="anonymous"></script>



<script src='../../../js/prism.js'></script>



  </body>
</html>
