<!doctype html>
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="generator" content="Hugo 0.80.0" /><META NAME="ROBOTS" CONTENT="INDEX, FOLLOW">


<link rel="shortcut icon" href="../../../favicons/favicon.ico" >
<link rel="apple-touch-icon" href="../../../favicons/apple-touch-icon-180x180.png" sizes="180x180">
<link rel="icon" type="image/png" href="../../../favicons/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="../../../favicons/favicon-32x32.png" sizes="32x32">
<link rel="icon" type="image/png" href="../../../favicons/android-36x36.png" sizes="36x36">
<link rel="icon" type="image/png" href="../../../favicons/android-48x48.png" sizes="48x48">
<link rel="icon" type="image/png" href="../../../favicons/android-72x72.png" sizes="72x72">
<link rel="icon" type="image/png" href="../../../favicons/android-96x96.png" sizes="96x96">
<link rel="icon" type="image/png" href="../../../favicons/android-144x144.png" sizes="144x144">
<link rel="icon" type="image/png" href="../../../favicons/android-192x192.png" sizes="192x192">

<title>Introduction to R | GEN242</title><meta property="og:title" content="Introduction to R" />
<meta property="og:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Overview What is R? R is a powerful statistical environment and programming language for the analysis and visualization of data. The associated Bioconductor and CRAN package repositories provide many additional R packages for statistical data analysis for a wide array of research areas." />
<meta property="og:type" content="article" />
<meta property="og:url" content="/manuals/rbasics/rbasics/" />
<meta property="article:modified_time" content="2021-05-15T21:10:10-07:00" /><meta property="og:site_name" content="GEN242" />
<meta itemprop="name" content="Introduction to R">
<meta itemprop="description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Overview What is R? R is a powerful statistical environment and programming language for the analysis and visualization of data. The associated Bioconductor and CRAN package repositories provide many additional R packages for statistical data analysis for a wide array of research areas.">
<meta itemprop="dateModified" content="2021-05-15T21:10:10-07:00" />
<meta itemprop="wordCount" content="10048">



<meta itemprop="keywords" content="" />
<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Introduction to R"/>
<meta name="twitter:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Overview What is R? R is a powerful statistical environment and programming language for the analysis and visualization of data. The associated Bioconductor and CRAN package repositories provide many additional R packages for statistical data analysis for a wide array of research areas."/>


<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-00000000-0', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>






<link rel="preload" href="../../../scss/main.min.1f416e6a5f646f231d085681bf2b4ebdad5d41f149a3977625affc97f735ab7e.css" as="style">
<link href="../../../scss/main.min.1f416e6a5f646f231d085681bf2b4ebdad5d41f149a3977625affc97f735ab7e.css" rel="stylesheet" integrity="">


<script
  src="https://code.jquery.com/jquery-3.5.1.min.js"
  integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0="
  crossorigin="anonymous"></script>

<script
  src="https://unpkg.com/lunr@2.3.8/lunr.min.js"
  integrity="sha384-vRQ9bDyE0Wnu+lMfm57BlYLO0/XauFuKpVsZPs7KEDwYKktWi5+Kz3MP8++DFlRY"
  crossorigin="anonymous"></script>



<link rel="stylesheet" href="../../../css/prism.css"/>


<script src="../../../js/driver0.9.8.min.js"></script>
<link rel="stylesheet" href="../../../css/driver0.9.8.min.css"/>


<link rel="stylesheet" href="../../../css/site.css"/>
<script src="../../../js/site.js"></script>



  </head>
  <body class="td-page line-numbers">
    <header>
     
<nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar">
  <a class="navbar-brand" href="../../../">
		<span class="font-weight-bold">GEN242</span>
	</a>
	<row class="header-switch">
	  	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="sidebarSwitch" checked=true>
        <label class="custom-control-label" for="sidebarSwitch">sidebar</label>
      </div>
    	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="tocSwitch" checked=true>
        <label class="custom-control-label" for="tocSwitch">toc</label>
      </div>
	</row>
	<div class="td-navbar-nav-scroll ml-md-auto" id="main_navbar">
		<ul class="navbar-nav mt-2 mt-lg-0">
			
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../about/" id="navbarDropdown" role="button">
          	About
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/introduction/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/syllabus/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Syllabus
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/schedule/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Schedule
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/internal/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Internal
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/internal_resources/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Resources
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/install/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Installation
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../slides/" id="navbarDropdown" role="button">
          	Slides
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_01"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Course Introduction (S1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_02"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Genome Basics (S2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_03"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Databases and Software (S3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_04"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 DNA Sequencing (S4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_05"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (S5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_06"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sequence Alignments (S6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_07"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (S7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_08"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Multiple Alignments (S8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_09"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Short Read Alignments (S9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_10"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R on HPC (S10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_11"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (S11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_12"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Gene Expression Analysis (S12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_13"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Workflows (S13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_14/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Overview (S14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_15/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Enrichment Analysis (S15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_16/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis (S16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_17/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (S17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_18/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics in R (S18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_19/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (S19)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_20/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (S20)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_21/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 More on Workflows (S21)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../tutorials/" id="navbarDropdown" role="button">
          	Tutorials
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/github/github/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Introduction (T1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/linux/linux/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux and HPCC Cluster (T2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rbasics/rbasics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (T3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rprogramming/rprogramming/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (T4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rparallel/rparallel/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Parallel R (T5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rsequences/rsequences/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (T6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/systempiper/systempiper/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflows (T7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/sprnaseq/sprnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 RNA-Seq Workflow (T8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/spchipseq/spchipseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Workflow (T9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rmarkdown/rmarkdown/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R Markdown (T10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rfea/rfea/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Functional Enrichment Analysis (T11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rclustering/rclustering/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis and Data Mining (T12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (T13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rgraphics/rgraphics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics and Visualization (T14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/shinyapps/shinyapps/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Shiny Apps (T15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rpackages/rpackages/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (T16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/dplyr/dplyr/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (T17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/cmdtocwl/cmdtocwl/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Automate CWL Creation (T18)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../manuals/" id="navbarDropdown" role="button">
          	Manuals
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/github/github/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/linux/linux/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux &amp; HPCC Cluster
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rbasics/rbasics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rbasics/sample/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sample R Markdown
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rprogramming/rprogramming/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rparallel/rparallel/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Parallel Evaluations in R
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rsequences/rsequences/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/systempiper/systempiper/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Workflow Overview
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/sprnaseq/sprnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 RNA-Seq Workflow
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/spchipseq/spchipseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Workflow
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rfea/rfea/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Functional Enrichment Analysis
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rclustering/rclustering/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis and Data Mining
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../manuals/rgraphics/rgraphics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics and Visualization
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../assignments/" id="navbarDropdown" role="button">
          	Assignments
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/homework/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Homework
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw01/hw01/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW1: Intro &amp; GitHub
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw02/hw02/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW2: Linux &amp; HPC
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw03/hw03/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW3: R Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw04/hw04/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW4: Alignments
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw05/hw05"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW5: R Programming
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw06/hw06"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW6: NGS Analysis Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw07/hw07"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW7: RNA-Seq Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw08/hw08"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW8: R Markdown
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_overview/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Overview of Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/01_rnaseq_aligners/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 RNA-Seq Aligners
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/02_rnaseq_deg/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEG Methods
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/03_cluster_analysis/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Cluster Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/04_scrnaseq_embedding/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 scRNA-Seq Embedding
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/05_chipseq_peakcaller/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 ChIP-Seq Peak Callers
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/06_functional_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Functional Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/07_motif_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Motif Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/08_genome_graphics/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Genome Graphics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Project Data
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Presentations
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/paper_presentations/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Paper Presentations
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../external_resources/" id="navbarDropdown" role="button">
          	Links
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="https://piazza.com/ucr/spring2021/gen242/home"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Piazza
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://classroom.github.com/classrooms/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Classroom
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://github.com/tgirke/GEN242"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Repo
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://bioconductor.org/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Bioconductor
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://www.dcassol.com/docsy_Rmarkdown/rmarkdown/install/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Hugo, Docsy &amp; R
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
				
				
					<a class="nav-link" href="../../../blog/" ><span>News</span></a>
				
			</li>
			
			
			
		</ul>
	</div>
	<div class="navbar-nav d-none d-lg-block">



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.207ec06ddfc2d3c0796bfb7ea03c9444.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>

</div>
</nav>

    </header>
    <div class="container-fluid td-outer">
      <div class="td-main">
        <div class="row flex-xl-nowrap">
          <div class="col-12 col-md-3 col-xl-2 td-sidebar d-print-none">
            




<div id="td-sidebar-menu" class="td-sidebar__inner">
  
  <form class="td-sidebar__search d-flex align-items-center">
    



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.207ec06ddfc2d3c0796bfb7ea03c9444.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>


    <button class="btn btn-link td-sidebar__toggle d-md-none p-0 ml-3 fas fa-bars" type="button" data-toggle="collapse" data-target="#td-section-nav" aria-controls="td-docs-nav" aria-expanded="false" aria-label="Toggle section navigation">
    </button>
  </form>
  
  <nav class="collapse td-sidebar-nav" id="td-section-nav">
    
    
    






<ul class="td-sidebar-nav__section pr-md-3">
  <li class="td-sidebar-nav__section-title">
    <a  href="../../../manuals/" class="align-left pl-0 pr-2 active td-sidebar-link td-sidebar-link__section">Manuals</a>
  </li>
  <ul>
    <li class="collapse show" id="manuals">
      
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsgithubgithub" href="../../../manuals/github/github/">GitHub</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualslinuxlinux" href="../../../manuals/linux/linux/">Linux &amp; HPCC Cluster</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page  active" id="m-manualsrbasicsrbasics" href="../../../manuals/rbasics/rbasics/">Introduction to R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrprogrammingrprogramming" href="../../../manuals/rprogramming/rprogramming/">Programming in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrbasicssample" href="../../../manuals/rbasics/sample/">Sample R Markdown</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrparallelrparallel" href="../../../manuals/rparallel/rparallel/">Parallel Evaluations in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrsequencesrsequences" href="../../../manuals/rsequences/rsequences/">NGS Analysis Basics</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualssystempipersystempiper" href="../../../manuals/systempiper/systempiper/">systemPipeR: Workflow design and reporting generation environment</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualssprnaseqsprnaseq" href="../../../manuals/sprnaseq/sprnaseq/">RNA-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsspchipseqspchipseq" href="../../../manuals/spchipseq/spchipseq/">ChIP-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrfearfea" href="../../../manuals/rfea/rfea/">Functional Enrichment Analysis</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrclusteringrclustering" href="../../../manuals/rclustering/rclustering/">Cluster Analysis in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-manualsrgraphicsrgraphics" href="../../../manuals/rgraphics/rgraphics/">Graphics and Data Visualization in R</a>
        
      
      
    </li>
  </ul>
</ul>

  </nav>
</div>




          </div>
          
          

          
            
          

          
          
              <div class="d-none d-xl-block col-xl-2 td-toc d-print-none">
                
  
  
  
  
  
  <div class="td-page-meta ml-2 pb-1 pt-2 mb-0">
  
    
    
      
    
    
    
    
    
    
    
    


    
    
      <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/manuals/rbasics/Rbasics.markdown" target="_blank" class="source-link"><i class="fa fa-code fa-fw"></i> View Rmd source</a>
    

    <a href="https://github.com/tgirke/GEN242//edit/main/content/en/manuals/rbasics/Rbasics.markdown" target="_blank"><i class="fa fa-edit fa-fw"></i> Edit this page</a>
    <a href="https://github.com/tgirke/GEN242//new/main/content/en/manuals/rbasics/Rbasics.markdown?filename=change-me.md&amp;value=---%0Atitle%3A&#43;%22Long&#43;Page&#43;Title%22%0AlinkTitle%3A&#43;%22Short&#43;Nav&#43;Title%22%0Aweight%3A&#43;100%0Adescription%3A&#43;%3E-%0A&#43;&#43;&#43;&#43;&#43;Page&#43;description&#43;for&#43;heading&#43;and&#43;indexes.%0A---%0A%0A%23%23&#43;Heading%0A%0AEdit&#43;this&#43;template&#43;to&#43;create&#43;your&#43;new&#43;page.%0A%0A%2A&#43;Give&#43;it&#43;a&#43;good&#43;name%2C&#43;ending&#43;in&#43;%60.md%60&#43;-&#43;e.g.&#43;%60getting-started.md%60%0A%2A&#43;Edit&#43;the&#43;%22front&#43;matter%22&#43;section&#43;at&#43;the&#43;top&#43;of&#43;the&#43;page&#43;%28weight&#43;controls&#43;how&#43;its&#43;ordered&#43;amongst&#43;other&#43;pages&#43;in&#43;the&#43;same&#43;directory%3B&#43;lowest&#43;number&#43;first%29.%0A%2A&#43;Add&#43;a&#43;good&#43;commit&#43;message&#43;at&#43;the&#43;bottom&#43;of&#43;the&#43;page&#43;%28%3C80&#43;characters%3B&#43;use&#43;the&#43;extended&#43;description&#43;field&#43;for&#43;more&#43;detail%29.%0A%2A&#43;Create&#43;a&#43;new&#43;branch&#43;so&#43;you&#43;can&#43;preview&#43;your&#43;new&#43;file&#43;and&#43;request&#43;a&#43;review&#43;via&#43;Pull&#43;Request.%0A" target="_blank"><i class="fa fa-edit fa-fw"></i> Create child page</a>
    <a href="https://github.com/tgirke/GEN242//issues/new?title=Introduction%20to%20R" target="_blank"><i class="fab fa-github fa-fw"></i> Create documentation issue</a>
    
      
      <a href="https://github.com/tgirke/GEN242/issues/new" target="_blank"><i class="fas fa-tasks fa-fw"></i> Create project issue</a>
    

  
  
  </div>





<nav id="TableOfContents">
  <ul>
    <li><a href="#overview">Overview</a></li>
    <li><a href="#what-is-r">What is R?</a></li>
    <li><a href="#why-using-r">Why Using R?</a></li>
    <li><a href="#books-and-documentation">Books and Documentation</a></li>
    <li><a href="#r-working-environments">R Working Environments</a>
      <ul>
        <li><a href="#example-rstudio">Example: RStudio</a></li>
        <li><a href="#example-nvim-r-tmux">Example: Nvim-R-Tmux</a></li>
      </ul>
    </li>
    <li><a href="#r-package-repositories">R Package Repositories</a></li>
    <li><a href="#installation-of-r-packages">Installation of R Packages</a></li>
    <li><a href="#getting-around">Getting Around</a>
      <ul>
        <li><a href="#startup-and-closing-behavior">Startup and Closing Behavior</a></li>
      </ul>
    </li>
    <li><a href="#navigating-directories">Navigating directories</a></li>
    <li><a href="#basic-syntax">Basic Syntax</a></li>
    <li><a href="#data-types">Data Types</a>
      <ul>
        <li><a href="#numeric-data">Numeric data</a></li>
        <li><a href="#character-data">Character data</a></li>
        <li><a href="#complex-data">Complex data</a></li>
        <li><a href="#logical-data">Logical data</a></li>
      </ul>
    </li>
    <li><a href="#data-objects">Data Objects</a>
      <ul>
        <li><a href="#object-types">Object types</a>
          <ul>
            <li><a href="#vectors-1d">Vectors (1D)</a></li>
            <li><a href="#factors-1d">Factors (1D)</a></li>
            <li><a href="#matrices-2d">Matrices (2D)</a></li>
            <li><a href="#data-frames-2d">Data Frames (2D)</a></li>
            <li><a href="#arrays">Arrays</a></li>
            <li><a href="#lists">Lists</a></li>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#functions">Functions</a></li>
    <li><a href="#subsetting-of-data-objects">Subsetting of data objects</a></li>
    <li><a href="#important-utilities">Important Utilities</a>
      <ul>
        <li><a href="#combining-objects">Combining Objects</a></li>
        <li><a href="#accessing-dimensions-of-objects">Accessing Dimensions of Objects</a></li>
        <li><a href="#accessing-name-slots-of-objects">Accessing Name Slots of Objects</a></li>
        <li><a href="#sorting-objects">Sorting Objects</a></li>
      </ul>
    </li>
    <li><a href="#operators-and-calculations">Operators and Calculations</a>
      <ul>
        <li><a href="#comparison-operators">Comparison Operators</a></li>
        <li><a href="#basic-calculations">Basic Calculations</a></li>
      </ul>
    </li>
    <li><a href="#reading-and-writing-external-data">Reading and Writing External Data</a>
      <ul>
        <li><a href="#import-of-tabular-data">Import of tabular data</a></li>
        <li><a href="#export-of-tabular-data">Export of tabular data</a></li>
        <li><a href="#line-wise-import">Line-wise import</a></li>
        <li><a href="#line-wise-export">Line-wise export</a></li>
        <li><a href="#export-r-object">Export R object</a></li>
        <li><a href="#import-r-object">Import R object</a></li>
        <li><a href="#copy-and-paste-into-r">Copy and paste into R</a></li>
        <li><a href="#copy-and-paste-from-r">Copy and paste from R</a></li>
        <li><a href="#homework-3a">Homework 3A</a></li>
      </ul>
    </li>
    <li><a href="#useful-r-functions">Useful R Functions</a>
      <ul>
        <li><a href="#unique-entries">Unique entries</a></li>
        <li><a href="#count-occurrences">Count occurrences</a></li>
        <li><a href="#aggregate-data">Aggregate data</a></li>
        <li><a href="#intersect-data">Intersect data</a></li>
        <li><a href="#merge-data-frames">Merge data frames</a></li>
      </ul>
    </li>
    <li><a href="#dplyr-environment">dplyr Environment</a>
      <ul>
        <li><a href="#installation">Installation</a></li>
        <li><a href="#construct-a-tibble-tibble">Construct a <code>tibble</code> (<code>tibble</code>)</a></li>
        <li><a href="#reading-and-writing-tabular-files">Reading and writing tabular files</a>
          <ul>
            <li><a href="#import-with-readr">Import with <code>readr</code></a></li>
            <li><a href="#fast-table-import-with-fread">Fast table import with <code>fread</code></a></li>
            <li><a href="#export-with-readr">Export with <code>readr</code></a></li>
          </ul>
        </li>
        <li><a href="#column-and-row-binds">Column and row binds</a></li>
        <li><a href="#extract-column-as-vector">Extract column as vector</a></li>
        <li><a href="#important-dplyr-functions">Important <code>dplyr</code> functions</a></li>
        <li><a href="#slice-and-filter-functions">Slice and filter functions</a>
          <ul>
            <li><a href="#filter-function">Filter function</a></li>
            <li><a href="#base-r-code-equivalent">Base R code equivalent</a></li>
            <li><a href="#including-boolean-operators">Including boolean operators</a></li>
            <li><a href="#subset-rows-by-position">Subset rows by position</a></li>
            <li><a href="#subset-rows-by-names">Subset rows by names</a></li>
          </ul>
        </li>
        <li><a href="#sorting-with-arrange">Sorting with <code>arrange</code></a></li>
        <li><a href="#select-columns-with-select">Select columns with <code>select</code></a></li>
        <li><a href="#renaming-columns-with-rename">Renaming columns with <code>rename</code></a></li>
        <li><a href="#obtain-unique-rows-with-distinct">Obtain unique rows with <code>distinct</code></a></li>
        <li><a href="#add-columns">Add columns</a>
          <ul>
            <li><a href="#mutate"><code>mutate</code></a></li>
            <li><a href="#transmute"><code>transmute</code></a></li>
            <li><a href="#bind_cols"><code>bind_cols</code></a></li>
          </ul>
        </li>
        <li><a href="#summarize-data">Summarize data</a></li>
        <li><a href="#merging-tibbles">Merging tibbles</a>
          <ul>
            <li><a href="#inner-join">Inner join</a></li>
            <li><a href="#left-join">Left join</a></li>
            <li><a href="#right-join">Right join</a></li>
            <li><a href="#full-join">Full join</a></li>
            <li><a href="#anti-join">Anti join</a></li>
          </ul>
        </li>
        <li><a href="#chaining">Chaining</a>
          <ul>
            <li><a href="#example-1">Example 1</a></li>
            <li><a href="#example-2">Example 2</a></li>
            <li><a href="#example-3">Example 3</a></li>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#sqlite-databases">SQLite Databases</a></li>
    <li><a href="#loading-data-into-sqlite-databases">Loading data into SQLite databases</a>
      <ul>
        <li><a href="#list-names-of-tables-in-database">List names of tables in database</a></li>
        <li><a href="#import-table-into-dataframe">Import table into <code>data.frame</code></a></li>
        <li><a href="#query-database">Query database</a></li>
        <li><a href="#join-tables">Join tables</a></li>
      </ul>
    </li>
    <li><a href="#graphics-in-r">Graphics in R</a>
      <ul>
        <li><a href="#advantages">Advantages</a></li>
        <li><a href="#documentation-for-r-graphics">Documentation for R Graphics</a></li>
        <li><a href="#graphics-environments">Graphics Environments</a></li>
        <li><a href="#base-graphics-overview">Base Graphics: Overview</a>
          <ul>
            <li><a href="#preferred-object-types">Preferred Object Types</a></li>
          </ul>
        </li>
        <li><a href="#scatter-plots">Scatter Plots</a></li>
        <li><a href="#basic-scatter-plot">Basic Scatter Plot</a>
          <ul>
            <li><a href="#all-pairs">All pairs</a></li>
            <li><a href="#with-labels">With labels</a></li>
          </ul>
        </li>
        <li><a href="#more-examples">More examples</a>
          <ul>
            <li><a href="#add-regression-line">Add regression line</a></li>
            <li><a href="#log-scale">Log scale</a></li>
            <li><a href="#add-a-mathematical-expression">Add a mathematical expression</a></li>
          </ul>
        </li>
        <li><a href="#homework-3b">Homework 3B</a></li>
        <li><a href="#line-plots">Line Plots</a>
          <ul>
            <li><a href="#single-data-set">Single data set</a></li>
            <li><a href="#many-data-sets">Many Data Sets</a></li>
          </ul>
        </li>
        <li><a href="#bar-plots">Bar Plots</a>
          <ul>
            <li><a href="#basics">Basics</a></li>
            <li><a href="#error-bars">Error Bars</a></li>
          </ul>
        </li>
        <li><a href="#histograms">Histograms</a></li>
        <li><a href="#density-plots">Density Plots</a></li>
        <li><a href="#pie-charts">Pie Charts</a></li>
        <li><a href="#color-selection-utilities">Color Selection Utilities</a></li>
        <li><a href="#saving-graphics-to-file">Saving Graphics to File</a></li>
        <li><a href="#homework-3c">Homework 3C</a></li>
      </ul>
    </li>
    <li><a href="#analysis-routine">Analysis Routine</a>
      <ul>
        <li><a href="#overview-1">Overview</a></li>
        <li><a href="#analysis-routine-data-import">Analysis Routine: Data Import</a></li>
        <li><a href="#merging-data-frames">Merging Data Frames</a></li>
        <li><a href="#filtering-data">Filtering Data</a></li>
        <li><a href="#string-substitutions">String Substitutions</a></li>
        <li><a href="#calculations-on-data-frames">Calculations on Data Frames</a></li>
        <li><a href="#plotting-example">Plotting Example</a></li>
        <li><a href="#export-results-and-run-entire-exercise-as-script">Export Results and Run Entire Exercise as Script</a></li>
      </ul>
    </li>
    <li><a href="#r-markdown">R Markdown</a>
      <ul>
        <li><a href="#overview-2">Overview</a></li>
        <li><a href="#quick-start">Quick Start</a>
          <ul>
            <li><a href="#install-r-markdown">Install R Markdown</a></li>
            <li><a href="#initialize-a-new-r-markdown-rmd-script">Initialize a new R Markdown (<code>Rmd</code>) script</a></li>
            <li><a href="#metadata-section">Metadata section</a></li>
            <li><a href="#render-rmd-script">Render <code>Rmd</code> script</a></li>
            <li><a href="#r-code-chunks">R code chunks</a></li>
            <li><a href="#learning-markdown">Learning Markdown</a></li>
            <li><a href="#tables">Tables</a></li>
            <li><a href="#figures">Figures</a></li>
            <li><a href="#inline-r-code">Inline R code</a></li>
            <li><a href="#mathematical-equations">Mathematical equations</a></li>
            <li><a href="#citations-and-bibliographies">Citations and bibliographies</a></li>
            <li><a href="#viewing-r-markdown-report-on-hpcc-cluster">Viewing R Markdown report on HPCC cluster</a></li>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#shiny-web-apps">Shiny Web Apps</a>
      <ul>
        <li><a href="#what-is-shiny">What is Shiny?</a></li>
        <li><a href="#develop-and-test-shiny-app-locally">Develop and test Shiny app locally</a></li>
        <li><a href="#deploy-on-web-server">Deploy on web server</a></li>
        <li><a href="#example-shiny-app">Example Shiny app</a></li>
        <li><a href="#learning-shiny">Learning Shiny</a></li>
      </ul>
    </li>
    <li><a href="#session-info">Session Info</a></li>
    <li><a href="#references">References</a></li>
  </ul>
</nav>



              </div>
              <main class="col-12 col-md-9 col-xl-8 pl-md-5" role="main">
                
  

                <nav aria-label="breadcrumb" class="d-none d-md-block d-print-none">
	<ol class="breadcrumb spb-1">
		







<li class="breadcrumb-item" >
	<a href="../../../manuals/">Manuals</a>
</li>




<li class="breadcrumb-item active" aria-current="page">
	<a href="../../../manuals/rbasics/rbasics/">Introduction to R</a>
</li>

	</ol>
</nav	>

                
<div class="td-content">
	<h1>Introduction to R</h1>
    
	
        <p class="reading-time"><i class="fa fa-clock" aria-hidden="true"></i> 48 minute read</p>
           
	<script src="../../../rmarkdown-libs/htmlwidgets/htmlwidgets.js"></script>
<script src="../../../rmarkdown-libs/jquery/jquery.min.js"></script>
<link href="../../../rmarkdown-libs/datatables-css/datatables-crosstalk.css" rel="stylesheet" />
<script src="../../../rmarkdown-libs/datatables-binding/datatables.js"></script>
<link href="../../../rmarkdown-libs/dt-core/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="../../../rmarkdown-libs/dt-core/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="../../../rmarkdown-libs/dt-core/js/jquery.dataTables.min.js"></script>
<link href="../../../rmarkdown-libs/nouislider/jquery.nouislider.min.css" rel="stylesheet" />
<script src="../../../rmarkdown-libs/nouislider/jquery.nouislider.min.js"></script>
<link href="../../../rmarkdown-libs/selectize/selectize.bootstrap3.css" rel="stylesheet" />
<script src="../../../rmarkdown-libs/selectize/selectize.min.js"></script>
<link href="../../../rmarkdown-libs/crosstalk/css/crosstalk.css" rel="stylesheet" />
<script src="../../../rmarkdown-libs/crosstalk/js/crosstalk.min.js"></script>
<!---
- Compile from command-line
Rscript -e "rmarkdown::render('Rbasics.Rmd', c('html_document'), clean=F); knitr::knit('Rbasics.Rmd', tangle=TRUE)"; Rscript ../md2jekyll.R Rbasics.knit.md 8; Rscript -e "rmarkdown::render('Rbasics.Rmd', c('pdf_document'))"
-->
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>
<h2 id="overview">Overview</h2>
<h2 id="what-is-r">What is R?</h2>
<p><a href="http://cran.at.r-project.org">R</a> is a powerful statistical environment and
programming language for the analysis and visualization of data. The
associated <a href="http://bioconductor.org/">Bioconductor</a> and CRAN package
repositories provide many additional R packages for statistical data analysis
for a wide array of research areas. The R software is free and runs on all
common operating systems.</p>
<h2 id="why-using-r">Why Using R?</h2>
<ul>
<li>Complete statistical environment and programming language</li>
<li>Efficient functions and data structures for data analysis</li>
<li>Powerful graphics</li>
<li>Access to fast growing number of analysis packages</li>
<li>Most widely used language in bioinformatics</li>
<li>Is standard for data mining and biostatistical analysis</li>
<li>Technical advantages: free, open-source, available for all OSs</li>
</ul>
<h2 id="books-and-documentation">Books and Documentation</h2>
<ul>
<li>simpleR - Using R for Introductory Statistics (John Verzani, 2004) - <a href="http://cran.r-project.org/doc/contrib/Verzani-SimpleR.pdf">URL</a></li>
<li>Bioinformatics and Computational Biology Solutions Using R and Bioconductor (Gentleman et al., 2005) - <a href="http://www.bioconductor.org/help/publications/books/bioinformatics-and-computational-biology-solutions/">URL</a></li>
<li>More on this see “Finding Help” section in UCR Manual - <a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual#TOC-Finding-Help">URL</a></li>
</ul>
<h2 id="r-working-environments">R Working Environments</h2>
<center>
<img title="R_Interfaces" src="../images/rinterface.png"/>
</center>
<center>
R Projects and Interfaces
</center>
<p>Some R working environments with support for syntax highlighting and utilities to send code
to the R console:</p>
<ul>
<li><a href="https://www.rstudio.com/products/rstudio/features">RStudio</a>: excellent choice for beginners (<a href="http://www.rstudio.com/wp-content/uploads/2016/01/rstudio-IDE-cheatsheet.pdf">Cheat Sheet</a>)</li>
<li>Basic R code editors provided by Rguis</li>
<li><a href="https://wiki.gnome.org/Apps/Gedit">gedit</a>, <a href="http://rgedit.sourceforge.net/">Rgedit</a>, <a href="https://rkward.kde.org/">RKWard</a>, <a href="http://www.walware.de/goto/statet">Eclipse</a>, <a href="http://www.sciviews.org/Tinn-R/">Tinn-R</a>, <a href="https://notepad-plus-plus.org/">Notepad++</a>, <a href="http://sourceforge.net/projects/npptor/">NppToR</a></li>
<li><a href="http://manuals.bioinformatics.ucr.edu/home/programming-in-r/vim-r">Vim-R-Tmux</a>: R working environment based on vim and tmux</li>
<li><a href="http://www.xemacs.org/Download/index.html">Emacs</a> (<a href="http://ess.r-project.org/">ESS add-on package</a>)</li>
</ul>
<h3 id="example-rstudio">Example: RStudio</h3>
<p>New integrated development environment (IDE) for <a href="http://www.rstudio.com/ide/download/">R</a>. Highly functional for both beginners and
advanced.</p>
<center>
<img title="RStudio" src="../images/rstudio.png"/>
</center>
<center>
RStudio IDE
</center>
<p>Some userful shortcuts: <code>Ctrl+Enter</code> (send code), <code>Ctrl+Shift+C</code> (comment/uncomment), <code>Ctrl+1/2</code> (switch window focus)</p>
<h3 id="example-nvim-r-tmux">Example: Nvim-R-Tmux</h3>
<p>Terminal-based Working Environment for R: <a href="http://girke.bioinformatics.ucr.edu/GEN242/mydoc_tutorial_02.html#nvim-r-tmux-essentials">Nvim-R-Tmux</a>.</p>
<center>
<img title="Nvim-R-Tmux" src="../images/Nvim-R.gif" >
</center>
<center>
Nvim-R-Tmux IDE for R
</center>
<h2 id="r-package-repositories">R Package Repositories</h2>
<ul>
<li>CRAN (&gt;11,000 packages) general data analysis - <a href="http://cran.at.r-project.org/">URL</a></li>
<li>Bioconductor (&gt;1,100 packages) bioscience data analysis - <a href="http://www.bioconductor.org/">URL</a></li>
<li>Omegahat (&gt;90 packages) programming interfaces - <a href="https://github.com/omegahat?tab=repositories">URL</a></li>
</ul>
<h2 id="installation-of-r-packages">Installation of R Packages</h2>
<ol>
<li>
<p>Install R for your operating system from <a href="http://cran.at.r-project.org/">CRAN</a>.</p>
</li>
<li>
<p>Install RStudio from <a href="http://www.rstudio.com/ide/download">RStudio</a>.</p>
</li>
<li>
<p>Install CRAN Packages from R console like this:</p>
<pre><code class="language-r">install.packages(c(&quot;pkg1&quot;, &quot;pkg2&quot;)) 
install.packages(&quot;pkg.zip&quot;, repos=NULL)
</code></pre>
</li>
<li>
<p>Install Bioconductor packages as follows:</p>
<pre><code class="language-r">if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;) # Installs BiocManager if not available yet
BiocManager::version() # Reports Bioconductor version
BiocManager::install(c(&quot;pkg1&quot;, &quot;pkg2&quot;)) # Installs packages specified under &quot;pkg1&quot; 
</code></pre>
</li>
<li>
<p>For more details consult the <a href="http://www.bioconductor.org/install/">Bioc Install page</a>
and <a href="http://www.bioconductor.org/packages/release/bioc/html/BiocInstaller.html">BiocInstaller</a> package.</p>
</li>
</ol>
<h2 id="getting-around">Getting Around</h2>
<h3 id="startup-and-closing-behavior">Startup and Closing Behavior</h3>
<ul>
<li>
<p><strong>Starting R</strong>:
The R GUI versions, including RStudio, under Windows and Mac OS X can be
opened by double-clicking their icons. Alternatively, one can start it by
typing <code>R</code> in a terminal (default under Linux).</p>
</li>
<li>
<p><strong>Startup/Closing Behavior</strong>:
The R environment is controlled by hidden files in the startup directory:
<code>.RData</code>, <code>.Rhistory</code> and <code>.Rprofile</code> (optional).</p>
</li>
<li>
<p><strong>Closing R</strong>:</p>
</li>
</ul>
<pre><code class="language-r">q()  
</code></pre>
<pre><code>Save workspace image? [y/n/c]:
</code></pre>
<ul>
<li><strong>Note</strong>:
When responding with <code>y</code>, then the entire R workspace will be written to
the <code>.RData</code> file which can become very large. Often it is sufficient to just
save an analysis protocol in an R source file. This way one can quickly
regenerate all data sets and objects.</li>
</ul>
<h2 id="navigating-directories">Navigating directories</h2>
<p>Create an object with the assignment operator <code>&lt;-</code> or <code>=</code></p>
<pre><code class="language-r">object &lt;- ...
</code></pre>
<p>Instead of the assignment operator one can use the <code>assign</code> function</p>
<pre><code class="language-r">assign(&quot;x&quot;, function(arguments))
</code></pre>
<p>List objects in current R session</p>
<pre><code class="language-r">ls()
</code></pre>
<p>Return content of current working directory</p>
<pre><code class="language-r">dir()
</code></pre>
<p>Return path of current working directory</p>
<pre><code class="language-r">getwd()
</code></pre>
<p>Change current working directory</p>
<pre><code class="language-r">setwd(&quot;/home/user&quot;)
</code></pre>
<h2 id="basic-syntax">Basic Syntax</h2>
<p>General R command syntax</p>
<pre><code class="language-r">object &lt;- function_name(arguments) 
object &lt;- object[arguments] 
</code></pre>
<p>Finding help</p>
<pre><code class="language-r">?function_name
</code></pre>
<p>Load a library/package</p>
<pre><code class="language-r">library(&quot;my_library&quot;) 
</code></pre>
<p>List functions defined by a library</p>
<pre><code class="language-r">library(help=&quot;my_library&quot;)
</code></pre>
<p>Load library manual (PDF or HTML file)</p>
<pre><code class="language-r">vignette(&quot;my_library&quot;) 
</code></pre>
<p>Execute an R script from within R</p>
<pre><code class="language-r">source(&quot;my_script.R&quot;)
</code></pre>
<p>Execute an R script from command-line (the first of the three options is preferred)</p>
<pre><code class="language-sh">$ Rscript my_script.R
$ R CMD BATCH my_script.R 
$ R --slave &lt; my_script.R 
</code></pre>
<h2 id="data-types">Data Types</h2>
<h3 id="numeric-data">Numeric data</h3>
<p>Example: <code>1, 2, 3, ...</code></p>
<pre><code class="language-r">x &lt;- c(1, 2, 3)
x
</code></pre>
<pre><code>## [1] 1 2 3
</code></pre>
<pre><code class="language-r">is.numeric(x)
</code></pre>
<pre><code>## [1] TRUE
</code></pre>
<pre><code class="language-r">as.character(x)
</code></pre>
<pre><code>## [1] &quot;1&quot; &quot;2&quot; &quot;3&quot;
</code></pre>
<h3 id="character-data">Character data</h3>
<p>Example: <code>&quot;a&quot;, &quot;b&quot;, &quot;c&quot;, ...</code></p>
<pre><code class="language-r">x &lt;- c(&quot;1&quot;, &quot;2&quot;, &quot;3&quot;)
x
</code></pre>
<pre><code>## [1] &quot;1&quot; &quot;2&quot; &quot;3&quot;
</code></pre>
<pre><code class="language-r">is.character(x)
</code></pre>
<pre><code>## [1] TRUE
</code></pre>
<pre><code class="language-r">as.numeric(x)
</code></pre>
<pre><code>## [1] 1 2 3
</code></pre>
<h3 id="complex-data">Complex data</h3>
<p>Example: mix of both</p>
<pre><code class="language-r">c(1, &quot;b&quot;, 3)
</code></pre>
<pre><code>## [1] &quot;1&quot; &quot;b&quot; &quot;3&quot;
</code></pre>
<h3 id="logical-data">Logical data</h3>
<p>Example: <code>TRUE</code> of <code>FALSE</code></p>
<pre><code class="language-r">x &lt;- 1:10 &lt; 5
x  
</code></pre>
<pre><code>##  [1]  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE
</code></pre>
<pre><code class="language-r">!x
</code></pre>
<pre><code>##  [1] FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
</code></pre>
<pre><code class="language-r">which(x) # Returns index for the 'TRUE' values in logical vector
</code></pre>
<pre><code>## [1] 1 2 3 4
</code></pre>
<h2 id="data-objects">Data Objects</h2>
<h3 id="object-types">Object types</h3>
<h4 id="vectors-1d">Vectors (1D)</h4>
<p>Definition: <code>numeric</code> or <code>character</code></p>
<pre><code class="language-r">myVec &lt;- 1:10; names(myVec) &lt;- letters[1:10]  
myVec &lt;- setNames(1:10, letters[1:10]) # Same as above in single step
myVec[1:5]
</code></pre>
<pre><code>## a b c d e 
## 1 2 3 4 5
</code></pre>
<pre><code class="language-r">myVec[c(2,4,6,8)]
</code></pre>
<pre><code>## b d f h 
## 2 4 6 8
</code></pre>
<pre><code class="language-r">myVec[c(&quot;b&quot;, &quot;d&quot;, &quot;f&quot;)]
</code></pre>
<pre><code>## b d f 
## 2 4 6
</code></pre>
<h4 id="factors-1d">Factors (1D)</h4>
<p>Definition: vectors with grouping information</p>
<pre><code class="language-r">factor(c(&quot;dog&quot;, &quot;cat&quot;, &quot;mouse&quot;, &quot;dog&quot;, &quot;dog&quot;, &quot;cat&quot;))
</code></pre>
<pre><code>## [1] dog   cat   mouse dog   dog   cat  
## Levels: cat dog mouse
</code></pre>
<h4 id="matrices-2d">Matrices (2D)</h4>
<p>Definition: two dimensional structures with data of same type</p>
<pre><code class="language-r">myMA &lt;- matrix(1:30, 3, 10, byrow = TRUE) 
class(myMA)
</code></pre>
<pre><code>## [1] &quot;matrix&quot; &quot;array&quot;
</code></pre>
<pre><code class="language-r">myMA[1:2,]
</code></pre>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## [1,]    1    2    3    4    5    6    7    8    9    10
## [2,]   11   12   13   14   15   16   17   18   19    20
</code></pre>
<pre><code class="language-r">myMA[1, , drop=FALSE]
</code></pre>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## [1,]    1    2    3    4    5    6    7    8    9    10
</code></pre>
<h4 id="data-frames-2d">Data Frames (2D)</h4>
<p>Definition: two dimensional objects with data of variable types</p>
<pre><code class="language-r">myDF &lt;- data.frame(Col1=1:10, Col2=10:1) 
myDF[1:2, ]
</code></pre>
<pre><code>##   Col1 Col2
## 1    1   10
## 2    2    9
</code></pre>
<h4 id="arrays">Arrays</h4>
<p>Definition: data structure with one, two or more dimensions</p>
<h4 id="lists">Lists</h4>
<p>Definition: containers for any object type</p>
<pre><code class="language-r">myL &lt;- list(name=&quot;Fred&quot;, wife=&quot;Mary&quot;, no.children=3, child.ages=c(4,7,9)) 
myL
</code></pre>
<pre><code>## $name
## [1] &quot;Fred&quot;
## 
## $wife
## [1] &quot;Mary&quot;
## 
## $no.children
## [1] 3
## 
## $child.ages
## [1] 4 7 9
</code></pre>
<pre><code class="language-r">myL[[4]][1:2] 
</code></pre>
<pre><code>## [1] 4 7
</code></pre>
<h2 id="functions">Functions</h2>
<p>Definition: piece of code</p>
<pre><code class="language-r">myfct &lt;- function(arg1, arg2, ...) { 
    function_body 
}
</code></pre>
<h2 id="subsetting-of-data-objects">Subsetting of data objects</h2>
<p><strong>(1.) Subsetting by positive or negative index/position numbers</strong></p>
<pre><code class="language-r">myVec &lt;- 1:26; names(myVec) &lt;- LETTERS 
myVec[1:4]
</code></pre>
<pre><code>## A B C D 
## 1 2 3 4
</code></pre>
<p><strong>(2.) Subsetting by same length logical vectors</strong></p>
<pre><code class="language-r">myLog &lt;- myVec &gt; 10
myVec[myLog] 
</code></pre>
<pre><code>##  K  L  M  N  O  P  Q  R  S  T  U  V  W  X  Y  Z 
## 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26
</code></pre>
<p><strong>(3.) Subsetting by field names</strong></p>
<pre><code class="language-r">myVec[c(&quot;B&quot;, &quot;K&quot;, &quot;M&quot;)]
</code></pre>
<pre><code>##  B  K  M 
##  2 11 13
</code></pre>
<p><strong>(4.) Subset with <code>$</code> sign</strong>: references a single column or list component by its name</p>
<pre><code class="language-r">iris$Species[1:8]
</code></pre>
<pre><code>## [1] setosa setosa setosa setosa setosa setosa setosa setosa
## Levels: setosa versicolor virginica
</code></pre>
<h2 id="important-utilities">Important Utilities</h2>
<h3 id="combining-objects">Combining Objects</h3>
<p>The <code>c</code> function combines vectors and lists</p>
<pre><code class="language-r">c(1, 2, 3)
</code></pre>
<pre><code>## [1] 1 2 3
</code></pre>
<pre><code class="language-r">x &lt;- 1:3; y &lt;- 101:103
c(x, y)
</code></pre>
<pre><code>## [1]   1   2   3 101 102 103
</code></pre>
<pre><code class="language-r">iris$Species[1:8]
</code></pre>
<pre><code>## [1] setosa setosa setosa setosa setosa setosa setosa setosa
## Levels: setosa versicolor virginica
</code></pre>
<p>The <code>cbind</code> and <code>rbind</code> functions can be used to append columns and rows, respecively.</p>
<pre><code class="language-r">ma &lt;- cbind(x, y)
ma
</code></pre>
<pre><code>##      x   y
## [1,] 1 101
## [2,] 2 102
## [3,] 3 103
</code></pre>
<pre><code class="language-r">rbind(ma, ma)
</code></pre>
<pre><code>##      x   y
## [1,] 1 101
## [2,] 2 102
## [3,] 3 103
## [4,] 1 101
## [5,] 2 102
## [6,] 3 103
</code></pre>
<h3 id="accessing-dimensions-of-objects">Accessing Dimensions of Objects</h3>
<p>Length and dimension information of objects</p>
<pre><code class="language-r">length(iris$Species)
</code></pre>
<pre><code>## [1] 150
</code></pre>
<pre><code class="language-r">dim(iris)
</code></pre>
<pre><code>## [1] 150   5
</code></pre>
<h3 id="accessing-name-slots-of-objects">Accessing Name Slots of Objects</h3>
<p>Accessing row and column names of 2D objects</p>
<pre><code class="language-r">rownames(iris)[1:8]
</code></pre>
<pre><code>## [1] &quot;1&quot; &quot;2&quot; &quot;3&quot; &quot;4&quot; &quot;5&quot; &quot;6&quot; &quot;7&quot; &quot;8&quot;
</code></pre>
<pre><code class="language-r">colnames(iris)
</code></pre>
<pre><code>## [1] &quot;Sepal.Length&quot; &quot;Sepal.Width&quot;  &quot;Petal.Length&quot; &quot;Petal.Width&quot;  &quot;Species&quot;
</code></pre>
<p>Return name field of vectors and lists</p>
<pre><code class="language-r">names(myVec)
</code></pre>
<pre><code>##  [1] &quot;A&quot; &quot;B&quot; &quot;C&quot; &quot;D&quot; &quot;E&quot; &quot;F&quot; &quot;G&quot; &quot;H&quot; &quot;I&quot; &quot;J&quot; &quot;K&quot; &quot;L&quot; &quot;M&quot; &quot;N&quot; &quot;O&quot; &quot;P&quot; &quot;Q&quot; &quot;R&quot; &quot;S&quot; &quot;T&quot; &quot;U&quot; &quot;V&quot; &quot;W&quot; &quot;X&quot;
## [25] &quot;Y&quot; &quot;Z&quot;
</code></pre>
<pre><code class="language-r">names(myL)
</code></pre>
<pre><code>## [1] &quot;name&quot;        &quot;wife&quot;        &quot;no.children&quot; &quot;child.ages&quot;
</code></pre>
<h3 id="sorting-objects">Sorting Objects</h3>
<p>The function <code>sort</code> returns a vector in ascending or descending order</p>
<pre><code class="language-r">sort(10:1)
</code></pre>
<pre><code>##  [1]  1  2  3  4  5  6  7  8  9 10
</code></pre>
<p>The function <code>order</code> returns a sorting index for sorting an object</p>
<pre><code class="language-r">sortindex &lt;- order(iris[,1], decreasing = FALSE)
sortindex[1:12]
</code></pre>
<pre><code>##  [1] 14  9 39 43 42  4  7 23 48  3 30 12
</code></pre>
<pre><code class="language-r">iris[sortindex,][1:2,]
</code></pre>
<pre><code>##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 14          4.3         3.0          1.1         0.1  setosa
## 9           4.4         2.9          1.4         0.2  setosa
</code></pre>
<pre><code class="language-r">sortindex &lt;- order(-iris[,1]) # Same as decreasing=TRUE
</code></pre>
<p>Sorting multiple columns</p>
<pre><code class="language-r">iris[order(iris$Sepal.Length, iris$Sepal.Width),][1:2,]
</code></pre>
<pre><code>##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 14          4.3         3.0          1.1         0.1  setosa
## 9           4.4         2.9          1.4         0.2  setosa
</code></pre>
<h2 id="operators-and-calculations">Operators and Calculations</h2>
<h3 id="comparison-operators">Comparison Operators</h3>
<p>Comparison operators: <code>==</code>, <code>!=</code>, <code>&lt;</code>, <code>&gt;</code>, <code>&lt;=</code>, <code>&gt;=</code></p>
<pre><code class="language-r">1==1
</code></pre>
<pre><code>## [1] TRUE
</code></pre>
<p>Logical operators: AND: <code>&amp;</code>, OR: <code>|</code>, NOT: <code>!</code></p>
<pre><code class="language-r">x &lt;- 1:10; y &lt;- 10:1
x &gt; y &amp; x &gt; 5
</code></pre>
<pre><code>##  [1] FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE
</code></pre>
<h3 id="basic-calculations">Basic Calculations</h3>
<p>To look up math functions, see Function Index <a href="http://cran.at.r-project.org/doc/manuals/R-intro.html#Function-and-variable-index">here</a></p>
<pre><code class="language-r">x + y
</code></pre>
<pre><code>##  [1] 11 11 11 11 11 11 11 11 11 11
</code></pre>
<pre><code class="language-r">sum(x)
</code></pre>
<pre><code>## [1] 55
</code></pre>
<pre><code class="language-r">mean(x)
</code></pre>
<pre><code>## [1] 5.5
</code></pre>
<pre><code class="language-r">apply(iris[1:6,1:3], 1, mean) 
</code></pre>
<pre><code>##        1        2        3        4        5        6 
## 3.333333 3.100000 3.066667 3.066667 3.333333 3.666667
</code></pre>
<h2 id="reading-and-writing-external-data">Reading and Writing External Data</h2>
<h3 id="import-of-tabular-data">Import of tabular data</h3>
<p>Import of a tab-delimited tabular file</p>
<pre><code class="language-r">myDF &lt;- read.delim(&quot;myData.xls&quot;, sep=&quot;\t&quot;)
</code></pre>
<p>Import of Excel file. Note: working with tab- or comma-delimited files is more flexible and preferred.</p>
<pre><code class="language-r">library(gdata)
myDF &lt;- read.xls&quot;myData.xls&quot;)
</code></pre>
<p>Import of Google Sheets. The following example imports a sample Google Sheet from <a href="https://docs.google.com/spreadsheets/d/1U-32UcwZP1k3saKeaH1mbvEAOfZRdNHNkWK2GI1rpPM/edit#gid=472150521">here</a>.
Detailed instructions for interacting from R with Google Sheets with the required <code>googlesheets</code> package are <a href="https://github.com/jennybc/googlesheets">here</a>.</p>
<pre><code class="language-r">library(&quot;googlesheets&quot;); library(&quot;dplyr&quot;); library(knitr)
gs_auth() # Creates authorizaton token (.httr-oauth) in current directory if not present
sheetid &lt;-&quot;1U-32UcwZP1k3saKeaH1mbvEAOfZRdNHNkWK2GI1rpPM&quot;
gap &lt;- gs_key(sheetid)
mysheet &lt;- gs_read(gap, skip=4)
myDF &lt;- as.data.frame(mysheet)
myDF
</code></pre>
<h3 id="export-of-tabular-data">Export of tabular data</h3>
<pre><code class="language-r">write.table(myDF, file=&quot;myfile.xls&quot;, sep=&quot;\t&quot;, quote=FALSE, col.names=NA)
</code></pre>
<h3 id="line-wise-import">Line-wise import</h3>
<pre><code class="language-r">myDF &lt;- readLines(&quot;myData.txt&quot;)
</code></pre>
<h3 id="line-wise-export">Line-wise export</h3>
<pre><code class="language-r">writeLines(month.name, &quot;myData.txt&quot;)
</code></pre>
<h3 id="export-r-object">Export R object</h3>
<pre><code class="language-r">mylist &lt;- list(C1=iris[,1], C2=iris[,2]) # Example to export
saveRDS(mylist, &quot;mylist.rds&quot;)
</code></pre>
<h3 id="import-r-object">Import R object</h3>
<pre><code class="language-r">mylist &lt;- readRDS(&quot;mylist.rds&quot;)
</code></pre>
<h3 id="copy-and-paste-into-r">Copy and paste into R</h3>
<p>On Windows/Linux systems</p>
<pre><code class="language-r">read.delim(&quot;clipboard&quot;) 
</code></pre>
<p>On Mac OS X systems</p>
<pre><code class="language-r">read.delim(pipe(&quot;pbpaste&quot;)) 
</code></pre>
<h3 id="copy-and-paste-from-r">Copy and paste from R</h3>
<p>On Windows/Linux systems</p>
<pre><code class="language-r">write.table(iris, &quot;clipboard&quot;, sep=&quot;\t&quot;, col.names=NA, quote=F) 
</code></pre>
<p>On Mac OS X systems</p>
<pre><code class="language-r">zz &lt;- pipe('pbcopy', 'w')
write.table(iris, zz, sep=&quot;\t&quot;, col.names=NA, quote=F)
close(zz) 
</code></pre>
<h3 id="homework-3a">Homework 3A</h3>
<p>Homework 3A: <a href="https://girke.bioinformatics.ucr.edu/GEN242/assignments/homework/hw3/">Object Subsetting Routines and Import/Export</a></p>
<h2 id="useful-r-functions">Useful R Functions</h2>
<h3 id="unique-entries">Unique entries</h3>
<p>Make vector entries unique with <code>unique</code></p>
<pre><code class="language-r">length(iris$Sepal.Length)
</code></pre>
<pre><code>## [1] 150
</code></pre>
<pre><code class="language-r">length(unique(iris$Sepal.Length))
</code></pre>
<pre><code>## [1] 35
</code></pre>
<h3 id="count-occurrences">Count occurrences</h3>
<p>Count occurrences of entries with <code>table</code></p>
<pre><code class="language-r">table(iris$Species)
</code></pre>
<pre><code>## 
##     setosa versicolor  virginica 
##         50         50         50
</code></pre>
<h3 id="aggregate-data">Aggregate data</h3>
<p>Compute aggregate statistics with <code>aggregate</code></p>
<pre><code class="language-r">aggregate(iris[,1:4], by=list(iris$Species), FUN=mean, na.rm=TRUE)
</code></pre>
<pre><code>##      Group.1 Sepal.Length Sepal.Width Petal.Length Petal.Width
## 1     setosa        5.006       3.428        1.462       0.246
## 2 versicolor        5.936       2.770        4.260       1.326
## 3  virginica        6.588       2.974        5.552       2.026
</code></pre>
<h3 id="intersect-data">Intersect data</h3>
<p>Compute intersect between two vectors with <code>%in%</code></p>
<pre><code class="language-r">month.name %in% c(&quot;May&quot;, &quot;July&quot;)
</code></pre>
<pre><code>##  [1] FALSE FALSE FALSE FALSE  TRUE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE
</code></pre>
<h3 id="merge-data-frames">Merge data frames</h3>
<p>Join two data frames by common field entries with <code>merge</code> (here row names <code>by.x=0</code>). To obtain only the common rows, change <code>all=TRUE</code> to <code>all=FALSE</code>. To merge on specific columns, refer to them by their position numbers or their column names.</p>
<pre><code class="language-r">frame1 &lt;- iris[sample(1:length(iris[,1]), 30), ]
frame1[1:2,]
</code></pre>
<pre><code>##    Sepal.Length Sepal.Width Petal.Length Petal.Width    Species
## 72          6.1         2.8          4.0         1.3 versicolor
## 45          5.1         3.8          1.9         0.4     setosa
</code></pre>
<pre><code class="language-r">dim(frame1)
</code></pre>
<pre><code>## [1] 30  5
</code></pre>
<pre><code class="language-r">my_result &lt;- merge(frame1, iris, by.x = 0, by.y = 0, all = TRUE)
dim(my_result)
</code></pre>
<pre><code>## [1] 150  11
</code></pre>
<h2 id="dplyr-environment">dplyr Environment</h2>
<p>Modern object classes and methods for handling <code>data.frame</code> like structures
are provided by the <code>dplyr</code> and <code>data.table</code> packages. The following gives a
short introduction to the usage and functionalities of the <code>dplyr</code> package.
More detailed tutorials on this topic can be found here:</p>
<ul>
<li><a href="https://rdrr.io/cran/dplyr/">dplyr: A Grammar of Data Manipulation</a></li>
<li><a href="https://cran.r-project.org/web/packages/dplyr/vignettes/dplyr.html">Introduction to <code>dplyr</code></a></li>
<li><a href="http://genomicsclass.github.io/book/pages/dplyr_tutorial.html">Tutorial on <code>dplyr</code></a></li>
<li><a href="http://stat545.com/bit001_dplyr-cheatsheet.html">Cheatsheet for Joins from Jenny Bryan</a></li>
<li><a href="https://cran.r-project.org/web/packages/tibble/vignettes/tibble.html">Tibbles</a></li>
<li><a href="https://www.r-bloggers.com/intro-to-the-data-table-package/">Intro to <code>data.table</code> package</a></li>
<li><a href="https://www.r-bloggers.com/working-with-large-datasets-with-dplyr-and-data-table/">Big data with <code>dplyr</code> and <code>data.table</code></a></li>
<li><a href="https://www.r-bloggers.com/fast-data-lookups-in-r-dplyr-vs-data-table/">Fast lookups with <code>dplyr</code> and <code>data.table</code></a></li>
</ul>
<h3 id="installation">Installation</h3>
<p>The <code>dplyr</code> environment has evolved into an ecosystem of packages. To simplify
package management, one can install and load the entire collection via the
<code>tidyverse</code> package. For more details on <code>tidyverse</code> see
<a href="http://tidyverse.org/">here</a>.</p>
<pre><code class="language-r">install.packages(&quot;tidyverse&quot;)
</code></pre>
<h3 id="construct-a-tibble-tibble">Construct a <code>tibble</code> (<code>tibble</code>)</h3>
<pre><code class="language-r">library(tidyverse)
as_tibble(iris) # coerce data.frame to tibble tbl
</code></pre>
<pre><code>## # A tibble: 150 x 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;fct&gt;  
##  1          5.1         3.5          1.4         0.2 setosa 
##  2          4.9         3            1.4         0.2 setosa 
##  3          4.7         3.2          1.3         0.2 setosa 
##  4          4.6         3.1          1.5         0.2 setosa 
##  5          5           3.6          1.4         0.2 setosa 
##  6          5.4         3.9          1.7         0.4 setosa 
##  7          4.6         3.4          1.4         0.3 setosa 
##  8          5           3.4          1.5         0.2 setosa 
##  9          4.4         2.9          1.4         0.2 setosa 
## 10          4.9         3.1          1.5         0.1 setosa 
## # … with 140 more rows
</code></pre>
<p>Alternative functions producing the same result include <code>as_data_frame</code> and <code>tbl_df</code>:</p>
<pre><code class="language-r">as_data_frame(iris) 
tbl_df(iris) 
</code></pre>
<h3 id="reading-and-writing-tabular-files">Reading and writing tabular files</h3>
<p>While the base R read/write utilities can be used for <code>data.frames</code>, best time
performance with the least amount of typing is achieved with the export/import
functions from the <code>readr</code> package. For very large files the <code>fread</code> function from
the <code>data.table</code> package achieves the best time performance.</p>
<h4 id="import-with-readr">Import with <code>readr</code></h4>
<p>Import functions provided by <code>readr</code> include:</p>
<ul>
<li><code>read_csv()</code>: comma separated (CSV) files</li>
<li><code>read_tsv()</code>: tab separated files</li>
<li><code>read_delim()</code>: general delimited files</li>
<li><code>read_fwf()</code>: fixed width files</li>
<li><code>read_table()</code>: tabular files where colums are separated by white-space.</li>
<li><code>read_log()</code>: web log files</li>
</ul>
<p>Create a sample tab delimited file for import</p>
<pre><code class="language-r">write_tsv(iris, &quot;iris.txt&quot;) # Creates sample file
</code></pre>
<p>Import with <code>read_tsv</code></p>
<pre><code class="language-r">iris_df &lt;- read_tsv(&quot;iris.txt&quot;) # Import with read_tbv from readr package
iris_df
</code></pre>
<pre><code>## # A tibble: 150 x 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
##  1          5.1         3.5          1.4         0.2 setosa 
##  2          4.9         3            1.4         0.2 setosa 
##  3          4.7         3.2          1.3         0.2 setosa 
##  4          4.6         3.1          1.5         0.2 setosa 
##  5          5           3.6          1.4         0.2 setosa 
##  6          5.4         3.9          1.7         0.4 setosa 
##  7          4.6         3.4          1.4         0.3 setosa 
##  8          5           3.4          1.5         0.2 setosa 
##  9          4.4         2.9          1.4         0.2 setosa 
## 10          4.9         3.1          1.5         0.1 setosa 
## # … with 140 more rows
</code></pre>
<p>To import Google Sheets directly into R, see <a href="http://girke.bioinformatics.ucr.edu/GEN242/mydoc_Rbasics_10.html">here</a>.</p>
<h4 id="fast-table-import-with-fread">Fast table import with <code>fread</code></h4>
<p>The <code>fread</code> function from the <code>data.table</code> package provides the best time performance for reading large
tabular files into R.</p>
<pre><code class="language-r">library(data.table)
iris_df &lt;- as_data_frame(fread(&quot;iris.txt&quot;)) # Import with fread and conversion to tibble
iris_df
</code></pre>
<pre><code>## # A tibble: 150 x 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
##  1          5.1         3.5          1.4         0.2 setosa 
##  2          4.9         3            1.4         0.2 setosa 
##  3          4.7         3.2          1.3         0.2 setosa 
##  4          4.6         3.1          1.5         0.2 setosa 
##  5          5           3.6          1.4         0.2 setosa 
##  6          5.4         3.9          1.7         0.4 setosa 
##  7          4.6         3.4          1.4         0.3 setosa 
##  8          5           3.4          1.5         0.2 setosa 
##  9          4.4         2.9          1.4         0.2 setosa 
## 10          4.9         3.1          1.5         0.1 setosa 
## # … with 140 more rows
</code></pre>
<p>Note: to ignore lines starting with comment signs, one can pass on to <code>fread</code> a shell
command for preprocessing the file. The following example illustrates this option.</p>
<pre><code class="language-r">fread(&quot;grep -v '^#' iris.txt&quot;) 
</code></pre>
<h4 id="export-with-readr">Export with <code>readr</code></h4>
<p>Export function provided by <code>readr</code> inlcude</p>
<ul>
<li><code>write_delim()</code>: general delimited files</li>
<li><code>write_csv()</code>: comma separated (CSV) files</li>
<li><code>write_excel_csv()</code>: excel style CSV files</li>
<li><code>write_tsv()</code>: tab separated files</li>
</ul>
<p>For instance, the <code>write_tsv</code> function writes a <code>data.frame</code> or <code>tibble</code> to a tab delimited file with much nicer
default settings than the base R <code>write.table</code> function.</p>
<pre><code class="language-r">write_tsv(iris_df, &quot;iris.txt&quot;)
</code></pre>
<h3 id="column-and-row-binds">Column and row binds</h3>
<p>The equivalents to base R’s <code>rbind</code> and <code>cbind</code> are <code>bind_rows</code> and <code>bind_cols</code>, respectively.</p>
<pre><code class="language-r">bind_cols(iris_df, iris_df)
</code></pre>
<pre><code>## # A tibble: 150 x 10
##    Sepal.Length...1 Sepal.Width...2 Petal.Length...3 Petal.Width...4 Species...5 Sepal.Length...6
##               &lt;dbl&gt;           &lt;dbl&gt;            &lt;dbl&gt;           &lt;dbl&gt; &lt;chr&gt;                  &lt;dbl&gt;
##  1              5.1             3.5              1.4             0.2 setosa                   5.1
##  2              4.9             3                1.4             0.2 setosa                   4.9
##  3              4.7             3.2              1.3             0.2 setosa                   4.7
##  4              4.6             3.1              1.5             0.2 setosa                   4.6
##  5              5               3.6              1.4             0.2 setosa                   5  
##  6              5.4             3.9              1.7             0.4 setosa                   5.4
##  7              4.6             3.4              1.4             0.3 setosa                   4.6
##  8              5               3.4              1.5             0.2 setosa                   5  
##  9              4.4             2.9              1.4             0.2 setosa                   4.4
## 10              4.9             3.1              1.5             0.1 setosa                   4.9
## # … with 140 more rows, and 4 more variables: Sepal.Width...7 &lt;dbl&gt;, Petal.Length...8 &lt;dbl&gt;,
## #   Petal.Width...9 &lt;dbl&gt;, Species...10 &lt;chr&gt;
</code></pre>
<pre><code class="language-r">bind_rows(iris_df, iris_df)
</code></pre>
<pre><code>## # A tibble: 300 x 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
##  1          5.1         3.5          1.4         0.2 setosa 
##  2          4.9         3            1.4         0.2 setosa 
##  3          4.7         3.2          1.3         0.2 setosa 
##  4          4.6         3.1          1.5         0.2 setosa 
##  5          5           3.6          1.4         0.2 setosa 
##  6          5.4         3.9          1.7         0.4 setosa 
##  7          4.6         3.4          1.4         0.3 setosa 
##  8          5           3.4          1.5         0.2 setosa 
##  9          4.4         2.9          1.4         0.2 setosa 
## 10          4.9         3.1          1.5         0.1 setosa 
## # … with 290 more rows
</code></pre>
<h3 id="extract-column-as-vector">Extract column as vector</h3>
<p>The subsetting operators <code>[[</code> and <code>$</code>can be used to extract from a <code>tibble</code> single columns as vector.</p>
<pre><code class="language-r">iris_df[[5]][1:12]
</code></pre>
<pre><code>##  [1] &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot;
## [11] &quot;setosa&quot; &quot;setosa&quot;
</code></pre>
<pre><code class="language-r">iris_df$Species[1:12]
</code></pre>
<pre><code>##  [1] &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot;
## [11] &quot;setosa&quot; &quot;setosa&quot;
</code></pre>
<h3 id="important-dplyr-functions">Important <code>dplyr</code> functions</h3>
<ol>
<li><code>filter()</code> and <code>slice()</code></li>
<li><code>arrange()</code></li>
<li><code>select()</code> and <code>rename()</code></li>
<li><code>distinct()</code></li>
<li><code>mutate()</code> and <code>transmute()</code></li>
<li><code>summarise()</code></li>
<li><code>sample_n()</code> and <code>sample_frac()</code></li>
</ol>
<h3 id="slice-and-filter-functions">Slice and filter functions</h3>
<h4 id="filter-function">Filter function</h4>
<pre><code class="language-r">filter(iris_df, Sepal.Length &gt; 7.5, Species==&quot;virginica&quot;)
</code></pre>
<pre><code>## # A tibble: 6 x 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species  
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;    
## 1          7.6         3            6.6         2.1 virginica
## 2          7.7         3.8          6.7         2.2 virginica
## 3          7.7         2.6          6.9         2.3 virginica
## 4          7.7         2.8          6.7         2   virginica
## 5          7.9         3.8          6.4         2   virginica
## 6          7.7         3            6.1         2.3 virginica
</code></pre>
<h4 id="base-r-code-equivalent">Base R code equivalent</h4>
<pre><code class="language-r">iris_df[iris_df[, &quot;Sepal.Length&quot;] &gt; 7.5 &amp; iris_df[, &quot;Species&quot;]==&quot;virginica&quot;, ]
</code></pre>
<pre><code>## # A tibble: 6 x 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species  
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;    
## 1          7.6         3            6.6         2.1 virginica
## 2          7.7         3.8          6.7         2.2 virginica
## 3          7.7         2.6          6.9         2.3 virginica
## 4          7.7         2.8          6.7         2   virginica
## 5          7.9         3.8          6.4         2   virginica
## 6          7.7         3            6.1         2.3 virginica
</code></pre>
<h4 id="including-boolean-operators">Including boolean operators</h4>
<pre><code class="language-r">filter(iris_df, Sepal.Length &gt; 7.5 | Sepal.Length &lt; 5.5, Species==&quot;virginica&quot;)
</code></pre>
<pre><code>## # A tibble: 7 x 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species  
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;    
## 1          7.6         3            6.6         2.1 virginica
## 2          4.9         2.5          4.5         1.7 virginica
## 3          7.7         3.8          6.7         2.2 virginica
## 4          7.7         2.6          6.9         2.3 virginica
## 5          7.7         2.8          6.7         2   virginica
## 6          7.9         3.8          6.4         2   virginica
## 7          7.7         3            6.1         2.3 virginica
</code></pre>
<h4 id="subset-rows-by-position">Subset rows by position</h4>
<p><code>dplyr</code> approach</p>
<pre><code class="language-r">slice(iris_df, 1:2)
</code></pre>
<pre><code>## # A tibble: 2 x 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
## 1          5.1         3.5          1.4         0.2 setosa 
## 2          4.9         3            1.4         0.2 setosa
</code></pre>
<p>Base R code equivalent</p>
<pre><code class="language-r">iris_df[1:2,]
</code></pre>
<pre><code>## # A tibble: 2 x 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
## 1          5.1         3.5          1.4         0.2 setosa 
## 2          4.9         3            1.4         0.2 setosa
</code></pre>
<h4 id="subset-rows-by-names">Subset rows by names</h4>
<p>Since <code>tibbles</code> do not contain row names, row wise subsetting via the <code>[,]</code> operator cannot be used.
However, the corresponding behavior can be achieved by passing to <code>select</code> a row position index
obtained by basic R intersect utilities such as <code>match</code>.</p>
<p>Create a suitable test <code>tibble</code></p>
<pre><code class="language-r">df1 &lt;- bind_cols(data_frame(ids1=paste0(&quot;g&quot;, 1:10)), as_data_frame(matrix(1:40, 10, 4, dimnames=list(1:10, paste0(&quot;CA&quot;, 1:4)))))
df1
</code></pre>
<pre><code>## # A tibble: 10 x 5
##    ids1    CA1   CA2   CA3   CA4
##    &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
##  1 g1        1    11    21    31
##  2 g2        2    12    22    32
##  3 g3        3    13    23    33
##  4 g4        4    14    24    34
##  5 g5        5    15    25    35
##  6 g6        6    16    26    36
##  7 g7        7    17    27    37
##  8 g8        8    18    28    38
##  9 g9        9    19    29    39
## 10 g10      10    20    30    40
</code></pre>
<p><code>dplyr</code> approach</p>
<pre><code class="language-r">slice(df1, match(c(&quot;g10&quot;, &quot;g4&quot;, &quot;g4&quot;), df1$ids1))
</code></pre>
<pre><code>## # A tibble: 3 x 5
##   ids1    CA1   CA2   CA3   CA4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1 g10      10    20    30    40
## 2 g4        4    14    24    34
## 3 g4        4    14    24    34
</code></pre>
<p>Base R equivalent</p>
<pre><code class="language-r">df1_old &lt;- as.data.frame(df1)
rownames(df1_old) &lt;- df1_old[,1]
df1_old[c(&quot;g10&quot;, &quot;g4&quot;, &quot;g4&quot;),]
</code></pre>
<pre><code>##      ids1 CA1 CA2 CA3 CA4
## g10   g10  10  20  30  40
## g4     g4   4  14  24  34
## g4.1   g4   4  14  24  34
</code></pre>
<h3 id="sorting-with-arrange">Sorting with <code>arrange</code></h3>
<p>Row-wise ordering based on specific columns</p>
<p><code>dplyr</code> approach</p>
<pre><code class="language-r">arrange(iris_df, Species, Sepal.Length, Sepal.Width)
</code></pre>
<pre><code>## # A tibble: 150 x 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
##  1          4.3         3            1.1         0.1 setosa 
##  2          4.4         2.9          1.4         0.2 setosa 
##  3          4.4         3            1.3         0.2 setosa 
##  4          4.4         3.2          1.3         0.2 setosa 
##  5          4.5         2.3          1.3         0.3 setosa 
##  6          4.6         3.1          1.5         0.2 setosa 
##  7          4.6         3.2          1.4         0.2 setosa 
##  8          4.6         3.4          1.4         0.3 setosa 
##  9          4.6         3.6          1           0.2 setosa 
## 10          4.7         3.2          1.3         0.2 setosa 
## # … with 140 more rows
</code></pre>
<p>For ordering descendingly use <code>desc()</code> function</p>
<pre><code class="language-r">arrange(iris_df, desc(Species), Sepal.Length, Sepal.Width)
</code></pre>
<pre><code>## # A tibble: 150 x 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species  
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;    
##  1          4.9         2.5          4.5         1.7 virginica
##  2          5.6         2.8          4.9         2   virginica
##  3          5.7         2.5          5           2   virginica
##  4          5.8         2.7          5.1         1.9 virginica
##  5          5.8         2.7          5.1         1.9 virginica
##  6          5.8         2.8          5.1         2.4 virginica
##  7          5.9         3            5.1         1.8 virginica
##  8          6           2.2          5           1.5 virginica
##  9          6           3            4.8         1.8 virginica
## 10          6.1         2.6          5.6         1.4 virginica
## # … with 140 more rows
</code></pre>
<p>Base R code equivalent</p>
<pre><code class="language-r">iris_df[order(iris_df$Species, iris_df$Sepal.Length, iris_df$Sepal.Width), ]
</code></pre>
<pre><code>## # A tibble: 150 x 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
##  1          4.3         3            1.1         0.1 setosa 
##  2          4.4         2.9          1.4         0.2 setosa 
##  3          4.4         3            1.3         0.2 setosa 
##  4          4.4         3.2          1.3         0.2 setosa 
##  5          4.5         2.3          1.3         0.3 setosa 
##  6          4.6         3.1          1.5         0.2 setosa 
##  7          4.6         3.2          1.4         0.2 setosa 
##  8          4.6         3.4          1.4         0.3 setosa 
##  9          4.6         3.6          1           0.2 setosa 
## 10          4.7         3.2          1.3         0.2 setosa 
## # … with 140 more rows
</code></pre>
<pre><code class="language-r">iris_df[order(iris_df$Species, decreasing=TRUE), ] 
</code></pre>
<pre><code>## # A tibble: 150 x 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species  
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;    
##  1          6.3         3.3          6           2.5 virginica
##  2          5.8         2.7          5.1         1.9 virginica
##  3          7.1         3            5.9         2.1 virginica
##  4          6.3         2.9          5.6         1.8 virginica
##  5          6.5         3            5.8         2.2 virginica
##  6          7.6         3            6.6         2.1 virginica
##  7          4.9         2.5          4.5         1.7 virginica
##  8          7.3         2.9          6.3         1.8 virginica
##  9          6.7         2.5          5.8         1.8 virginica
## 10          7.2         3.6          6.1         2.5 virginica
## # … with 140 more rows
</code></pre>
<h3 id="select-columns-with-select">Select columns with <code>select</code></h3>
<p>Select specific columns</p>
<pre><code class="language-r">select(iris_df, Species, Petal.Length, Sepal.Length)
</code></pre>
<pre><code>## # A tibble: 150 x 3
##    Species Petal.Length Sepal.Length
##    &lt;chr&gt;          &lt;dbl&gt;        &lt;dbl&gt;
##  1 setosa           1.4          5.1
##  2 setosa           1.4          4.9
##  3 setosa           1.3          4.7
##  4 setosa           1.5          4.6
##  5 setosa           1.4          5  
##  6 setosa           1.7          5.4
##  7 setosa           1.4          4.6
##  8 setosa           1.5          5  
##  9 setosa           1.4          4.4
## 10 setosa           1.5          4.9
## # … with 140 more rows
</code></pre>
<p>Select range of columns by name</p>
<pre><code class="language-r">select(iris_df, Sepal.Length : Petal.Width)
</code></pre>
<pre><code>## # A tibble: 150 x 4
##    Sepal.Length Sepal.Width Petal.Length Petal.Width
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;
##  1          5.1         3.5          1.4         0.2
##  2          4.9         3            1.4         0.2
##  3          4.7         3.2          1.3         0.2
##  4          4.6         3.1          1.5         0.2
##  5          5           3.6          1.4         0.2
##  6          5.4         3.9          1.7         0.4
##  7          4.6         3.4          1.4         0.3
##  8          5           3.4          1.5         0.2
##  9          4.4         2.9          1.4         0.2
## 10          4.9         3.1          1.5         0.1
## # … with 140 more rows
</code></pre>
<p>Drop specific columns (here range)</p>
<pre><code class="language-r">select(iris_df, -(Sepal.Length : Petal.Width))
</code></pre>
<pre><code>## # A tibble: 150 x 1
##    Species
##    &lt;chr&gt;  
##  1 setosa 
##  2 setosa 
##  3 setosa 
##  4 setosa 
##  5 setosa 
##  6 setosa 
##  7 setosa 
##  8 setosa 
##  9 setosa 
## 10 setosa 
## # … with 140 more rows
</code></pre>
<h3 id="renaming-columns-with-rename">Renaming columns with <code>rename</code></h3>
<p><code>dplyr</code> approach</p>
<pre><code class="language-r">rename(iris_df, new_col_name = Species)
</code></pre>
<pre><code>## # A tibble: 150 x 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width new_col_name
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;       
##  1          5.1         3.5          1.4         0.2 setosa      
##  2          4.9         3            1.4         0.2 setosa      
##  3          4.7         3.2          1.3         0.2 setosa      
##  4          4.6         3.1          1.5         0.2 setosa      
##  5          5           3.6          1.4         0.2 setosa      
##  6          5.4         3.9          1.7         0.4 setosa      
##  7          4.6         3.4          1.4         0.3 setosa      
##  8          5           3.4          1.5         0.2 setosa      
##  9          4.4         2.9          1.4         0.2 setosa      
## 10          4.9         3.1          1.5         0.1 setosa      
## # … with 140 more rows
</code></pre>
<p>Base R code approach</p>
<pre><code class="language-r">colnames(iris_df)[colnames(iris_df)==&quot;Species&quot;] &lt;- &quot;new_col_names&quot;
</code></pre>
<h3 id="obtain-unique-rows-with-distinct">Obtain unique rows with <code>distinct</code></h3>
<p><code>dplyr</code> approach</p>
<pre><code class="language-r">distinct(iris_df, Species, .keep_all=TRUE)
</code></pre>
<pre><code>## # A tibble: 3 x 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species   
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;     
## 1          5.1         3.5          1.4         0.2 setosa    
## 2          7           3.2          4.7         1.4 versicolor
## 3          6.3         3.3          6           2.5 virginica
</code></pre>
<p>Base R code approach</p>
<pre><code class="language-r">iris_df[!duplicated(iris_df$Species),]
</code></pre>
<pre><code>## # A tibble: 3 x 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species   
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;     
## 1          5.1         3.5          1.4         0.2 setosa    
## 2          7           3.2          4.7         1.4 versicolor
## 3          6.3         3.3          6           2.5 virginica
</code></pre>
<h3 id="add-columns">Add columns</h3>
<h4 id="mutate"><code>mutate</code></h4>
<p>The <code>mutate</code> function allows to append columns to existing ones.</p>
<pre><code class="language-r">mutate(iris_df, Ratio = Sepal.Length / Sepal.Width, Sum = Sepal.Length + Sepal.Width)
</code></pre>
<pre><code>## # A tibble: 150 x 7
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species Ratio   Sum
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;
##  1          5.1         3.5          1.4         0.2 setosa   1.46   8.6
##  2          4.9         3            1.4         0.2 setosa   1.63   7.9
##  3          4.7         3.2          1.3         0.2 setosa   1.47   7.9
##  4          4.6         3.1          1.5         0.2 setosa   1.48   7.7
##  5          5           3.6          1.4         0.2 setosa   1.39   8.6
##  6          5.4         3.9          1.7         0.4 setosa   1.38   9.3
##  7          4.6         3.4          1.4         0.3 setosa   1.35   8  
##  8          5           3.4          1.5         0.2 setosa   1.47   8.4
##  9          4.4         2.9          1.4         0.2 setosa   1.52   7.3
## 10          4.9         3.1          1.5         0.1 setosa   1.58   8  
## # … with 140 more rows
</code></pre>
<h4 id="transmute"><code>transmute</code></h4>
<p>The <code>transmute</code> function does the same as <code>mutate</code> but drops existing columns</p>
<pre><code class="language-r">transmute(iris_df, Ratio = Sepal.Length / Sepal.Width, Sum = Sepal.Length + Sepal.Width)
</code></pre>
<pre><code>## # A tibble: 150 x 2
##    Ratio   Sum
##    &lt;dbl&gt; &lt;dbl&gt;
##  1  1.46   8.6
##  2  1.63   7.9
##  3  1.47   7.9
##  4  1.48   7.7
##  5  1.39   8.6
##  6  1.38   9.3
##  7  1.35   8  
##  8  1.47   8.4
##  9  1.52   7.3
## 10  1.58   8  
## # … with 140 more rows
</code></pre>
<h4 id="bind_cols"><code>bind_cols</code></h4>
<p>The <code>bind_cols</code> function is the equivalent of <code>cbind</code> in base R. To add rows, use the corresponding
<code>bind_rows</code> function.</p>
<pre><code class="language-r">bind_cols(iris_df, iris_df)
</code></pre>
<pre><code>## # A tibble: 150 x 10
##    Sepal.Length...1 Sepal.Width...2 Petal.Length...3 Petal.Width...4 Species...5 Sepal.Length...6
##               &lt;dbl&gt;           &lt;dbl&gt;            &lt;dbl&gt;           &lt;dbl&gt; &lt;chr&gt;                  &lt;dbl&gt;
##  1              5.1             3.5              1.4             0.2 setosa                   5.1
##  2              4.9             3                1.4             0.2 setosa                   4.9
##  3              4.7             3.2              1.3             0.2 setosa                   4.7
##  4              4.6             3.1              1.5             0.2 setosa                   4.6
##  5              5               3.6              1.4             0.2 setosa                   5  
##  6              5.4             3.9              1.7             0.4 setosa                   5.4
##  7              4.6             3.4              1.4             0.3 setosa                   4.6
##  8              5               3.4              1.5             0.2 setosa                   5  
##  9              4.4             2.9              1.4             0.2 setosa                   4.4
## 10              4.9             3.1              1.5             0.1 setosa                   4.9
## # … with 140 more rows, and 4 more variables: Sepal.Width...7 &lt;dbl&gt;, Petal.Length...8 &lt;dbl&gt;,
## #   Petal.Width...9 &lt;dbl&gt;, Species...10 &lt;chr&gt;
</code></pre>
<h3 id="summarize-data">Summarize data</h3>
<p>Summary calculation on single column</p>
<pre><code class="language-r">summarize(iris_df, mean(Petal.Length))
</code></pre>
<pre><code>## # A tibble: 1 x 1
##   `mean(Petal.Length)`
##                  &lt;dbl&gt;
## 1                 3.76
</code></pre>
<p>Summary calculation on many columns</p>
<pre><code class="language-r">summarize_all(iris_df[,1:4], mean)
</code></pre>
<pre><code>## # A tibble: 1 x 4
##   Sepal.Length Sepal.Width Petal.Length Petal.Width
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;
## 1         5.84        3.06         3.76        1.20
</code></pre>
<p>Summarize by grouping column</p>
<pre><code class="language-r">summarize(group_by(iris_df, Species), mean(Petal.Length))
</code></pre>
<pre><code>## # A tibble: 3 x 2
##   Species    `mean(Petal.Length)`
##   &lt;chr&gt;                     &lt;dbl&gt;
## 1 setosa                     1.46
## 2 versicolor                 4.26
## 3 virginica                  5.55
</code></pre>
<p>Aggregate summaries</p>
<pre><code class="language-r">summarize_all(group_by(iris_df, Species), mean) 
</code></pre>
<pre><code>## # A tibble: 3 x 5
##   Species    Sepal.Length Sepal.Width Petal.Length Petal.Width
##   &lt;chr&gt;             &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;
## 1 setosa             5.01        3.43         1.46       0.246
## 2 versicolor         5.94        2.77         4.26       1.33 
## 3 virginica          6.59        2.97         5.55       2.03
</code></pre>
<p>Note: <code>group_by</code> does the looping for the user similar to <code>aggregate</code> or <code>tapply</code>.</p>
<h3 id="merging-tibbles">Merging tibbles</h3>
<p>The <code>dplyr</code> package provides several join functions for merging <code>tibbles</code> by a common key column
similar to the <code>merge</code> function in base R. These <code>*_join</code> functions include:</p>
<ul>
<li><code>inner_join()</code>: returns join only for rows matching among both <code>tibbles</code></li>
<li><code>full_join()</code>: returns join for all (matching and non-matching) rows of two <code>tibbles</code></li>
<li><code>left_join()</code>: returns join for all rows in first <code>tibble</code></li>
<li><code>right_join()</code>: returns join for all rows in second <code>tibble</code></li>
<li><code>anti_join()</code>: returns for first <code>tibble</code> only those rows that have no match in the second one</li>
</ul>
<p>Sample <code>tibbles</code> to illustrate <code>*.join</code> functions.</p>
<pre><code class="language-r">df1 &lt;- bind_cols(data_frame(ids1=paste0(&quot;g&quot;, 1:10)), as_data_frame(matrix(1:40, 10, 4, dimnames=list(1:10, paste0(&quot;CA&quot;, 1:4)))))
df1
</code></pre>
<pre><code>## # A tibble: 10 x 5
##    ids1    CA1   CA2   CA3   CA4
##    &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
##  1 g1        1    11    21    31
##  2 g2        2    12    22    32
##  3 g3        3    13    23    33
##  4 g4        4    14    24    34
##  5 g5        5    15    25    35
##  6 g6        6    16    26    36
##  7 g7        7    17    27    37
##  8 g8        8    18    28    38
##  9 g9        9    19    29    39
## 10 g10      10    20    30    40
</code></pre>
<pre><code class="language-r">df2 &lt;- bind_cols(data_frame(ids2=paste0(&quot;g&quot;, c(2,5,11,12))), as_data_frame(matrix(1:16, 4, 4, dimnames=list(1:4, paste0(&quot;CB&quot;, 1:4)))))
df2
</code></pre>
<pre><code>## # A tibble: 4 x 5
##   ids2    CB1   CB2   CB3   CB4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1 g2        1     5     9    13
## 2 g5        2     6    10    14
## 3 g11       3     7    11    15
## 4 g12       4     8    12    16
</code></pre>
<h4 id="inner-join">Inner join</h4>
<pre><code class="language-r">inner_join(df1, df2, by=c(&quot;ids1&quot;=&quot;ids2&quot;))
</code></pre>
<pre><code>## # A tibble: 2 x 9
##   ids1    CA1   CA2   CA3   CA4   CB1   CB2   CB3   CB4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1 g2        2    12    22    32     1     5     9    13
## 2 g5        5    15    25    35     2     6    10    14
</code></pre>
<h4 id="left-join">Left join</h4>
<pre><code class="language-r">left_join(df1, df2, by=c(&quot;ids1&quot;=&quot;ids2&quot;))
</code></pre>
<pre><code>## # A tibble: 10 x 9
##    ids1    CA1   CA2   CA3   CA4   CB1   CB2   CB3   CB4
##    &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
##  1 g1        1    11    21    31    NA    NA    NA    NA
##  2 g2        2    12    22    32     1     5     9    13
##  3 g3        3    13    23    33    NA    NA    NA    NA
##  4 g4        4    14    24    34    NA    NA    NA    NA
##  5 g5        5    15    25    35     2     6    10    14
##  6 g6        6    16    26    36    NA    NA    NA    NA
##  7 g7        7    17    27    37    NA    NA    NA    NA
##  8 g8        8    18    28    38    NA    NA    NA    NA
##  9 g9        9    19    29    39    NA    NA    NA    NA
## 10 g10      10    20    30    40    NA    NA    NA    NA
</code></pre>
<h4 id="right-join">Right join</h4>
<pre><code class="language-r">right_join(df1, df2, by=c(&quot;ids1&quot;=&quot;ids2&quot;))
</code></pre>
<pre><code>## # A tibble: 4 x 9
##   ids1    CA1   CA2   CA3   CA4   CB1   CB2   CB3   CB4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1 g2        2    12    22    32     1     5     9    13
## 2 g5        5    15    25    35     2     6    10    14
## 3 g11      NA    NA    NA    NA     3     7    11    15
## 4 g12      NA    NA    NA    NA     4     8    12    16
</code></pre>
<h4 id="full-join">Full join</h4>
<pre><code class="language-r">full_join(df1, df2, by=c(&quot;ids1&quot;=&quot;ids2&quot;))
</code></pre>
<pre><code>## # A tibble: 12 x 9
##    ids1    CA1   CA2   CA3   CA4   CB1   CB2   CB3   CB4
##    &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
##  1 g1        1    11    21    31    NA    NA    NA    NA
##  2 g2        2    12    22    32     1     5     9    13
##  3 g3        3    13    23    33    NA    NA    NA    NA
##  4 g4        4    14    24    34    NA    NA    NA    NA
##  5 g5        5    15    25    35     2     6    10    14
##  6 g6        6    16    26    36    NA    NA    NA    NA
##  7 g7        7    17    27    37    NA    NA    NA    NA
##  8 g8        8    18    28    38    NA    NA    NA    NA
##  9 g9        9    19    29    39    NA    NA    NA    NA
## 10 g10      10    20    30    40    NA    NA    NA    NA
## 11 g11      NA    NA    NA    NA     3     7    11    15
## 12 g12      NA    NA    NA    NA     4     8    12    16
</code></pre>
<h4 id="anti-join">Anti join</h4>
<pre><code class="language-r">anti_join(df1, df2, by=c(&quot;ids1&quot;=&quot;ids2&quot;))
</code></pre>
<pre><code>## # A tibble: 8 x 5
##   ids1    CA1   CA2   CA3   CA4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1 g1        1    11    21    31
## 2 g3        3    13    23    33
## 3 g4        4    14    24    34
## 4 g6        6    16    26    36
## 5 g7        7    17    27    37
## 6 g8        8    18    28    38
## 7 g9        9    19    29    39
## 8 g10      10    20    30    40
</code></pre>
<p>For additional join options users want to cosult the <code>*_join</code> help pages.</p>
<h3 id="chaining">Chaining</h3>
<p>To simplify chaining of serveral operations, <code>dplyr</code> provides the <code>%&gt;%</code>
operator, where <code>x %&gt;% f(y)</code> turns into <code>f(x, y)</code>. This way one can pipe
together multiple operations by writing them from left-to-right or
top-to-bottom. This makes for easy to type and readable code.</p>
<h4 id="example-1">Example 1</h4>
<p>Series of data manipulations and export</p>
<pre><code class="language-r">read_tsv(&quot;iris.txt&quot;) %&gt;% # Import with read_tbv from readr package
    as_tibble() %&gt;% # Declare tibble to use
    select(Sepal.Length:Species) %&gt;% # Select columns
    filter(Species==&quot;setosa&quot;) %&gt;% # Filter rows by some value
    arrange(Sepal.Length) %&gt;% # Sort by some column
    mutate(Subtract=Petal.Length - Petal.Width) # Calculate and append
</code></pre>
<pre><code>## # A tibble: 50 x 6
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species Subtract
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;
##  1          4.3         3            1.1         0.1 setosa      1   
##  2          4.4         2.9          1.4         0.2 setosa      1.2 
##  3          4.4         3            1.3         0.2 setosa      1.1 
##  4          4.4         3.2          1.3         0.2 setosa      1.1 
##  5          4.5         2.3          1.3         0.3 setosa      1   
##  6          4.6         3.1          1.5         0.2 setosa      1.3 
##  7          4.6         3.4          1.4         0.3 setosa      1.10
##  8          4.6         3.6          1           0.2 setosa      0.8 
##  9          4.6         3.2          1.4         0.2 setosa      1.2 
## 10          4.7         3.2          1.3         0.2 setosa      1.1 
## # … with 40 more rows
</code></pre>
<pre><code class="language-r">    # write_tsv(&quot;iris.txt&quot;) # Export to file, omitted here to show result 
</code></pre>
<h4 id="example-2">Example 2</h4>
<p>Series of summary calculations for grouped data (<code>group_by</code>)</p>
<pre><code class="language-r">iris_df %&gt;% # Declare tibble to use 
    group_by(Species) %&gt;% # Group by species
    summarize(Mean_Sepal.Length=mean(Sepal.Length), 
              Max_Sepal.Length=max(Sepal.Length),
              Min_Sepal.Length=min(Sepal.Length),
              SD_Sepal.Length=sd(Sepal.Length),
              Total=n()) 
</code></pre>
<pre><code>## # A tibble: 3 x 6
##   Species    Mean_Sepal.Length Max_Sepal.Length Min_Sepal.Length SD_Sepal.Length Total
##   &lt;chr&gt;                  &lt;dbl&gt;            &lt;dbl&gt;            &lt;dbl&gt;           &lt;dbl&gt; &lt;int&gt;
## 1 setosa                  5.01              5.8              4.3           0.352    50
## 2 versicolor              5.94              7                4.9           0.516    50
## 3 virginica               6.59              7.9              4.9           0.636    50
</code></pre>
<h4 id="example-3">Example 3</h4>
<p>Combining <code>dplyr</code> chaining with <code>ggplot</code></p>
<pre><code class="language-r">iris_df %&gt;% 
    group_by(Species) %&gt;% 
    summarize_all(mean) %&gt;% 
    reshape2::melt(id.vars=c(&quot;Species&quot;), variable.name = &quot;Samples&quot;, value.name=&quot;Values&quot;) %&gt;%
    ggplot(aes(Samples, Values, fill = Species)) + 
        geom_bar(position=&quot;dodge&quot;, stat=&quot;identity&quot;)
</code></pre>
<p><img src="../../../en/manuals/rbasics/Rbasics_files/figure-html/plyr_chaining3-1.png" width="672" /></p>
<h2 id="sqlite-databases">SQLite Databases</h2>
<p><code>SQLite</code> is a lightweight relational database solution. The <code>RSQLite</code> package provides an easy to use interface to create, manage and query <code>SQLite</code> databases directly from R. Basic instructions
for using <code>SQLite</code> from the command-line are available <a href="https://www.sqlite.org/cli.html">here</a>. A short introduction to <code>RSQLite</code> is available <a href="https://github.com/rstats-db/RSQLite/blob/master/vignettes/RSQLite.Rmd">here</a>.</p>
<h2 id="loading-data-into-sqlite-databases">Loading data into SQLite databases</h2>
<p>The following loads two <code>data.frames</code> derived from the <code>iris</code> data set (here <code>mydf1</code> and <code>mydf2</code>)
into an SQLite database (here <code>test.db</code>).</p>
<pre><code class="language-r">library(RSQLite)
unlink(&quot;test.db&quot;) # Delete any existing test.db
mydb &lt;- dbConnect(SQLite(), &quot;test.db&quot;) # Creates database file test.db
mydf1 &lt;- data.frame(ids=paste0(&quot;id&quot;, seq_along(iris[,1])), iris)
mydf2 &lt;- mydf1[sample(seq_along(mydf1[,1]), 10),]
dbWriteTable(mydb, &quot;mydf1&quot;, mydf1)
dbWriteTable(mydb, &quot;mydf2&quot;, mydf2)
</code></pre>
<h3 id="list-names-of-tables-in-database">List names of tables in database</h3>
<pre><code class="language-r">dbListTables(mydb)
</code></pre>
<pre><code>## [1] &quot;mydf1&quot; &quot;mydf2&quot;
</code></pre>
<h3 id="import-table-into-dataframe">Import table into <code>data.frame</code></h3>
<pre><code class="language-r">dbGetQuery(mydb, 'SELECT * FROM mydf2')
</code></pre>
<pre><code>##      ids Sepal.Length Sepal.Width Petal.Length Petal.Width    Species
## 1  id111          6.5         3.2          5.1         2.0  virginica
## 2  id106          7.6         3.0          6.6         2.1  virginica
## 3  id117          6.5         3.0          5.5         1.8  virginica
## 4   id61          5.0         2.0          3.5         1.0 versicolor
## 5   id50          5.0         3.3          1.4         0.2     setosa
## 6  id123          7.7         2.8          6.7         2.0  virginica
## 7   id60          5.2         2.7          3.9         1.4 versicolor
## 8   id92          6.1         3.0          4.6         1.4 versicolor
## 9   id32          5.4         3.4          1.5         0.4     setosa
## 10  id84          6.0         2.7          5.1         1.6 versicolor
</code></pre>
<h3 id="query-database">Query database</h3>
<pre><code class="language-r">dbGetQuery(mydb, 'SELECT * FROM mydf1 WHERE &quot;Sepal.Length&quot; &lt; 4.6')
</code></pre>
<pre><code>##    ids Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 1  id9          4.4         2.9          1.4         0.2  setosa
## 2 id14          4.3         3.0          1.1         0.1  setosa
## 3 id39          4.4         3.0          1.3         0.2  setosa
## 4 id42          4.5         2.3          1.3         0.3  setosa
## 5 id43          4.4         3.2          1.3         0.2  setosa
</code></pre>
<h3 id="join-tables">Join tables</h3>
<p>The two tables can be joined on the shared <code>ids</code> column as follows.</p>
<pre><code class="language-r">dbGetQuery(mydb, 'SELECT * FROM mydf1, mydf2 WHERE mydf1.ids = mydf2.ids')
</code></pre>
<pre><code>##      ids Sepal.Length Sepal.Width Petal.Length Petal.Width    Species   ids Sepal.Length
## 1   id32          5.4         3.4          1.5         0.4     setosa  id32          5.4
## 2   id50          5.0         3.3          1.4         0.2     setosa  id50          5.0
## 3   id60          5.2         2.7          3.9         1.4 versicolor  id60          5.2
## 4   id61          5.0         2.0          3.5         1.0 versicolor  id61          5.0
## 5   id84          6.0         2.7          5.1         1.6 versicolor  id84          6.0
## 6   id92          6.1         3.0          4.6         1.4 versicolor  id92          6.1
## 7  id106          7.6         3.0          6.6         2.1  virginica id106          7.6
## 8  id111          6.5         3.2          5.1         2.0  virginica id111          6.5
## 9  id117          6.5         3.0          5.5         1.8  virginica id117          6.5
## 10 id123          7.7         2.8          6.7         2.0  virginica id123          7.7
##    Sepal.Width Petal.Length Petal.Width    Species
## 1          3.4          1.5         0.4     setosa
## 2          3.3          1.4         0.2     setosa
## 3          2.7          3.9         1.4 versicolor
## 4          2.0          3.5         1.0 versicolor
## 5          2.7          5.1         1.6 versicolor
## 6          3.0          4.6         1.4 versicolor
## 7          3.0          6.6         2.1  virginica
## 8          3.2          5.1         2.0  virginica
## 9          3.0          5.5         1.8  virginica
## 10         2.8          6.7         2.0  virginica
</code></pre>
<h2 id="graphics-in-r">Graphics in R</h2>
<h3 id="advantages">Advantages</h3>
<ul>
<li>Powerful environment for visualizing scientific data</li>
<li>Integrated graphics and statistics infrastructure</li>
<li>Publication quality graphics</li>
<li>Fully programmable</li>
<li>Highly reproducible</li>
<li>Full <a href="http://www.latex-project.org/">LaTeX</a> and Markdown support via <code>knitr</code> and <code>R markdown</code></li>
<li>Vast number of R packages with graphics utilities</li>
</ul>
<h3 id="documentation-for-r-graphics">Documentation for R Graphics</h3>
<p><strong>General</strong></p>
<ul>
<li>Graphics Task Page - <a href="http://cran.r-project.org/web/views/Graphics.html">URL</a></li>
<li>R Graph Gallery - <a href="http://addictedtor.free.fr/graphiques/allgraph.php">URL</a></li>
<li>R Graphical Manual - <a href="http://cged.genes.nig.ac.jp/RGM2/index.php">URL</a></li>
<li>Paul Murrell’s book R (Grid) Graphics - <a href="http://www.stat.auckland.ac.nz/~paul/RGraphics/rgraphics.html">URL</a></li>
</ul>
<p><strong>Interactive graphics</strong></p>
<ul>
<li>rggobi` (GGobi) - <a href="http://www.ggobi.org/">URL</a></li>
<li><code>iplots</code> - <a href="http://www.rosuda.org/iplots/">URL</a></li>
<li>Open GL (<code>rgl</code>) - <a href="http://rgl.neoscientists.org/gallery.shtml">URL</a></li>
</ul>
<h3 id="graphics-environments">Graphics Environments</h3>
<p><strong>Viewing and saving graphics in R</strong></p>
<ul>
<li>On-screen graphics</li>
<li>postscript, pdf, svg</li>
<li>jpeg, png, wmf, tiff, …</li>
</ul>
<p><strong>Four major graphic environments</strong></p>
<ol>
<li>Low-level infrastructure</li>
</ol>
<ul>
<li>R Base Graphics (low- and high-level)</li>
<li><code>grid</code>: <a href="http://www.stat.auckland.ac.nz/~paul/grid/grid.html">Manual</a></li>
</ul>
<ol start="2">
<li>High-level infrastructure
\begin{itemize}</li>
</ol>
<ul>
<li><code>lattice</code>: <a href="http://lmdvr.r-forge.r-project.org">Manual</a>, <a href="http://www.his.sunderland.ac.uk/~cs0her/Statistics/UsingLatticeGraphicsInR.htm">Intro</a>, <a href="http://www.amazon.com/Lattice-Multivariate-Data-Visualization-Use/dp/0387759689">Book</a></li>
<li><code>ggplot2</code>: <a href="http://had.co.nz/ggplot2/">Manual</a>, <a href="http://www.ling.upenn.edu/~joseff/rstudy/summer2010_ggplot2_intro.html">Intro</a>, <a href="http://had.co.nz/ggplot2/book/">Book</a></li>
</ul>
<h3 id="base-graphics-overview">Base Graphics: Overview</h3>
<p><strong>Important high-level plotting functions</strong></p>
<ul>
<li><code>plot</code>: generic x-y plotting</li>
<li><code>barplot</code>: bar plots</li>
<li><code>boxplot</code>: box-and-whisker plot</li>
<li><code>hist</code>: histograms</li>
<li><code>pie</code>: pie charts</li>
<li><code>dotchart</code>: cleveland dot plots</li>
<li><code>image, heatmap, contour, persp</code>: functions to generate image-like plots</li>
<li><code>qqnorm, qqline, qqplot</code>: distribution comparison plots</li>
<li><code>pairs, coplot</code>: display of multivariant data</li>
</ul>
<p><strong>Help on graphics functions</strong></p>
<ul>
<li><code>?myfct</code></li>
<li><code>?plot</code></li>
<li><code>?par</code></li>
</ul>
<h4 id="preferred-object-types">Preferred Object Types</h4>
<ul>
<li>Matrices and data frames</li>
<li>Vectors</li>
<li>Named vectors</li>
</ul>
<h3 id="scatter-plots">Scatter Plots</h3>
<h3 id="basic-scatter-plot">Basic Scatter Plot</h3>
<p>Sample data set for subsequent plots</p>
<pre><code class="language-r">set.seed(1410)
y &lt;- matrix(runif(30), ncol=3, dimnames=list(letters[1:10], LETTERS[1:3]))
</code></pre>
<p>Plot data</p>
<pre><code class="language-r">plot(y[,1], y[,2]) 
</code></pre>
<p><img src="../../../en/manuals/rbasics/Rbasics_files/figure-html/basic_scatter_plot-1.png" width="672" /></p>
<h4 id="all-pairs">All pairs</h4>
<pre><code class="language-r">pairs(y) 
</code></pre>
<p><img src="../../../en/manuals/rbasics/Rbasics_files/figure-html/pairs_scatter_plot-1.png" width="672" /></p>
<h4 id="with-labels">With labels</h4>
<pre><code class="language-r">plot(y[,1], y[,2], pch=20, col=&quot;red&quot;, main=&quot;Symbols and Labels&quot;)
text(y[,1]+0.03, y[,2], rownames(y))
</code></pre>
<p><img src="../../../en/manuals/rbasics/Rbasics_files/figure-html/labels_scatter_plot-1.png" width="672" /></p>
<h3 id="more-examples">More examples</h3>
<p><strong>Print instead of symbols the row names</strong></p>
<pre><code class="language-r">plot(y[,1], y[,2], type=&quot;n&quot;, main=&quot;Plot of Labels&quot;)
text(y[,1], y[,2], rownames(y)) 
</code></pre>
<p><img src="../../../en/manuals/rbasics/Rbasics_files/figure-html/row_scatter_plot-1.png" width="672" /></p>
<p><strong>Usage of important plotting parameters</strong></p>
<pre><code class="language-r">grid(5, 5, lwd = 2) 
op &lt;- par(mar=c(8,8,8,8), bg=&quot;lightblue&quot;)
plot(y[,1], y[,2], type=&quot;p&quot;, col=&quot;red&quot;, cex.lab=1.2, cex.axis=1.2, 
     cex.main=1.2, cex.sub=1, lwd=4, pch=20, xlab=&quot;x label&quot;, 
     ylab=&quot;y label&quot;, main=&quot;My Main&quot;, sub=&quot;My Sub&quot;)
par(op)
</code></pre>
<p>__Important arguments_</p>
<ul>
<li><code>mar</code>: specifies the margin sizes around the plotting area in order: <code>c(bottom, left, top, right)</code></li>
<li><code>col</code>: color of symbols</li>
<li><code>pch</code>: type of symbols, samples: <code>example(points)</code></li>
<li><code>lwd</code>: size of symbols</li>
<li><code>cex.*</code>: control font sizes</li>
<li>For details see <code>?par</code></li>
</ul>
<h4 id="add-regression-line">Add regression line</h4>
<pre><code class="language-r">plot(y[,1], y[,2])
myline &lt;- lm(y[,2]~y[,1]); abline(myline, lwd=2) 
</code></pre>
<p><img src="../../../en/manuals/rbasics/Rbasics_files/figure-html/plot_regression-1.png" width="672" /></p>
<pre><code class="language-r">summary(myline) 
</code></pre>
<pre><code>## 
## Call:
## lm(formula = y[, 2] ~ y[, 1])
## 
## Residuals:
##      Min       1Q   Median       3Q      Max 
## -0.40357 -0.17912 -0.04299  0.22147  0.46623 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)  
## (Intercept)   0.5764     0.2110   2.732   0.0258 *
## y[, 1]       -0.3647     0.3959  -0.921   0.3839  
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## Residual standard error: 0.3095 on 8 degrees of freedom
## Multiple R-squared:  0.09589,    Adjusted R-squared:  -0.01712 
## F-statistic: 0.8485 on 1 and 8 DF,  p-value: 0.3839
</code></pre>
<h4 id="log-scale">Log scale</h4>
<p>Same plot as above, but on log scale</p>
<pre><code class="language-r">plot(y[,1], y[,2], log=&quot;xy&quot;) 
</code></pre>
<p><img src="../../../en/manuals/rbasics/Rbasics_files/figure-html/plot_regression_log-1.png" width="672" /></p>
<h4 id="add-a-mathematical-expression">Add a mathematical expression</h4>
<pre><code class="language-r">plot(y[,1], y[,2]); text(y[1,1], y[1,2], expression(sum(frac(1,sqrt(x^2*pi)))), cex=1.3) 
</code></pre>
<p><img src="../../../en/manuals/rbasics/Rbasics_files/figure-html/plot_regression_math-1.png" width="672" /></p>
<h3 id="homework-3b">Homework 3B</h3>
<p>Homework 3B: <a href="http://girke.bioinformatics.ucr.edu/GEN242/mydoc_homework_03.html">Scatter Plots</a></p>
<h3 id="line-plots">Line Plots</h3>
<h4 id="single-data-set">Single data set</h4>
<pre><code class="language-r">plot(y[,1], type=&quot;l&quot;, lwd=2, col=&quot;blue&quot;) 
</code></pre>
<p><img src="../../../en/manuals/rbasics/Rbasics_files/figure-html/plot_line_single-1.png" width="672" /></p>
<h4 id="many-data-sets">Many Data Sets</h4>
<p>Plots line graph for all columns in data frame <code>y</code>. The <code>split.screen</code> function is used in this example in a for loop to overlay several line graphs in the same plot.</p>
<pre><code class="language-r">split.screen(c(1,1)) 
</code></pre>
<pre><code>## [1] 1
</code></pre>
<pre><code class="language-r">plot(y[,1], ylim=c(0,1), xlab=&quot;Measurement&quot;, ylab=&quot;Intensity&quot;, type=&quot;l&quot;, lwd=2, col=1)
for(i in 2:length(y[1,])) { 
    screen(1, new=FALSE)
    plot(y[,i], ylim=c(0,1), type=&quot;l&quot;, lwd=2, col=i, xaxt=&quot;n&quot;, yaxt=&quot;n&quot;, ylab=&quot;&quot;, xlab=&quot;&quot;, main=&quot;&quot;, bty=&quot;n&quot;) 
}
</code></pre>
<p><img src="../../../en/manuals/rbasics/Rbasics_files/figure-html/plot_line_many-1.png" width="672" /></p>
<pre><code class="language-r">close.screen(all=TRUE) 
</code></pre>
<h3 id="bar-plots">Bar Plots</h3>
<h4 id="basics">Basics</h4>
<pre><code class="language-r">barplot(y[1:4,], ylim=c(0, max(y[1:4,])+0.3), beside=TRUE, legend=letters[1:4]) 
text(labels=round(as.vector(as.matrix(y[1:4,])),2), x=seq(1.5, 13, by=1) + sort(rep(c(0,1,2), 4)), y=as.vector(as.matrix(y[1:4,]))+0.04) 
</code></pre>
<p><img src="../../../en/manuals/rbasics/Rbasics_files/figure-html/plot_bar_simple-1.png" width="672" /></p>
<h4 id="error-bars">Error Bars</h4>
<pre><code class="language-r">bar &lt;- barplot(m &lt;- rowMeans(y) * 10, ylim=c(0, 10))
stdev &lt;- sd(t(y))
arrows(bar, m, bar, m + stdev, length=0.15, angle = 90)
</code></pre>
<p><img src="../../../en/manuals/rbasics/Rbasics_files/figure-html/plot_bar_error-1.png" width="672" /></p>
<h3 id="histograms">Histograms</h3>
<pre><code class="language-r">hist(y, freq=TRUE, breaks=10)
</code></pre>
<p><img src="../../../en/manuals/rbasics/Rbasics_files/figure-html/plot_hist-1.png" width="672" /></p>
<h3 id="density-plots">Density Plots</h3>
<pre><code class="language-r">plot(density(y), col=&quot;red&quot;)
</code></pre>
<p><img src="../../../en/manuals/rbasics/Rbasics_files/figure-html/plot_dens-1.png" width="672" /></p>
<h3 id="pie-charts">Pie Charts</h3>
<pre><code class="language-r">pie(y[,1], col=rainbow(length(y[,1]), start=0.1, end=0.8), clockwise=TRUE)
legend(&quot;topright&quot;, legend=row.names(y), cex=1.3, bty=&quot;n&quot;, pch=15, pt.cex=1.8, 
col=rainbow(length(y[,1]), start=0.1, end=0.8), ncol=1) 
</code></pre>
<p><img src="../../../en/manuals/rbasics/Rbasics_files/figure-html/plot_pie-1.png" width="672" /></p>
<h3 id="color-selection-utilities">Color Selection Utilities</h3>
<p>Default color palette and how to change it</p>
<pre><code class="language-r">palette()
</code></pre>
<pre><code>## [1] &quot;black&quot;   &quot;#DF536B&quot; &quot;#61D04F&quot; &quot;#2297E6&quot; &quot;#28E2E5&quot; &quot;#CD0BBC&quot; &quot;#F5C710&quot; &quot;gray62&quot;
</code></pre>
<pre><code class="language-r">palette(rainbow(5, start=0.1, end=0.2))
palette()
</code></pre>
<pre><code>## [1] &quot;#FF9900&quot; &quot;#FFBF00&quot; &quot;#FFE600&quot; &quot;#F2FF00&quot; &quot;#CCFF00&quot;
</code></pre>
<pre><code class="language-r">palette(&quot;default&quot;)
</code></pre>
<p>The <code>gray</code> function allows to select any type of gray shades by providing values from 0 to 1</p>
<pre><code class="language-r">gray(seq(0.1, 1, by= 0.2))
</code></pre>
<pre><code>## [1] &quot;#1A1A1A&quot; &quot;#4D4D4D&quot; &quot;#808080&quot; &quot;#B3B3B3&quot; &quot;#E6E6E6&quot;
</code></pre>
<p>Color gradients with <code>colorpanel</code> function from <code>gplots</code> library`</p>
<pre><code class="language-r">library(gplots)
colorpanel(5, &quot;darkblue&quot;, &quot;yellow&quot;, &quot;white&quot;)
</code></pre>
<pre><code>## [1] &quot;#00008B&quot; &quot;#808046&quot; &quot;#FFFF00&quot; &quot;#FFFF80&quot; &quot;#FFFFFF&quot;
</code></pre>
<p>Much more on colors in R see Earl Glynn’s color chart <a href="http://research.stowers-institute.org/efg/R/Color/Chart/">here</a></p>
<h3 id="saving-graphics-to-file">Saving Graphics to File</h3>
<p>After the <code>pdf()</code> command all graphs are redirected to file <code>test.pdf</code>. Works for all common formats similarly: jpeg, png, ps, tiff, …</p>
<pre><code class="language-r">pdf(&quot;test.pdf&quot;)
plot(1:10, 1:10)
dev.off() 
</code></pre>
<p>Generates Scalable Vector Graphics (SVG) files that can be edited in vector graphics programs, such as InkScape.</p>
<pre><code class="language-r">library(&quot;RSvgDevice&quot;)
devSVG(&quot;test.svg&quot;)
plot(1:10, 1:10)
dev.off() 
</code></pre>
<h3 id="homework-3c">Homework 3C</h3>
<p>Homework 3C: <a href="https://girke.bioinformatics.ucr.edu/GEN242/assignments/homework/hw3/">Bar Plots</a></p>
<h2 id="analysis-routine">Analysis Routine</h2>
<h3 id="overview-1">Overview</h3>
<p>The following exercise introduces a variety of useful data analysis utilities in R.</p>
<h3 id="analysis-routine-data-import">Analysis Routine: Data Import</h3>
<ul>
<li>
<p><strong>Step 1</strong>: To get started with this exercise, direct your R session to a dedicated workshop directory and download into this directory the following sample tables. Then import the files into Excel and save them as tab delimited text files.</p>
<ul>
<li><a href="http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/Samples/MolecularWeight_tair7.xls">MolecularWeight_tair7.xls</a></li>
<li><a href="http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/Samples/TargetP_analysis_tair7.xls">TargetP_analysis_tair7.xls</a></li>
</ul>
</li>
</ul>
<p><strong>Import the tables into R</strong></p>
<p>Import molecular weight table</p>
<pre><code class="language-r">my_mw &lt;- read.delim(file=&quot;MolecularWeight_tair7.xls&quot;, header=T, sep=&quot;\t&quot;) 
my_mw[1:2,]
</code></pre>
<p>Import subcelluar targeting table</p>
<pre><code class="language-r">my_target &lt;- read.delim(file=&quot;TargetP_analysis_tair7.xls&quot;, header=T, sep=&quot;\t&quot;) 
my_target[1:2,]
</code></pre>
<p>Online import of molecular weight table</p>
<pre><code class="language-r">my_mw &lt;- read.delim(file=&quot;http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/Samples/MolecularWeight_tair7.xls&quot;, header=T, sep=&quot;\t&quot;) 
my_mw[1:2,]
</code></pre>
<pre><code>##   Sequence.id Molecular.Weight.Da. Residues
## 1 AT1G08520.1                83285      760
## 2 AT1G08530.1                27015      257
</code></pre>
<p>Online import of subcelluar targeting table</p>
<pre><code class="language-r">my_target &lt;- read.delim(file=&quot;http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/Samples/TargetP_analysis_tair7.xls&quot;, header=T, sep=&quot;\t&quot;) 
my_target[1:2,]
</code></pre>
<pre><code>##      GeneName Loc   cTP   mTP    SP other
## 1 AT1G08520.1   C 0.822 0.137 0.029 0.039
## 2 AT1G08530.1   C 0.817 0.058 0.010 0.100
</code></pre>
<h3 id="merging-data-frames">Merging Data Frames</h3>
<ul>
<li><strong>Step 2</strong>: Assign uniform gene ID column titles</li>
</ul>
<pre><code class="language-r">colnames(my_target)[1] &lt;- &quot;ID&quot;
colnames(my_mw)[1] &lt;- &quot;ID&quot; 
</code></pre>
<ul>
<li><strong>Step 3</strong>: Merge the two tables based on common ID field</li>
</ul>
<pre><code class="language-r">my_mw_target &lt;- merge(my_mw, my_target, by.x=&quot;ID&quot;, by.y=&quot;ID&quot;, all.x=T)
</code></pre>
<ul>
<li><strong>Step 4</strong>: Shorten one table before the merge and then remove the non-matching rows (NAs) in the merged file</li>
</ul>
<pre><code class="language-r">my_mw_target2a &lt;- merge(my_mw, my_target[1:40,], by.x=&quot;ID&quot;, by.y=&quot;ID&quot;, all.x=T)  # To remove non-matching rows, use the argument setting 'all=F'.
my_mw_target2 &lt;- na.omit(my_mw_target2a) # Removes rows containing &quot;NAs&quot; (non-matching rows).
</code></pre>
<ul>
<li><strong>Homework 3D</strong>: How can the merge function in the previous step be executed so that only the common rows among the two data frames are returned? Prove that both methods - the two step version with <code>na.omit</code> and your method - return identical results.</li>
<li><strong>Homework 3E</strong>: Replace all <code>NAs</code> in the data frame <code>my_mw_target2a</code> with zeros.</li>
</ul>
<h3 id="filtering-data">Filtering Data</h3>
<ul>
<li><strong>Step 5</strong>: Retrieve all records with a value of greater than 100,000 in ‘MW’ column and ‘C’ value in ‘Loc’ column (targeted to chloroplast).</li>
</ul>
<pre><code class="language-r">query &lt;- my_mw_target[my_mw_target[, 2] &gt; 100000 &amp; my_mw_target[, 4] == &quot;C&quot;, ] 
query[1:4, ]
</code></pre>
<pre><code>##              ID Molecular.Weight.Da. Residues Loc   cTP   mTP    SP other
## 219 AT1G02730.1               132588     1181   C 0.972 0.038 0.008 0.045
## 243 AT1G02890.1               136825     1252   C 0.748 0.529 0.011 0.013
## 281 AT1G03160.1               100732      912   C 0.871 0.235 0.011 0.007
## 547 AT1G05380.1               126360     1138   C 0.740 0.099 0.016 0.358
</code></pre>
<pre><code class="language-r">dim(query)
</code></pre>
<pre><code>## [1] 170   8
</code></pre>
<ul>
<li><strong>Homework 3F</strong>: How many protein entries in the <code>my</code>_mw<code>_target</code> data frame have a MW of greater then 4,000 and less then 5,000. Subset the data frame accordingly and sort it by MW to check that your result is correct.</li>
</ul>
<h3 id="string-substitutions">String Substitutions</h3>
<ul>
<li><strong>Step 6</strong>: Use a regular expression in a substitute function to generate a separate ID column that lacks the gene model extensions.
&lt;&lt;label=Exercise 4.7, eval=TRUE, echo=TRUE, keep.source=TRUE&gt;&gt;=</li>
</ul>
<pre><code class="language-r">my_mw_target3 &lt;- data.frame(loci=gsub(&quot;\\..*&quot;, &quot;&quot;, as.character(my_mw_target[,1]), perl = TRUE), my_mw_target)
my_mw_target3[1:3,1:8]
</code></pre>
<pre><code>##        loci          ID Molecular.Weight.Da. Residues Loc  cTP   mTP    SP
## 1 AT1G01010 AT1G01010.1                49426      429   _ 0.10 0.090 0.075
## 2 AT1G01020 AT1G01020.1                28092      245   * 0.01 0.636 0.158
## 3 AT1G01020 AT1G01020.2                21711      191   * 0.01 0.636 0.158
</code></pre>
<ul>
<li><strong>Homework 3G</strong>: Retrieve those rows in <code>my_mw_target3</code> where the second column contains the following identifiers: <code>c(&quot;AT5G52930.1&quot;, &quot;AT4G18950.1&quot;, &quot;AT1G15385.1&quot;, &quot;AT4G36500.1&quot;, &quot;AT1G67530.1&quot;)</code>. Use the <code>%in%</code> function for this query. As an alternative approach, assign the second column to the row index of the data frame and then perform the same query again using the row index. Explain the difference of the two methods.</li>
</ul>
<h3 id="calculations-on-data-frames">Calculations on Data Frames</h3>
<ul>
<li><strong>Step 7</strong>: Count the number of duplicates in the loci column with the <code>table</code> function and append the result to the data frame with the <code>cbind</code> function.</li>
</ul>
<pre><code class="language-r">mycounts &lt;- table(my_mw_target3[,1])[my_mw_target3[,1]]
my_mw_target4 &lt;- cbind(my_mw_target3, Freq=mycounts[as.character(my_mw_target3[,1])]) 
</code></pre>
<ul>
<li><strong>Step 8</strong>: Perform a vectorized devision of columns 3 and 4 (average AA weight per protein)</li>
</ul>
<pre><code class="language-r">data.frame(my_mw_target4, avg_AA_WT=(my_mw_target4[,3] / my_mw_target4[,4]))[1:2,5:11] 
</code></pre>
<pre><code>##   Loc  cTP   mTP    SP other Freq.Var1 Freq.Freq
## 1   _ 0.10 0.090 0.075 0.925 AT1G01010         1
## 2   * 0.01 0.636 0.158 0.448 AT1G01020         2
</code></pre>
<ul>
<li><strong>Step 9</strong>: Calculate for each row the mean and standard deviation across several columns</li>
</ul>
<pre><code class="language-r">mymean &lt;- apply(my_mw_target4[,6:9], 1, mean)
mystdev &lt;- apply(my_mw_target4[,6:9], 1, sd, na.rm=TRUE)
data.frame(my_mw_target4, mean=mymean, stdev=mystdev)[1:2,5:12] 
</code></pre>
<pre><code>##   Loc  cTP   mTP    SP other Freq.Var1 Freq.Freq   mean
## 1   _ 0.10 0.090 0.075 0.925 AT1G01010         1 0.2975
## 2   * 0.01 0.636 0.158 0.448 AT1G01020         2 0.3130
</code></pre>
<h3 id="plotting-example">Plotting Example</h3>
<ul>
<li><strong>Step 10</strong>: Generate scatter plot columns: ‘MW’ and ‘Residues’</li>
</ul>
<pre><code class="language-r">plot(my_mw_target4[1:500,3:4], col=&quot;red&quot;)
</code></pre>
<p><img src="../../../en/manuals/rbasics/Rbasics_files/figure-html/plot_example-1.png" width="672" /></p>
<h3 id="export-results-and-run-entire-exercise-as-script">Export Results and Run Entire Exercise as Script</h3>
<ul>
<li><strong>Step 11</strong>: Write the data frame <code>my_mw_target4</code> into a tab-delimited text file and inspect it in Excel.</li>
</ul>
<pre><code class="language-r">write.table(my_mw_target4, file=&quot;my_file.xls&quot;, quote=F, sep=&quot;\t&quot;, col.names = NA) 
</code></pre>
<ul>
<li><strong>Homework 3H</strong>: Write all commands from this exercise into an R script named <code>exerciseRbasics.R</code>, or download it from <a href="http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/My_R_Scripts/exerciseRbasics.R">here</a>. Then execute the script with the <code>source</code> function like this: <code>source(&quot;exerciseRbasics.R&quot;)</code>. This will run all commands of this exercise and generate the corresponding output files in the current working directory.</li>
</ul>
<pre><code class="language-r">source(&quot;exerciseRbasics.R&quot;)
</code></pre>
<h2 id="r-markdown">R Markdown</h2>
<h3 id="overview-2">Overview</h3>
<p>R Markdown combines markdown (an easy to write plain text format) with embedded
R code chunks. When compiling R Markdown documents, the code components can be
evaluated so that both the code and its output can be included in the final
document. This makes analysis reports highly reproducible by allowing to automatically
regenerate them when the underlying R code or data changes. R Markdown
documents (<code>.Rmd</code> files) can be rendered to various formats including HTML and
PDF. The R code in an <code>.Rmd</code> document is processed by <code>knitr</code>, while the
resulting <code>.md</code> file is rendered by <code>pandoc</code> to the final output formats
(<em>e.g.</em> HTML or PDF). Historically, R Markdown is an extension of the older
<code>Sweave/Latex</code> environment. Rendering of mathematical expressions and reference
management is also supported by R Markdown using embedded Latex syntax and
Bibtex, respectively.</p>
<h3 id="quick-start">Quick Start</h3>
<h4 id="install-r-markdown">Install R Markdown</h4>
<pre><code class="language-r">install.packages(&quot;rmarkdown&quot;)
</code></pre>
<h4 id="initialize-a-new-r-markdown-rmd-script">Initialize a new R Markdown (<code>Rmd</code>) script</h4>
<p>To minimize typing, it can be helful to start with an R Markdown template and
then modify it as needed. Note the file name of an R Markdown scirpt needs to
have the extension <code>.Rmd</code>. Template files for the following examples are available
here:</p>
<ul>
<li>R Markdown sample script: <a href="https://raw.githubusercontent.com/tgirke/GEN242/gh-pages/_vignettes/07_Rbasics/sample.Rmd"><code>sample.Rmd</code></a></li>
<li>Bibtex file for handling citations and reference section: <a href="https://raw.githubusercontent.com/tgirke/GEN242/gh-pages/_vignettes/07_Rbasics/bibtex.bib"><code>bibtex.bib</code></a></li>
</ul>
<p>Users want to download these files, open the <code>sample.Rmd</code> file with their preferred R IDE
(<em>e.g.</em> RStudio, vim or emacs), initilize an R session and then direct their R session to
the location of these two files.</p>
<h4 id="metadata-section">Metadata section</h4>
<p>The metadata section (YAML header) in an R Markdown script defines how it will be processed and
rendered. The metadata section also includes both title, author, and date information as well as
options for customizing the output format. For instance, PDF and HTML output can be defined
with <code>pdf_document</code> and <code>html_document</code>, respectively. The <code>BiocStyle::</code> prefix will use the
formatting style of the <a href="http://bioconductor.org/packages/release/bioc/html/BiocStyle.html"><code>BiocStyle</code></a>
package from Bioconductor.</p>
<pre><code> ---
title: &quot;My First R Markdown Document&quot;
author: &quot;Author: First Last&quot;
date: &quot;Last update: 15 May, 2021&quot;
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 3
    fig_caption: yes

fontsize: 14pt
bibliography: bibtex.bib
 ---
</code></pre>
<h4 id="render-rmd-script">Render <code>Rmd</code> script</h4>
<p>An R Markdown script can be evaluated and rendered with the following <code>render</code> command or by pressing the <code>knit</code> button in RStudio.
The <code>output_format</code> argument defines the format of the output (<em>e.g.</em> <code>html_document</code>). The setting <code>output_format=&quot;all&quot;</code> will generate
all supported output formats. Alternatively, one can specify several output formats in the metadata section as shown in the above example.</p>
<pre><code class="language-r">rmarkdown::render(&quot;sample.Rmd&quot;, clean=TRUE, output_format=&quot;html_document&quot;)
</code></pre>
<p>The following shows two options how to run the rendering from the command-line.</p>
<pre><code class="language-sh">$ Rscript -e &quot;rmarkdown::render('sample.Rmd', clean=TRUE)&quot;
</code></pre>
<p>Alternatively, one can use a Makefile to evaluate and render an R Markdown
script. A sample Makefile for rendering the above <code>sample.Rmd</code> can be
downloaded <a href="https://raw.githubusercontent.com/tgirke/GEN242/gh-pages/_vignettes/07_Rbasics/Makefile"><code>here</code></a>.
To apply it to a custom <code>Rmd</code> file, one needs open the Makefile in a text
editor and change the value assigned to <code>MAIN</code> (line 13) to the base name of
the corresponding <code>.Rmd</code> file (<em>e.g.</em> assign <code>systemPipeRNAseq</code> if the file
name is <code>systemPipeRNAseq.Rmd</code>). To execute the <code>Makefile</code>, run the following
command from the command-line.</p>
<pre><code class="language-sh">$ make -B
</code></pre>
<h4 id="r-code-chunks">R code chunks</h4>
<p>R Code Chunks can be embedded in an R Markdown script by using three backticks
at the beginning of a new line along with arguments enclosed in curly braces
controlling the behavior of the code. The following lines contain the
plain R code. A code chunk is terminated by a new line starting with three backticks.
The following shows an example of such a code chunk. Note the backslashes are
not part of it. They have been added to print the code chunk syntax in this document.</p>
<pre><code>    ```\{r code_chunk_name, eval=FALSE\}
    x &lt;- 1:10
    ```
</code></pre>
<p>The following lists the most important arguments to control the behavior of R code chunks:</p>
<ul>
<li><code>r</code>: specifies language for code chunk, here R</li>
<li><code>chode_chunk_name</code>: name of code chunk; this name needs to be unique</li>
<li><code>eval</code>: if assigned <code>TRUE</code> the code will be evaluated</li>
<li><code>warning</code>: if assigned <code>FALSE</code> warnings will not be shown</li>
<li><code>message</code>: if assigned <code>FALSE</code> messages will not be shown</li>
<li><code>cache</code>: if assigned <code>TRUE</code> results will be cached to reuse in future rendering instances</li>
<li><code>fig.height</code>: allows to specify height of figures in inches</li>
<li><code>fig.width</code>: allows to specify width of figures in inches</li>
</ul>
<p>For more details on code chunk options see <a href="https://www.rstudio.com/wp-content/uploads/2015/03/rmarkdown-reference.pdf">here</a>.</p>
<h4 id="learning-markdown">Learning Markdown</h4>
<p>The basic syntax of Markdown and derivatives like kramdown is extremely easy to learn. Rather
than providing another introduction on this topic, here are some useful sites for learning Markdown:</p>
<ul>
<li><a href="https://guides.github.com/features/mastering-markdown/">Markdown Intro on GitHub</a></li>
<li><a href="https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet">Markdown Cheet Sheet</a></li>
<li><a href="http://rmarkdown.rstudio.com/authoring_basics.html">Markdown Basics from RStudio</a></li>
<li><a href="http://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf">R Markdown Cheat Sheet</a></li>
<li><a href="http://kramdown.gettalong.org/syntax.html">kramdown Syntax</a></li>
</ul>
<h4 id="tables">Tables</h4>
<p>There are several ways to render tables. First, they can be printed within the R code chunks. Second,
much nicer formatted tables can be generated with the functions <code>kable</code>, <code>pander</code> or <code>xtable</code>. The following
example uses <code>kable</code> from the <code>knitr</code> package.</p>
<pre><code class="language-r">library(knitr)
kable(iris[1:12,])
</code></pre>
<table>
<thead>
<tr>
<th style="text-align:right">Sepal.Length</th>
<th style="text-align:right">Sepal.Width</th>
<th style="text-align:right">Petal.Length</th>
<th style="text-align:right">Petal.Width</th>
<th style="text-align:left">Species</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right">5.1</td>
<td style="text-align:right">3.5</td>
<td style="text-align:right">1.4</td>
<td style="text-align:right">0.2</td>
<td style="text-align:left">setosa</td>
</tr>
<tr>
<td style="text-align:right">4.9</td>
<td style="text-align:right">3.0</td>
<td style="text-align:right">1.4</td>
<td style="text-align:right">0.2</td>
<td style="text-align:left">setosa</td>
</tr>
<tr>
<td style="text-align:right">4.7</td>
<td style="text-align:right">3.2</td>
<td style="text-align:right">1.3</td>
<td style="text-align:right">0.2</td>
<td style="text-align:left">setosa</td>
</tr>
<tr>
<td style="text-align:right">4.6</td>
<td style="text-align:right">3.1</td>
<td style="text-align:right">1.5</td>
<td style="text-align:right">0.2</td>
<td style="text-align:left">setosa</td>
</tr>
<tr>
<td style="text-align:right">5.0</td>
<td style="text-align:right">3.6</td>
<td style="text-align:right">1.4</td>
<td style="text-align:right">0.2</td>
<td style="text-align:left">setosa</td>
</tr>
<tr>
<td style="text-align:right">5.4</td>
<td style="text-align:right">3.9</td>
<td style="text-align:right">1.7</td>
<td style="text-align:right">0.4</td>
<td style="text-align:left">setosa</td>
</tr>
<tr>
<td style="text-align:right">4.6</td>
<td style="text-align:right">3.4</td>
<td style="text-align:right">1.4</td>
<td style="text-align:right">0.3</td>
<td style="text-align:left">setosa</td>
</tr>
<tr>
<td style="text-align:right">5.0</td>
<td style="text-align:right">3.4</td>
<td style="text-align:right">1.5</td>
<td style="text-align:right">0.2</td>
<td style="text-align:left">setosa</td>
</tr>
<tr>
<td style="text-align:right">4.4</td>
<td style="text-align:right">2.9</td>
<td style="text-align:right">1.4</td>
<td style="text-align:right">0.2</td>
<td style="text-align:left">setosa</td>
</tr>
<tr>
<td style="text-align:right">4.9</td>
<td style="text-align:right">3.1</td>
<td style="text-align:right">1.5</td>
<td style="text-align:right">0.1</td>
<td style="text-align:left">setosa</td>
</tr>
<tr>
<td style="text-align:right">5.4</td>
<td style="text-align:right">3.7</td>
<td style="text-align:right">1.5</td>
<td style="text-align:right">0.2</td>
<td style="text-align:left">setosa</td>
</tr>
<tr>
<td style="text-align:right">4.8</td>
<td style="text-align:right">3.4</td>
<td style="text-align:right">1.6</td>
<td style="text-align:right">0.2</td>
<td style="text-align:left">setosa</td>
</tr>
</tbody>
</table>
<p>A much more elegant and powerful solution is to create fully interactive tables
with the <a href="https://rstudio.github.io/DT/"><code>DT</code> package</a>. This JavaScirpt based environment provides a wrapper to the DataTables library
using jQuery. The resulting tables can be sorted, queried and resized by the
user. For an example see <a href="http://girke.bioinformatics.ucr.edu/GEN242/pages/mydoc/Rbasics.html#1627_tables">here</a>.</p>
<pre><code class="language-r">library(DT)
datatable(iris, filter = 'top', options = list(
      pageLength = 100, scrollX = TRUE, scrollY = &quot;600px&quot;, autoWidth = TRUE
     ))
</code></pre>
<div id="htmlwidget-1" class="datatables html-widget" style="width:100%;height:auto;">
</div>
<script type="application/json" data-for="htmlwidget-1">{"x":{"filter":"top","filterHTML":"<tr>\n  <td><\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"4.3\" data-max=\"7.9\" data-scale=\"1\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"2\" data-max=\"4.4\" data-scale=\"1\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"1\" data-max=\"6.9\" data-scale=\"1\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"0.1\" data-max=\"2.5\" data-scale=\"1\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"factor\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"width: 100%; display: none;\">\n      <select multiple=\"multiple\" style=\"width: 100%;\" data-options=\"[&quot;setosa&quot;,&quot;versicolor&quot;,&quot;virginica&quot;]\"><\/select>\n    <\/div>\n  <\/td>\n<\/tr>","data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150"],[5.1,4.9,4.7,4.6,5,5.4,4.6,5,4.4,4.9,5.4,4.8,4.8,4.3,5.8,5.7,5.4,5.1,5.7,5.1,5.4,5.1,4.6,5.1,4.8,5,5,5.2,5.2,4.7,4.8,5.4,5.2,5.5,4.9,5,5.5,4.9,4.4,5.1,5,4.5,4.4,5,5.1,4.8,5.1,4.6,5.3,5,7,6.4,6.9,5.5,6.5,5.7,6.3,4.9,6.6,5.2,5,5.9,6,6.1,5.6,6.7,5.6,5.8,6.2,5.6,5.9,6.1,6.3,6.1,6.4,6.6,6.8,6.7,6,5.7,5.5,5.5,5.8,6,5.4,6,6.7,6.3,5.6,5.5,5.5,6.1,5.8,5,5.6,5.7,5.7,6.2,5.1,5.7,6.3,5.8,7.1,6.3,6.5,7.6,4.9,7.3,6.7,7.2,6.5,6.4,6.8,5.7,5.8,6.4,6.5,7.7,7.7,6,6.9,5.6,7.7,6.3,6.7,7.2,6.2,6.1,6.4,7.2,7.4,7.9,6.4,6.3,6.1,7.7,6.3,6.4,6,6.9,6.7,6.9,5.8,6.8,6.7,6.7,6.3,6.5,6.2,5.9],[3.5,3,3.2,3.1,3.6,3.9,3.4,3.4,2.9,3.1,3.7,3.4,3,3,4,4.4,3.9,3.5,3.8,3.8,3.4,3.7,3.6,3.3,3.4,3,3.4,3.5,3.4,3.2,3.1,3.4,4.1,4.2,3.1,3.2,3.5,3.6,3,3.4,3.5,2.3,3.2,3.5,3.8,3,3.8,3.2,3.7,3.3,3.2,3.2,3.1,2.3,2.8,2.8,3.3,2.4,2.9,2.7,2,3,2.2,2.9,2.9,3.1,3,2.7,2.2,2.5,3.2,2.8,2.5,2.8,2.9,3,2.8,3,2.9,2.6,2.4,2.4,2.7,2.7,3,3.4,3.1,2.3,3,2.5,2.6,3,2.6,2.3,2.7,3,2.9,2.9,2.5,2.8,3.3,2.7,3,2.9,3,3,2.5,2.9,2.5,3.6,3.2,2.7,3,2.5,2.8,3.2,3,3.8,2.6,2.2,3.2,2.8,2.8,2.7,3.3,3.2,2.8,3,2.8,3,2.8,3.8,2.8,2.8,2.6,3,3.4,3.1,3,3.1,3.1,3.1,2.7,3.2,3.3,3,2.5,3,3.4,3],[1.4,1.4,1.3,1.5,1.4,1.7,1.4,1.5,1.4,1.5,1.5,1.6,1.4,1.1,1.2,1.5,1.3,1.4,1.7,1.5,1.7,1.5,1,1.7,1.9,1.6,1.6,1.5,1.4,1.6,1.6,1.5,1.5,1.4,1.5,1.2,1.3,1.4,1.3,1.5,1.3,1.3,1.3,1.6,1.9,1.4,1.6,1.4,1.5,1.4,4.7,4.5,4.9,4,4.6,4.5,4.7,3.3,4.6,3.9,3.5,4.2,4,4.7,3.6,4.4,4.5,4.1,4.5,3.9,4.8,4,4.9,4.7,4.3,4.4,4.8,5,4.5,3.5,3.8,3.7,3.9,5.1,4.5,4.5,4.7,4.4,4.1,4,4.4,4.6,4,3.3,4.2,4.2,4.2,4.3,3,4.1,6,5.1,5.9,5.6,5.8,6.6,4.5,6.3,5.8,6.1,5.1,5.3,5.5,5,5.1,5.3,5.5,6.7,6.9,5,5.7,4.9,6.7,4.9,5.7,6,4.8,4.9,5.6,5.8,6.1,6.4,5.6,5.1,5.6,6.1,5.6,5.5,4.8,5.4,5.6,5.1,5.1,5.9,5.7,5.2,5,5.2,5.4,5.1],[0.2,0.2,0.2,0.2,0.2,0.4,0.3,0.2,0.2,0.1,0.2,0.2,0.1,0.1,0.2,0.4,0.4,0.3,0.3,0.3,0.2,0.4,0.2,0.5,0.2,0.2,0.4,0.2,0.2,0.2,0.2,0.4,0.1,0.2,0.2,0.2,0.2,0.1,0.2,0.2,0.3,0.3,0.2,0.6,0.4,0.3,0.2,0.2,0.2,0.2,1.4,1.5,1.5,1.3,1.5,1.3,1.6,1,1.3,1.4,1,1.5,1,1.4,1.3,1.4,1.5,1,1.5,1.1,1.8,1.3,1.5,1.2,1.3,1.4,1.4,1.7,1.5,1,1.1,1,1.2,1.6,1.5,1.6,1.5,1.3,1.3,1.3,1.2,1.4,1.2,1,1.3,1.2,1.3,1.3,1.1,1.3,2.5,1.9,2.1,1.8,2.2,2.1,1.7,1.8,1.8,2.5,2,1.9,2.1,2,2.4,2.3,1.8,2.2,2.3,1.5,2.3,2,2,1.8,2.1,1.8,1.8,1.8,2.1,1.6,1.9,2,2.2,1.5,1.4,2.3,2.4,1.8,1.8,2.1,2.4,2.3,1.9,2.3,2.5,2.3,1.9,2,2.3,1.8],["setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Sepal.Length<\/th>\n      <th>Sepal.Width<\/th>\n      <th>Petal.Length<\/th>\n      <th>Petal.Width<\/th>\n      <th>Species<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":100,"scrollX":true,"scrollY":"600px","autoWidth":true,"columnDefs":[{"className":"dt-right","targets":[1,2,3,4]},{"orderable":false,"targets":0}],"order":[],"orderClasses":false,"orderCellsTop":true}},"evals":[],"jsHooks":[]}</script>
<h4 id="figures">Figures</h4>
<p>Plots generated by the R code chunks in an R Markdown document can be automatically
inserted in the output file. The size of the figure can be controlled with the <code>fig.height</code>
and <code>fig.width</code> arguments.</p>
<pre><code class="language-r">library(ggplot2)
dsmall &lt;- diamonds[sample(nrow(diamonds), 1000), ]
ggplot(dsmall, aes(color, price/carat)) + geom_jitter(alpha = I(1 / 2), aes(color=color))
</code></pre>
<p><img src="../../../en/manuals/rbasics/Rbasics_files/figure-html/some_jitter_plot-1.png" width="672" /></p>
<p>Sometimes it can be useful to explicitly write an image to a file and then insert that
image into the final document by referencing its file name in the R Markdown source. For
instance, this can be useful for time consuming analyses. The following code will generate a
file named <code>myplot.png</code>. To insert the file in the final document, one can use standard
Markdown or HTML syntax, <em>e.g.</em>: <code>&lt;img src=&quot;myplot.png&quot;/&gt;</code>.</p>
<pre><code class="language-r">png(&quot;images/myplot.png&quot;)
ggplot(dsmall, aes(color, price/carat)) + geom_jitter(alpha = I(1 / 2), aes(color=color))
dev.off()
</code></pre>
<pre><code>## png 
##   2
</code></pre>
<center>
<img title="some_title" src="./images/myplot.png"/>
</center>
<h4 id="inline-r-code">Inline R code</h4>
<p>To evaluate R code inline, one can enclose an R expression with a single back-tick
followed by <code>r</code> and then the actual expression. For instance, the back-ticked version
of ‘r 1 + 1’ evaluates to 2 and ‘r pi’ evaluates to 3.1415927.</p>
<h4 id="mathematical-equations">Mathematical equations</h4>
<p>To render mathematical equations, one can use standard Latex syntax. When expressions are
enclosed with single <code>$</code> signs then they will be shown inline, while
enclosing them with double <code>$$</code> signs will show them in display mode. For instance, the following
Latex syntax <code>d(X,Y) = \sqrt[]{ \sum_{i=1}^{n}{(x_{i}-y_{i})^2} }</code> renders in display mode as follows:</p>
<p><code>$$d(X,Y) = \sqrt[]{ \sum_{i=1}^{n}{(x_{i}-y_{i})^2} }$$</code></p>
<h4 id="citations-and-bibliographies">Citations and bibliographies</h4>
<p>Citations and bibliographies can be autogenerated in R Markdown in a similar
way as in Latex/Bibtex. Reference collections should be stored in a separate
file in Bibtex or other supported formats. To cite a publication in an R Markdown
script, one uses the syntax <code>[@&lt;id1&gt;]</code> where <code>&lt;id1&gt;</code> needs to be replaced with a
reference identifier present in the Bibtex database listed in the metadata section
of the R Markdown script (<em>e.g.</em> <code>bibtex.bib</code>). For instance, to cite Lawrence et al. 
(2013), one uses its reference identifier (<em>e.g.</em> <code>Lawrence2013-kt</code>) as <code>&lt;id1&gt;</code> (Lawrence et al. 2013).
This will place the citation inline in the text and add the corresponding
reference to a reference list at the end of the output document. For the latter a
special section called <code>References</code> needs to be specified at the end of the R Markdown script.
To fine control the formatting of citations and reference lists, users want to consult this
the corresponding <a href="http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html">R Markdown page</a>.
Also, for general reference management and outputting references in Bibtex format <a href="https://paperpile.com/features">Paperpile</a>
can be very helpful.</p>
<h4 id="viewing-r-markdown-report-on-hpcc-cluster">Viewing R Markdown report on HPCC cluster</h4>
<p>R Markdown reports located on UCR’s HPCC Cluster can be viewed locally in a web browser (without moving
the source HTML) by creating a symbolic link from a user’s <code>.html</code> directory. This way any updates to
the report can be viewed immediately without creating another copy of the HTML file. For instance, if user
<code>ttest</code> has generated an R Markdown report under <code>~/bigdata/today/rmarkdown/sample.html</code>, then the
proper symbolic link to this file can be created as follows:</p>
<pre><code class="language-r">cd ~/.html
ln -s ~/bigdata/today/rmarkdown/sample.html sample.html
</code></pre>
<p>After this one can view the report in a web browser using this URL <a href="http://biocluster.ucr.edu/~ttest/rmarkdown/sample.html">http://biocluster.ucr.edu/~ttest/rmarkdown/sample.html</a>.
If necessary access to the URL can be restricted with a password following the instructions <a href="http://hpcc.ucr.edu/manuals_linux-cluster_sharing.html#sharing-files-on-the-web">here</a>.</p>
<p>A sample R Markdown report for an RNA-Seq project is given here:</p>
<ul>
<li><a href="https://girke.bioinformatics.ucr.edu/GEN242/pages/mydoc/demo_files/systemPipeRNAseq.html">RNASeq.html</a>{:target=&quot;_blank&quot;}</li>
<li><a href="https://raw.githubusercontent.com/tgirke/GEN242/gh-pages/pages/mydoc/demo_files/systemPipeRNAseq.Rmd">RNASeq.Rmd</a>{:target=&quot;_blank&quot;}</li>
</ul>
<h2 id="shiny-web-apps">Shiny Web Apps</h2>
<h3 id="what-is-shiny">What is Shiny?</h3>
<p><a href="https://shiny.rstudio.com/gallery/">Shiny</a> is an R-based environment for building interactive web applications for
data analysis and exploration. Since most JavaScript code is autogenerated by
the environment, basic R knowledge is sufficient for developing Shiny apps.
They can be deployed on local computers or web servers including custom and cloud-based servers (e.g.
AWS, GCP, <a href="http://www.shinyapps.io/">shinyapp.io</a> service). The basic structure of a Shiny app is an
<code>app.R</code> script containing the following components:</p>
<ol>
<li>
<p>User interface</p>
<pre><code class="language-r">ui &lt;- fluidPage()
</code></pre>
</li>
<li>
<p>Server function</p>
<pre><code class="language-r">server &lt;- function(input, output) {}
</code></pre>
</li>
<li>
<p>Statement to run shiny app</p>
<pre><code class="language-r">shinyApp(ui = ui, server = server)
</code></pre>
</li>
</ol>
<p>Alternatively, the <code>ui</code> and <code>server</code> functions can be organized in two script files, a <code>ui.R</code> and a <code>server.R</code> script, respectively.</p>
<h3 id="develop-and-test-shiny-app-locally">Develop and test Shiny app locally</h3>
<p>Open R and set session to parent directory (here <code>myappdir</code>) containing shiny script <code>app.R</code>, and the
run it with the <code>runApp()</code> function. A sample <code>app.R</code> script for testing can be downloaded from <a href="https://raw.githubusercontent.com/tgirke/GEN242/main/static/custom/scripts/app.R">here</a>.</p>
<pre><code class="language-r">library(shiny)
runApp(&quot;myappdir&quot;) # To show code in app, add argument: display.mode=&quot;showcase&quot; 
</code></pre>
<p>This will open the app in a web browser.</p>
<h3 id="deploy-on-web-server">Deploy on web server</h3>
<p>This can be done on local or cloud systems. An easy solution is to get an account on <a href="http://www.shinyapps.io/">shinyapps.io</a>
and then deploy Shiny apps there. For details, see <a href="https://shiny.rstudio.com/deploy/">here</a>.</p>
<pre><code class="language-r">setwd(&quot;myappdir&quot;)
library(rsconnect)
deployApp()
</code></pre>
<h3 id="example-shiny-app">Example Shiny app</h3>
<p>The following Shiny app is hosted on <code>shinyapps.io</code> and embedded into the markdown (or html) source of this page
using the following iframe syntax:</p>
<pre><code class="language-r">&lt;iframe src=&quot;https://tgirke.shinyapps.io/diamonds/&quot; style=&quot;border: none; width: 880px; height: 900px&quot;&gt;&lt;/iframe&gt;
</code></pre>
<iframe src="https://tgirke.shinyapps.io/diamonds/" style="border: none; width: 880px; height: 900px">
</iframe>
<h3 id="learning-shiny">Learning Shiny</h3>
<p>The Shiny section on the Rstudio site contains excellent <a href="https://shiny.rstudio.com/tutorial/">tutorials</a>.
In addition, users may want to explore the example apps included in the <code>shiny</code> package. This can be
done by loading the individual examples (see <a href="https://shiny.rstudio.com/tutorial/written-tutorial/lesson1/">here</a>) or saving
the code to a user writable directory like this:</p>
<pre><code class="language-r">mydir &lt;- system.file(&quot;examples&quot;, package=&quot;shiny&quot;)
dir.create('my_shiny_test_dir')
file.copy(mydir, &quot;my_shiny_test_dir&quot;, recursive=TRUE)
setwd(&quot;my_shiny_test_dir/examples&quot;)
runApp(&quot;01_hello&quot;) # Runs first example app in directory 
dir() # Lists available Shiny examples (directories). 
</code></pre>
<h2 id="session-info">Session Info</h2>
<pre><code class="language-r">sessionInfo()
</code></pre>
<pre><code>## R version 4.0.4 (2021-02-15)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 10 (buster)
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.8.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.8.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
##  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] DT_0.16           knitr_1.30        gplots_3.1.1      RSQLite_2.2.1     data.table_1.13.2
##  [6] forcats_0.5.0     stringr_1.4.0     dplyr_1.0.2       purrr_0.3.4       readr_1.4.0      
## [11] tidyr_1.1.2       tibble_3.0.4      tidyverse_1.3.0   ggplot2_3.3.2     limma_3.46.0     
## [16] BiocStyle_2.18.0 
## 
## loaded via a namespace (and not attached):
##  [1] httr_1.4.2          viridisLite_0.3.0   bit64_4.0.5         jsonlite_1.7.1     
##  [5] modelr_0.1.8        gtools_3.8.2        assertthat_0.2.1    BiocManager_1.30.10
##  [9] highr_0.8           blob_1.2.1          cellranger_1.1.0    yaml_2.2.1         
## [13] pillar_1.4.7        backports_1.2.0     glue_1.4.2          digest_0.6.27      
## [17] rvest_0.3.6         colorspace_2.0-0    htmltools_0.5.1.1   plyr_1.8.6         
## [21] pkgconfig_2.0.3     broom_0.7.2         haven_2.3.1         bookdown_0.21      
## [25] scales_1.1.1        generics_0.1.0      farver_2.0.3        ellipsis_0.3.1     
## [29] withr_2.3.0         cli_2.2.0           magrittr_2.0.1      crayon_1.3.4       
## [33] readxl_1.3.1        memoise_1.1.0       evaluate_0.14       ps_1.4.0           
## [37] fs_1.5.0            fansi_0.4.1         xml2_1.3.2          blogdown_1.1.7     
## [41] tools_4.0.4         hms_0.5.3           lifecycle_0.2.0     munsell_0.5.0      
## [45] reprex_0.3.0        compiler_4.0.4      caTools_1.18.1      rlang_0.4.8        
## [49] grid_4.0.4          rstudioapi_0.13     htmlwidgets_1.5.2   crosstalk_1.1.0.1  
## [53] bitops_1.0-6        labeling_0.4.2      rmarkdown_2.5       gtable_0.3.0       
## [57] codetools_0.2-18    DBI_1.1.0           reshape2_1.4.4      R6_2.5.0           
## [61] lubridate_1.7.9.2   bit_4.0.4           utf8_1.1.4          KernSmooth_2.23-18 
## [65] stringi_1.5.3       Rcpp_1.0.5          vctrs_0.3.5         dbplyr_2.0.0       
## [69] tidyselect_1.1.0    xfun_0.20
</code></pre>
<h2 id="references">References</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Lawrence2013-kt" class="csl-entry">
<p>Lawrence, Michael, Wolfgang Huber, Hervé Pagès, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin T Morgan, and Vincent J Carey. 2013. “Software for Computing and Annotating Genomic Ranges.” <em>PLoS Comput. Biol.</em> 9 (8): e1003118. <a href="https://doi.org/10.1371/journal.pcbi.1003118">https://doi.org/10.1371/journal.pcbi.1003118</a>.</p>
</div>
</div>

	
	
	<div class="text-muted mt-5 pt-3 border-top">Last modified 2021-05-15: <a  href="https://github.com/tgirke/GEN242//commit/880bf3b75bfee47401cda87d5d4cae39c03952aa">some edits (880bf3b75)</a>
</div>
</div>


              </main>
          


        </div>
      </div>
      
<footer class="bg-dark py-5 row d-print-none">
  <div class="container-fluid mx-sm-5">
    <div class="row">
      <div class="col-6 col-sm-4 text-xs-center order-sm-2">
        
      </div>
      <div class="col-6 col-sm-4 text-right text-xs-center order-sm-3">
        
      </div>
      <div class="col-12 col-sm-4 text-center py-2 order-sm-2">
        <small class="text-white">&copy; 2021 The GEN242 Instructors All Rights Reserved</small>
        <small class="ml-1"><a href="https://www.ucr.edu/privacy.html" target="_blank">Privacy Policy</a></small>
	
      </div>
    </div>
  </div>
</footer>


    </div>
    
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>


<script>
  (function() {
    var i, text, code, codes = document.getElementsByTagName('code');
    for (i = 0; i < codes.length;) {
      code = codes[i];
      if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
        text = code.textContent;
        if (/^\$[^$]/.test(text) && /[^$]\$$/.test(text)) {
          text = text.replace(/^\$/, '\\(').replace(/\$$/, '\\)');
          code.textContent = text;
        }
        if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
            /^\$(.|\s)+\$$/.test(text) ||
            /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
          code.outerHTML = code.innerHTML;  
          continue;
        }
      }
      i++;
    }
  })();
</script>

<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script>













<script src="../../../js/main.min.f820bfc507c3ebefe17f8881ec4ee10013f3b89ad9555a50acf9bb1771e6823f.js" integrity="sha256-&#43;CC/xQfD6&#43;/hf4iB7E7hABPzuJrZVVpQrPm7F3Hmgj8=" crossorigin="anonymous"></script>



<script src='../../../js/prism.js'></script>



  </body>
</html>
